## Compared Transcript Quantification from multiple STAR/RSEM mapping and quantification approaches

## Overview
I was having trouble getting the RSEM transcript quantification command to work properly with STAR produced bam files. The error created mentioned that there were unequal transcripts identified in the STAR bam file vs. the transcriptomic index created for RSEM (I used the star mapper to create this index initially, and used the same GTF files to produce both the STAR and RSEM index folders). 

To fix this I decided to try three things. 

* 1) Use a GTF file that only uses transcripts identified by GNOMON. This excludes ~ 24-50 identified by refSeq. These tended to be poorly annotated anyways.

* 2) I tried running STAR using either an index created using the STAR command for generating a genomic index or the RSEM command with the star flag. 

* 3) I used the one step RSEM command with the RSEM index using the same GTF described above to perform the STAR mapping and RSEM quantification in a single command.

**Table**
![](https://github.com/epigeneticstoocean/2017OAExp_Oysters/blob/master/notebook/img/RSEM_methodComp_table_20190726.png)

[Link to full analysis]()

## Conclusions
It doesn't seem to really matter which approach I used. All methods should high degrees of correlation. Specifically, using either index (one generated by STAR or RSEM) for the STAR mapping step didn't change the results (they were perfectly correlated). There were some subtle differences in the the one-step process, this could be driven by the small changes in the default criterion that are present in the STAR mapping step when we run it separately. This options aren't aviable in the RSEM one liner and so im guessing it reverts to the default values. **Overall, I think going forward I will plan to keep the two steps separate and use the single RSEM created index to avoide and confusion.** This will also enable the bam files mapped to the genome by STAR to be used separately (something the one liner doesn't generate).
