# 2 Step STAR pipeline with BOTH RSEM and salmon transcript quantification

## Inputs
* Raw files after adapter removal and QC step
* GTF file with only features identified with `Gnomon` on NCBI database.
* Index / `.fa` file used for `STAR`, `RSEM`, and `salmon` generated using `` command from `RSEM`.

## Major Steps
* 2 Pass `STAR` alignment
  * 1 pass - preliminary mapping and identify novel junctions
  * 2 pass - remap but include discovered junction for improved mapping
* `RSEM` using `.bam` files with transcript coordinates output from `STAR`
  * Command ` rsem-calculate-expression`: defaults used and a bam file was provided rather than doing the mapping in a single step.
* `salmon` using `.bam` files with transcript coordinates output from `STAR`
  * Command `salmon quant` : alignment mode was used and the STAR bam file was provided.
* Gene count aggregation
  * Custom script was used to aggregate the transcript quantifications from STAR output.
  * `Tximport` package (R) was used in a custom script for salmon to create gene and isoform level count matrices, custom `TXimport` R objects and the `DESeqDataSetFromTximport()` function was used to convert the `Tximport` object into a `DESeqDataSet` object for downstream `DESeq2` analysis.

## Outputs

Base Directory : `/shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/run_20190726_RSEMGnomonIndex`
* `/m2` : outputs from STAR 1 Pass
* `/m3` : outputs from STAR 2 Pass
* `/RSEM` : outputs from RSEM quantification including quantification matrices
* `/salmon`: folders for each sample with `.quant.sf` files for each quantification.
  * ex : `/salmon/17005/quant.sf`
* `/salmon_outputs` :  `Tximport` and `DESeq2` R objects for downstreamt analysis based on salmon quantification.

**NOTE**: there should a `README.txt` file in each of these folder with a more complete description of the contents.

### Index folder generation

**Script on Server**: `/shared_lab/20180226_RNAseq_2017OAExp/RNA/scripts/STAR_scripts/RSEM_createRefFromStar.sh`

```
!#/bin/bash

rsem-prepare-reference \
--gtf /shared_lab/20180226_RNAseq_2017OAExp/RNA/references/gene_annotation/KM_CV_genome_edit_Gnomon.gtf \
--star \
-p 8 \
/shared_lab/20180226_RNAseq_2017OAExp/RNA/references/genome/GCF_002022765.2_C_virginica-3.0_genomic.fna \
/shared_lab/20180226_RNAseq_2017OAExp/RNA/references/RSEM_gnomon/RSEM_gnomon
```

## STAR two pass

**First Pass - Script on Server** : `/shared_lab/20180226_RNAseq_2017OAExp/RNA/scripts/STAR_scripts/STAR_1Pass_all.sh`
```
#!/bin/bash

echo "Please put in raw file directory:"
read raw
echo "Please put in name of new folder for output"
read output

base="/shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/"

echo "Outputs saving to : " $base$output

if [ -d "$base$output" ]; then
    echo "Directory Already Exists, please use another name"
else
    echo "Directory Created"
    mkdir "$base$output"
fi

echo "Processing the following samples: "
echo ls $raw/*.R1.*

# This will loop through each sample in the raw folder directory
for i in $( ls $raw/*.R1.* ); do
        for j in $( ls $raw/*.R2.* ); do
                file1=$( echo $i | rev | cut -d'/' -f 1 | rev | cut -d'.' -f 1)
                file2=$( echo $j | rev | cut -d'/' -f 1 | rev | cut -d'.' -f 1)

                if [ "$file1" == "$file2" ]
                then
                    	echo $i and $j
                        echo RNA"$file1"_m2
                        echo yes these files match
                        /shared_lab/scripts/STAR --runThreadN 10 \
                        --genomeDir /shared_lab/20180226_RNAseq_2017OAExp/RNA/references/RSEM_gnomon \
                        --outFilterMatchNminOverLread 0.17 --outFilterScoreMinOverLread 0.17 \
                        --readFilesIn $i $j \
                        --outSAMmapqUnique 40 \
                        --outSAMtype BAM Unsorted SortedByCoordinate \
                        --outFileNamePrefix $base$output/"$file1"_m2_ \
                        --readFilesCommand zcat
                fi
        done
done
```

**Second Pass - Script on Server** : `/shared_lab/20180226_RNAseq_2017OAExp/RNA/scripts/STAR_scripts/STAR_2Pass_all.sh`
```
#!/bin/bash

echo "Please put in raw file directory:"
read raw
#echo "Please put in name of new folder for output"
#read output

#base="/shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/"

#finaloutput=$base$output/m3
#m2=$base$output/m2

#echo "Outputs saving to:" $finaloutput

#echo ls "$m2"

#if [ -d "$finaloutput" ]; then
#    echo "Directory Already Exists, please use another name"
#    exit 1
#else
#	if [ -d "$base$output/m2"]; then
#               echo "Directory Created"
#               mkdir "$finaloutput"
#	else
#               echo "First pass output doesn't exist, make sure the correct output was specified or the 1Pass script was run first."
#               exit 1
#	fi
#fi

for i in $( ls $raw/*.R1.* ); do
        for j in $( ls $raw/*.R2.* ); do
                file1=$( echo $i | rev | cut -d'/' -f 1 | rev | cut -d'.' -f 1)
                file2=$( echo $j | rev | cut -d'/' -f 1 | rev | cut -d'.' -f 1)
                if [ "$file1" == "$file2" ]
                then
                    	m2_files=$( ls /shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/run_20190726_RSEMGnomonIndex/m2/*m2_SJ.out.tab)
                        echo $i and $j
                        echo RNA"$file1"_m3
                        echo yes these files match
                        echo $m2_files
                        /shared_lab/scripts/STAR --runThreadN 19 \
                        --genomeDir /shared_lab/20180226_RNAseq_2017OAExp/RNA/references/RSEM_gnomon \
                        --readFilesIn $i $j \
                        --outSAMmapqUnique 40 \
                        --outSAMtype BAM Unsorted SortedByCoordinate \
                        --quantMode TranscriptomeSAM GeneCounts --limitSjdbInsertNsj 1500000 \
                        --outFileNamePrefix /shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/run_20190726_RSEMGnomonIndex/m3/"$file1"_m3_ \
                        --readFilesCommand zcat \
                        --sjdbFileChrStartEnd $m2_files
                fi
        done
done
```

## RSEM

**Script on Server**: `/shared_lab/20180226_RNAseq_2017OAExp/RNA/scripts/STAR_scripts/RSEM_calcExpBam_range.sh`

```
#!/bin/bash

counter=1

echo "Start sample # (1-24)"
read lower
echo "End sample # (1-24)"
read upper

for i in $( ls /shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/run_20190726_RSEMGnomonIndex/m3/*toTranscriptome.out.bam); do
        file=$( echo $i | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 2)

        if [ "$counter" -ge "$lower" ]
        then

            	if [ "$counter" -le "$upper" ]
                then

                    	echo "Sample $counter"
                        echo "File Path : $i"
                        echo "File Name : $file"

                        rsem-calculate-expression --alignments --paired-end --output-genome-bam -p 20 \
                        $i \
                        /shared_lab/20180226_RNAseq_2017OAExp/RNA/references/RSEM_gnomon/RSEM_gnomon \
                        /shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/run_20190726_RSEMGnomonIndex/RSEM/"$file"_

                        echo "Output : /shared_lab/20180226_RNAseq_2017OAExp/RNA/STAR_files/run_20190726_RSEMGnomonIndex/RSEM/"$file"_"
                fi
        fi

	counter=$((counter+1))
done
```

**NOTE**: This script was written to perfor RSEM quantification in a specific range of sampels, allowing me to run it multiple times and specify a new range of samples. I did this because the RSEM quantification procss was taking a long time, and running the samples in parallel greatly accelerated the processing time.

## salmon

**Script on Server** : `/shared_lab/20180226_RNAseq_2017OAExp/RNA/scripts/salmon_scripts/quant_Salmon_AlignmentMode.sh`

```
#!/bin/bash

cd $1;
mkdir $2;

echo "Creating folder $3 for output"
echo "Saving in directory $2"
echo "Retrieving fastq files from $1 directory"

for fn in $(ls m3/*Aligned.toTranscriptome.out.bam);
do
  	samp=$(echo ${fn} | rev | cut -d'/' -f 1 | rev | cut -d'_' -f 2)
        echo "Processing Sample ${samp}"

        salmon quant -t /shared_lab/20180226_RNAseq_2017OAExp/RNA/references/RSEM_gnomon/RSEM_gnomon.transcripts.fa \
        -l A \
	-a m3/P1_${samp}_m3_Aligned.toTranscriptome.out.bam \
        -p 20 \
        --rangeFactorizationBins 4 \
        --numBootstraps 1000 \
        --gcBias \
        --seqBias \
        -o $2/${samp}
done

```
