---
title: "01D_CV17_Alignment_Filtering"
author: "adowneywall"
date: "March 5, 2019"
output: html_document
---

### Filter Reads that Poorly Mapped to Genome
To filter reads that poorly mapped, you can either run the following code each individual:
```{bash, eval=F}
samtools view -@64 -q4 -h -F 0x100 -F 0x400 RNA17005_m3Aligned.sortedByCoord.out.bam| mawk '$6 !~/[8-9].[SH]/ && $6 !~ /[1-9][0-9].[SH]/' | samtools view -b > RNA17005.q4.bam 2> RNA17005.q4_error.log
```
  
**The arguments:**  
- ```-q4```   - Skip alignments with MAPQ smaller than INT [0]. 
- ```-h```    - argument specifies to include header in output
- ```-F```    - Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0]. Specifies to not output alignments with with any flag between 100 and 400  
- ```mawk```  - line filters out reads that had significant hard or soft clipping >79 bp.  
- ```b```     - Output in the BAM format. 

Script for filtering reads from all files within a directory:
```{bash, eval=F}
#!/bin/bash

for i in $( ls RNA*m3Aligned.sortedByCoord.out.bam ); do
        ind=$( echo $i | cut -d'_' -f 1)
        samtools view -@64 -q4 -h -F 0x100 -F 0x400 "$i"| \
        mawk '$6 !~/[8-9].[SH]/ && $6 !~ /[1-9][0-9].[SH]/' | \
        samtools view -b > "$ind".q4.bam 2> "$ind".q4_error.log
done
```


