---
title: "03D_salmon_DiffExpression_comparison"
author: "adowneywall"
date: "6/17/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#source("http://bioconductor.org/biocLite.R")
if(!require("edgeR")) biocLite("edgeR")
if(!require("limma")) biocLite("limma")
library(edgeR)
library(limma)
library(ggplot2)
library(psych)
```

## Data

```{r}
gene_list <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/results/Salmon_Gene/run20190610_geneList_DESeq2Results.RData")
tran_list <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/results/Salmon_Gene/run20190610_tranList_DESeq2Results.RData")

# Meta Data
model<-read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Time <- as.character(model$timepoint)
model$Time[model$Time == "3"] <- "09"
model$Time[model$Time == "6"] <- "80"
model$Time <- as.factor(model$Time)
model$Pop <- as.factor(model$population)
model$Lane <- as.factor(model$lane)
model$SFV <-interaction(model$Time,model$Treatment) # Creates single factor variable for combination of time and treatment

# Transcript abundances (only needed for timepoint specific analysis)
ta <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/salmon_pipeline/run20190610_counts/tranMatrixCounts.RData")
# Round counts before create DESeq matrix
ta_round <- ta #round(ta,digits=0)
ta_filter <- ta_round[rowSums(ta_round)>80,]

#cpm_out <- cpm(ta_filter)
#cpm <- ta_filter[which(rowMins(cpm_out)>0.5),]

# Additional info about each gene
# DESeq Data Object created directly from tximport using `DESeqDataSetFromTximport()` function.
t_obj <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/salmon_pipeline/run20190610_counts/tranDESeqDataObj.RData")
# Transcriptome annotations
trans <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/transcriptome_table.RData")
```

```{r}
f <- paste(paste0("CO2_", model$treatment), paste0("T_",model$timepoint),sep=".")
f <- factor(f)

design <-model.matrix(~0+f)
colnames(design) <- levels(f)

dge <- DGEList(counts=ta_filter, group=f)
dgeNorm<- calcNormFactors(dge,method=("TMM"))

v1 <- voomWithQualityWeights(dgeNorm, design=design, lib.size=dgeNorm$samples$lib.size, normalize.method="none")
```

```{r}
ph_P <- NULL
#nrow(v1$E)
for(i in 1:nrow(v1$E)){
  temp <- glm(model$epf_pH~v1$E[i,]+model$Pop+model$Lane)
  temp_sum <- summary(temp)
  ph_P <- c(ph_P,temp_sum$coefficients[2,4])
}
ph_BH <- p.adjust(ph_P,method = "BH")
min(ph_BH)
hist(ph_BH)
```
