---
title: "01E_CV17_Alignment_DataVisualization"
author: "adowneywall"
date: "March 5, 2019"
output: html_document
---

##Additional code to visualize overall mapping efficiency

#### Create bed files to compare coverage after mapping reads
Make sure BEDOPS is installed on your system (see here: https://github.com/bedops for installation instructions). Convert gff file to bed file and then reduce to a sorted bed file with only exons.
```{bash, eval=FALSE}
gff2bed < ref_C_virginica-3.0_top_level.gff3 > ref_C_virginica-3.0_top.bed
mawk '$8 ~ /exon/' ref_C_virginica-3.0_top.bed > ref3.0.exon.bed
bedtools sort -i ref3.0.exon.bed -faidx <(cut -f1 genome.file) > sorted.ref3.0.exon.bed
```
Remove duplicates and concatenate overlapping intervals from transcript variants
```{bash, eval=FALSE}
bedtools merge -i sorted.ref3.0.exon.bed > sorted.ref3.0.exon.sc.bed
```
Create bed file for all gene regions
```{bash, eval=FALSE}
mawk '$8 ~ /gene/' ref_C_virginica-3.0_top.bed > ref3.0.gene.bed
bedtools sort -i ref3.0.gene.bed -faidx <(cut -f1 genome.file) > sorted.ref3.0.gene.bed
bedtools merge -i sorted.ref3.0.gene.bed > sorted.ref3.0.gene.sc.bed
```
Create bed file for intergenic, intron, non-coding, and CDS regions
```{bash, eval=FALSE}
bedtools complement -i sorted.ref3.0.gene.bed -g genome.file -sorted | bedtools subtract -b sorted.ref3.0.exon.sc.bed -a - > cv.ref3.intergenic.bed
bedtools complement -i sorted.ref3.0.exon.sc.bed -g genome.file -sorted > cv.ref3.noncoding.bed
bedtools intersect -a cv.ref3.noncoding.bed -b sorted.ref3.0.gene.sc.bed -sorted > cv.ref3.intron.bed
mawk '$8 ~ /CDS/' ref_C_virginica-3.0_top.bed > ref3.0.CDS.bed
bedtools sort -i ref3.0.CDS.bed -faidx <(cut -f1 genome.file) > sorted.ref3.0.CDS.bed
bedtools merge -i sorted.ref3.0.CDS.bed > sorted.ref3.0.CDS.sc.b
bedtools sort -i sorted.ref3.0.CDS.sc.b -faidx <(cut -f1 genome.file) > sorted.ref3.0.CDS.sc.bed
```
Create bed for untranslated regions (UTRs) of exons
```{bash, eval=FALSE}
bedtools subtract -a sorted.ref3.0.exon.bed -b sorted.ref3.0.CDS.bed -g genome.file -sorted > sorted.ref3.0.UTR.bed
bedtools merge -i <(bedtools sort -i sorted.ref3.0.UTR.bed -faidx <(cut -f1 genome.file))> sorted.ref3.0.UTR.sc.b
bedtools sort -i sorted.ref3.0.UTR.sc.b -faidx <(cut -f1 genome.file) > sorted.ref3.0.UTR.sc.bed
```
Create bed file for mtDNA
```{bash, eval=FALSE}
mawk '$1 ~ /NC_007175.2/' ref_C_virginica-3.0_top.bed > mtDNA.bed
```

Merge all alignment files into a single merged file
```{bash, eval=FALSE}
samtools merge -@64 m4.merged.bam RNA17005_m3Aligned.sortedByCoord.out.bam RNA17007_m3Aligned.sortedByCoord.out.bam RNA17013_m3Aligned.sortedByCoord.out.bam RNA17019_m3Aligned.sortedByCoord.out.bam RNA17069_m3Aligned.sortedByCoord.out.bam RNA17070_m3Aligned.sortedByCoord.out.bam RNA17072_m3Aligned.sortedByCoord.out.bam RNA17079_m3Aligned.sortedByCoord.out.bam RNA17090_m3Aligned.sortedByCoord.out.bam RNA17094_m3Aligned.sortedByCoord.out.bam RNA17099_m3Aligned.sortedByCoord.out.bam RNA17108_m3Aligned.sortedByCoord.out.bam RNA17122_m3Aligned.sortedByCoord.out.bam RNA17130_m3Aligned.sortedByCoord.out.bam RNA17142_m3Aligned.sortedByCoord.out.bam RNA17145_m3Aligned.sortedByCoord.out.bam RNA17162_m3Aligned.sortedByCoord.out.bam RNA17174_m3Aligned.sortedByCoord.out.bam RNA17176_m3Aligned.sortedByCoord.out.bam RNA17178_m3Aligned.sortedByCoord.out.bam RNA17181_m3Aligned.sortedByCoord.out.bam RNA17203_m3Aligned.sortedByCoord.out.bam RNA17211_m3Aligned.sortedByCoord.out.bam RNA17213_m3Aligned.sortedByCoord.out.bam
```

Again, filter reads that did not uniquely map to the genome from this merged file
```{bash, eval=FALSE}
samtools view -@64 -q4 -h -F 0x100 -F 0x400 m4.merged.bam| mawk '$6 !~/[8-9].[SH]/ && $6 !~ /[1-9][0-9].[SH]/' | samtools view -b > m4.q4.merged.bam
```

Now we want to look at coverage of our reads across the genome. To do so, first create links to the bed files in the genome features file (GFF) we downloaded earlier to see can look at coverage for various genome regions.
```{bash, eval=FALSE}
ln -s /shared_lab/20180226_RNAseq_2017OAExp/Genome/GFF/*.bed .
ln -s /shared_lab/20180226_RNAseq_2017OAExp/Genome/GFF/genome.file .
```

Then, let's get some basic numbers on the number of reads mapping, the number mapping to genes, and the number mapping to exons:
```{bash, eval=FALSE}
paste <(samtools view -@32 -c m4.q4.merged.bam) <(samtools view -@32 -c -L sorted.ref3.0.gene.bed m4.q4.merged.bam) <(samtools view -@32 -c -L sorted.ref3.0.exon.sc.bed m4.q4.merged.bam)
```

Next we'll calculate per base pair coverage across genomic regions and output the data in txt format:
```{bash, eval=FALSE}
#introns
bedtools coverage -hist -b m4.q4.merged.bam -a cv.ref3.intron.bed -g genome.file -sorted  -split | grep ^all > AllRNAm4q4.hist.AllIntron.all.split.txt
#intergenic
bedtools coverage -hist -b m4.q4.merged.bam -a cv.ref3.intergenic.bed -g genome.file -sorted -split | grep ^all > AllRNAm4q4.hist.AllIntergenic.all.split.txt
#exons (genes-introns)
bedtools coverage -hist -b m4.q4.merged.bam -a sorted.ref3.0.exon.sc.bed -g genome.file -sorted  -split | grep ^all > AllRNAm4q4.hist.AllExon.all.split.txt
#untraslated regions
bedtools coverage -hist -b m4.q4.merged.bam -a sorted.ref3.0.UTR.sc.bed -g genome.file -sorted  -split | grep ^all > AllRNAm4q4.hist.allUTR.all.split.txt
#coding sequences (genes-introns-UTR)
bedtools coverage -hist -b m4.q4.merged.bam -a sorted.ref3.0.CDS.sc.bed -g genome.file -sorted  -split | grep ^all > AllRNAm4q4.hist.AllCDS.all.split.txt
```

Next we'll use an Rscript to plot the coverage across genome regions. This was modified from the original code on the EecSeq GitHub page:
```{r, eval=FALSE}
setwd("/home/downeyam/Github/2017OAExp_Oysters/RNA")
library(ggplot2)
library(grid)
library(plyr)
library(dplyr)
library(scales)
library(zoo)

print(files <- list.files(pattern="AllRNAm4q4.hist."))

labs <- c("CDS","Exon","Intergenic","Intron","UTR")


cbPalette <- c("#D55E00", "#009E73", "#56B4E9" ,"#0072B2" ,"#E69F00" ,"#F0E442" ,"#999999" ,"#CC79A7","#7570B3")

cov <- list()
for (i in 1:length(files)) {
  cov[[i]] <- read.table(files[i])[,c(2,5)]
  cov_cumul=1-cumsum(cov[[i]][,2])
  cov[[i]]$cov_cumul <- c(1,cov_cumul[-length(cov_cumul)])
  cov[[i]]$sample=labs[i]
}

cov_df=do.call("rbind",cov)
names(cov_df)[1:2]=c("depth","fraction")

pcbPalette <- c("#009E73" ,"#D55E00","#CC79A7","#0072B2", "#56B4E9")

p <- ggplot(cov_df, aes(x= depth, y=cov_cumul, color=sample))  + xlim(0,100)+
  scale_alpha(guide = 'none') +
  geom_line(size=1.5)+ 
  #geom_segment(aes(x=20, y=0, xend=20, yend=1, color="red"))+
  scale_color_manual(values=pcbPalette) +
  scale_fill_manual(values=pcbPalette) +
  ylab("% of Bases > Depth")+
  xlab("Depth")+
  theme_bw() +
  theme(plot.title = element_text(size = 12, face = "bold",hjust = 0.5)) +
  theme(legend.title = element_blank()) + 
  theme(legend.position=c(0.75,0.75))

png(filename="Figure2_BFedited.png", type="cairo",units="px", width=5600, 
    height=3000, res=600, bg="transparent")

dev.off()
```