---
title: "Target Gene Queries"
author: "adowneywall"
date: "4/22/2019"
output: 
  html_document:
     keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
```

## Target Gene Overview  
  
Several target proteins have been thought to be associated with oyster calcification in the literature. Unfortunately, these didn't appear to be significantly associated with OA or time in our experiment when we perform a 'blind' association test. Still, it is useful to see if this is due to a lack of putative gene features coding for these proteins are present within out count data or if there gene features a simply not correlated with environment or time.  

Gene features associated with the proteins of interest were identified by perform a target search of the eastern oyster genome using the Genome View tool in NCBI and quering each protein name. All putative gene locations were used.  

#### Target proteins:  
  
* **Calcitonin**: Involved in calcium metabolism and regulation of calcium (studied mostly in context of osmoregulation)  
  * Papers:  
    * Molecular and physiological characterization of an invertebrate homologue of a calcitonin-related receptor (Dubos et al. 2003)  
    * Osmoregulation and mRNA expression of calcitonin-related receptor in the Pacific oyster Crassostrea gigas (Jo et al. 2008)  
    
* **Nacrein**: Involved I shell formation, have carbonic anhydrase like domains. Observed in Pacific oyster mantle.  
  * Paper: Identification two novel nacrein-like proteins involved in the shell formation of the Pacific oyster Crassostrea gigas (Song et al. 2014)  
  
* **Carbonic Anhydrase**: Important roles in biomineralization process.  
  * Papers:  
    * A shell-formation related carbonic anhydrase in Crassostrea gigas modulates intracellular calcium against CO2 exposure: Implication for impacts of ocean acidification on mollusk calcification (Wang et al. 2017)  
    * A Carbonic Anhydrase Serves as an Important Acid-Base Regulator in Pacific Oyster Crassostrea gigas Exposed to Elevated CO2: Implication for Physiological Responses of Mollusk to Ocean Acidification (Wang et al. 2017)  
    
* **Sodium bicarbonate transporter**: Acid -base regulator, responsible for modulating intra and inter cellular pH.  

* **Adenylyl cyclase**: Involved with pH regulation and responsive to OA in Cras. Gigas oysters  
  * Paper: Ocean acidification stimulates alkali signal pathway: A bicarbonate sensing soluble adenylyl cyclase from oyster Crassostrea gigas mediates physiological changes induced by CO2 exposure (Wang et al 2016 )  
  
### Data  
```{r data}
GeneCounts <- read.delim("/home/downeyam/Github/2017OAExp_Oysters/results/C_virginica_gene_count_final.txt",header=TRUE,sep="",row.names=1)
counts <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/Normalized_CountMatrix/Scenario1_normalizedVoom.RData")
model <-  read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Day <- as.factor(model$Day)
```
  
### Calcitonin  
```{r calcitonin}
# Had to use GeneCount matrix because this locus was filtered during countAnalysis

#Test if kept in trimmed data 
match("LOC111112092",row.names(counts$E))
#Nope

# Check if present in whole matrix
cal_text <- GeneCounts[match("LOC111112092",row.names(GeneCounts)),]
sum(cal_text)
# Nope not present

```
No reads seemed to align to location :/
  
### Nacrein  
```{r nacrein}
#Test if kept in trimmed data 
(nac_text <- counts$E[match("LOC111136431",row.names(counts$E)),])
# Yes its retained!

nac_data <- data.frame(Treatment=model$Treatment,Day=model$Day,EPF=model$epf_pH,locus=(nac_text))
nac_data
ggplot(nac_data,aes(x=interaction(Treatment,Day),y=locus)) + 
  geom_boxplot() +
  labs(title="LOC111136431",x="Treatment.Day Combination")  

# ggplot(nac_data,aes(x=EPF,y=LOC111136431)) + 
#   geom_point() +
#   labs(title="LOC111136431",x="EPF pH")
```
No significant differences between treatment_time combinations.  

### Carbonic Anhydrase  
```{r carbonic_anhydrase}
# All potential carbonic anhydrase gene locations based on CV genome on NCBI 
ca_LOC <- c("LOC111134699",
  "LOC111133036",
  "LOC111130107",
  "LOC111134700",
  "LOC111123083",
  "LOC111113402",
  "LOC111137424",
  "LOC111117514",
  "LOC111114059",
  "LOC111112162",
  "LOC111133640",
  "LOC111134104",
  "LOC111133247",
  "LOC111130106",
  "LOC111127254",
  "LOC111127093",
  "LOC111122701",
  "LOC111120606",
  "LOC111120607",
  "LOC11125708")

# Subset for locations matching known carbonic anhydrase genes
match(ca_LOC,row.names(counts),nomatch = FALSE)
out_ca <- match(ca_LOC,row.names(counts))
out_ca <- out_ca[!is.na(out_ca)]
ca_data <- data.frame(t(as.matrix(counts[out_ca,])))
ca_data$Treatment <- model$Treatment
ca_data$Day <- model$Day

ca_data

#LOC111120607 
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=LOC111120607)) + geom_boxplot()
#LOC111134700 
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=LOC111134700)) + geom_boxplot()
#LOC111134104 
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=LOC111134104 )) + geom_boxplot()
#LOC111133247 
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=LOC111133247)) + geom_boxplot()

# Q-vals for locus from linear mixed model
#### UPDATED THIS
# Treatment:Time
#time_qvals[1732,1]
# Treatment:Time
#treatment_qvals[1732,1]
# Treatment:Time
#timeTrt_qvals[1732,1]
```
  
### Sodium Bicarbonate Transporter  
```{r sodium_bicarbonate_trans}
bi_LOC <- c("LOC111118314",
  "LOC111125089",
  "LOC111124629",
  "LOC111126568",
  "LOC111126835",
  "LOC111123835",
  "LOC111120445",
  "LOC111120427",
  "LOC111119272",
  "LOC111115910",
  "LOC111126460")

match(bi_LOC,row.names(counts),nomatch = FALSE)
out_bi <- match(bi_LOC,row.names(counts))
out_bi <- out_bi[!is.na(out_bi)]
bi_data <- data.frame(t(as.matrix(counts[out_bi,])))
bi_data$Treatment <- model$Treatment
bi_data$Day <- model$Day
bi_data

ggplot(bi_data,aes(x=interaction(Treatment,Day),y=LOC111124629)) + geom_boxplot()
ggplot(bi_data,aes(x=interaction(Treatment,Day),y=LOC111126568)) + geom_boxplot()
ggplot(bi_data,aes(x=interaction(Treatment,Day),y=LOC111123835)) + geom_boxplot()
```

### Adenylyl cyclase  
```{r Adenylyl cyclase}
 
ac_LOC <- c("LOC111131100",
  "LOC111134509",
  "LOC111130107",
  "LOC111111874",
  "LOC111109301")

match(ac_LOC,row.names(counts),nomatch = FALSE)
out_ac <- match(ac_LOC,row.names(counts))
out_ac <- out_ac[!is.na(out_ac)]
ac_data <- data.frame(t(as.matrix(counts[out_ac,])))
ac_data$Treatment <- model$Treatment
ac_data$Day <- model$Day
ac_data

ggplot(ac_data,aes(x=interaction(Treatment,Day),y=LOC111131100)) + geom_boxplot()
ggplot(ac_data,aes(x=interaction(Treatment,Day),y=LOC111134509)) + geom_boxplot()
ggplot(ac_data,aes(x=interaction(Treatment,Day),y=LOC111109301)) + geom_boxplot()

#ggplot(ac_data,aes(x=EPF,y=LOC111131100,colour=interaction(Treatment,Time))) + geom_point()
#ggplot(ac_data,aes(x=EPF,y=LOC111109300,colour=interaction(Treatment,Time))) + geom_point()
```

### Calbindin - 32  
```{r Calbindin}
(cb_text <- counts$E[match("LOC111136252",row.names(counts$E)),])
# Yes its retained!
cb_data <- data.frame(Treatment=model$Treatment,Day=model$Day,EPF=model$epf_pH,locus=(cb_text))

ggplot(cb_data,aes(x=interaction(Treatment,Day),y=locus)) + 
  geom_boxplot() + theme_classic() + 
  labs(x="Treatment.Day Combination",y="Gene Expression (log2)")

#ggsave("/home/downeyam/Github/2017OAExp_Oysters/figures/Calbindin_Trt.png",width=20,height=14,units = "cm")
# ggplot(cb_data,aes(x=EPF_dff,y=locus,colour=Day)) + geom_smooth(method = "lm") +
#   geom_point(aes(shape=Treatment)) + 
#   labs(title="putative Calbindin-32 protein",x="EPF pH")
# ggsave("/home/downeyam/Github/2017OAExp_Oysters/figures/calbindin.png",width=20,height=12,units="cm")
```

### Beta-ureidopropionase  
```{r beta-ureidopropionase}
(bu_text <- counts$E[match("LOC111132068",row.names(counts$E)),])
# Yes its retained!
bu_data <- data.frame(Treatment=model$Treatment,Day=model$Day,EPF=model$epf_pH,locus=(bu_text))

ggplot(bu_data,aes(x=interaction(Treatment,Day),y=locus)) + 
  geom_boxplot() +
  labs(x="Treatment.Day Combination",y="Gene Expression (log2)")
#ggsave("/home/downeyam/Github/2017OAExp_Oysters/figures/Calbindin_Trt.png",width=20,height=14,units = "cm")
# ggplot(bu_data,aes(x=EPF_dff,y=locus,colour=Day)) + geom_smooth(method = "lm") +
#   geom_point(aes(shape=Treatment)) + 
#   labs(title="beta-ureidopropionase",x="EPF pH")
```