---
title: "DESeq2 Analysis"
author: "adowneywall"
date: "5/9/2019"
output: 
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: console
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(DESeq2)
library(ashr)
```

## About DESeq  

The DESeq2 model and all the steps taken in the software are described in detail in our publication ([Love, Huber, and Anders 2014](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8)), and we include the formula and descriptions in this section as well. The differential expression analysis in DESeq2 uses a generalized linear model of the form:
$$K_{ij} \sim NB(\mu_{ij},\alpha_i$$  
$$\mu_{ij}=s_jq_{ij}$$  
$$log_2(q_{ij})=x_j\beta_i$$
where counts $K_{ij}$ for gene $i$, sample $j$ are modeled using a negative binomial distribution with fitted mean $\mu_{ij}$ and a gene-specific dispersion parameter $\alpha_i$. The fitted mean is composed of a sample-specific size factor $s_j$ and a parameter $q_{ij}$ proportional to the expected true concentration of fragments for sample $j$. The coefficients $\beta_i$ give the log2 fold changes for gene i for each column of the model matrix $X$. Note that the model can be generalized to use sample- and gene-dependent normalization factors $s_{ij}$.  
  
The dispersion parameter $\alpha_i$ defines the relationship between the variance of the observed count and its mean value. In other words, how far do we expected the observed count will be from the mean value, which depends both on the size factor $s_j$ and the covariate-dependent part $q_{ij}$ as defined above.  
$$Var(K_{ij})=E[(K_{ij}-u_{ij})^2] = u_ij + \alpha_i\mu^2_{ij}$$  
An option in DESeq2 is to provide maximum a posteriori estimates of the log2 fold changes in $\beta_i$ after incorporating a zero-centered Normal prior (```betaPrior```). While previously, these moderated, or shrunken, estimates were generated by DESeq or ```nbinomWaldTest``` functions, they are now produced by the ```lfcShrink``` function. Dispersions are estimated using expected mean values from the maximum likelihood estimate of log2 fold changes, and optimizing the Cox-Reid adjusted profile likelihood, as first implemented for RNA-seq data in edgeR (Cox and Reid 1987,edgeR_GLM). The steps performed by the DESeq function are documented in its manual page ```?DESeq```; briefly, they are:

* estimation of size factors $s_j$ by ```estimateSizeFactors```
* estimation of dispersion $\alpha_i$ by ```estimateDispersions```
* negative binomial GLM fitting for $\beta_i$ and Wald statistics by ```nbinomWaldTest```

Description from the [```DESeq``` manual](https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#interactions)

**Data**  
```{r echo = FALSE}
GeneCounts <- as.matrix(read.delim("/home/downeyam/Github/2017OAExp_Oysters/results/C_virginica_gene_count_final.txt",sep = " "))
scenario1 <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/Filter_CountMatrix/scenario1_GeneCount.RData")
cts <- as.matrix(scenario1$countMatrix)
model<-read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Time <- as.character(model$timepoint)
model$Time[model$Time == "3"] <- "09"
model$Time[model$Time == "6"] <- "80"
model$Time <- as.factor(model$Time)
model$Pop <- as.factor(model$population)
model$Lane <- as.factor(model$lane)
model$SFV <-  interaction(model$Time,model$Treatment) # Creates single factor variable for combination of time and treatment
col_sel <- c("SFV") #,"Time"
coldata <- model #subset(model,select=col_sel)
row.names(coldata) <- colnames(cts)
```
  
## Filtered (non-normalized data)  
  
**Full Model** - Lane, Pop, Treatment, Time, Treatment:Time as factors  
```{r}
# Full Model
filtered_full <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ Lane + Pop + Treatment + Time + Treatment:Time)
dds_filtered_full <- DESeq(filtered_full)

# Interaction
res_filtered_full <- results(dds_filtered_full)
summary(res_filtered_full)
# Time
res_filtered_full_Time <- results(dds_filtered_full,contrast=c("Time","09","80"))
summary(res_filtered_full_Time)
# Treatment
res_filtered_full_Treatment <- results(dds_filtered_full,contrast=c("Treatment","400","2800"))
summary(res_filtered_full_Treatment)
# Pop
res_filtered_full_Pop <- results(dds_filtered_full,contrast=c("Pop","1","2"))
summary(res_filtered_full_Pop)
# Lane
res_filtered_full_Lane <- results(dds_filtered_full,contrast=c("Lane","1","2"))
summary(res_filtered_full_Lane)
```
  
**Just Fixed Factor Model** - Treatment, Time, Treatment:Time as factors  
```{r}
# Model with just fixed factors
filtered_fixed <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ Treatment + Time + Treatment:Time)
dds_filtered_fixed <- DESeq(filtered_fixed,test="LRT",reduced=~Time)
#Interaction
res_filtered_fixed <- results(dds_filtered_fixed)
summary(res_filtered_fixed)
# Time
res_filtered_fixed_Time <- results(dds_filtered_fixed,contrast=c("Time","09","80"))
summary(res_filtered_full_Time)
# Treatment
res_filtered_fixed_Treatment <- results(dds_filtered_fixed,contrast=c("Treatment","400","2800"))
summary(res_filtered_full_Treatment)

#Applying Shrinkage Terms -Ashr
#Interaction
Ashr_filtered_fixed <- lfcShrink(dds_filtered_fixed, coef=2, type="ashr")
summary(Ashr_filtered_fixed)
#Treatment
Ashr_filtered_fixed_Treatment <- lfcShrink(dds_filtered_fixed, coef=2, type="ashr",contrast = c("Treatment","400","2800"))
summary(Ashr_filtered_fixed_Treatment)
```
  
## Full Gene Count Matrix
  
**Full Model** - Lane, Pop, Treatment, Time, Treatment:Time as factors  
```{r}
whole_full <- DESeqDataSetFromMatrix(countData = GeneCounts,
                              colData = coldata,
                              design = ~ Lane + Pop + Treatment + Time + Treatment:Time)
dds_whole_full <- DESeq(whole_full)

# Diagnostic Plots
## Dispersion Plot
plotDispEsts(dds_whole_full)

# Interaction
res_whole_full <- results(dds_whole_full)
summary(res_whole_full)
ordered_val_full <- order(res_whole_full$padj)
plotCounts(dds_whole_full, gene=ordered_val_full[1], intgroup="SFV",normalized = FALSE)
# Treatment
res_whole_full_Treatment <- results(dds_whole_full,contrast=c("Treatment","400","2800"))
summary(res_whole_full_Treatment)
ordered_val_treatment <- order(res_whole_full_Treatment$padj)
plotCounts(dds_whole_full, gene=ordered_val_treatment[1], intgroup="SFV",normalized = FALSE)
```
  
**Just Fixed Factor Model** - Treatment, Time, Treatment:Time as factors  
```{r}
whole_fixed <- DESeqDataSetFromMatrix(countData = GeneCounts,
                              colData = coldata,
                              design = ~ Treatment + Time + Treatment:Time)
dds_whole_fixed <- DESeq(whole_fixed)
res_whole_fixed <- results(dds_whole_fixed)
summary(res_whole_fixed)

plotCounts(dds_whole_fixed, gene=which.min(res_whole_fixed$padj), intgroup="Pop")
```
  