---
title: "AE17_ExpressionPCA_withLoadings"
author: "adowneywall"
date: "9/10/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# If need be
library(BiocManager)
#BiocManager::install("rtracklayer")
library(rtracklayer)
library(dplyr)
library(factoextra) # for the fviz_eig function

wd <- "/home/downeyam/Github/2017OAExp_Oysters"
```

## Data Read in  

**Gene Expression**  
```{r}
## Gene Expression
RSEM <-  readRDS(paste0(wd,"/input_files/RNA/STAR_mapping/RSEM_output/RSEM_gene_Summary.Rdata"))
# Separate out RSEM counts and rename rows with LOC ID
rsem_c <- RSEM$Count # Stick with raw gene counts  (not TPM or some other normalizaed count)
rm(RSEM)
```

**Gene Annotation**  
```{r}
# Transcript file
tran <- readRDS(paste0(wd,"/input_files/RNA/references/STAR_gnomon_tximportGeneFile.RData"))
# Gene File
gene <- tran[!duplicated(tran$GENEID),]
gene$gene_length <- gene$stop-gene$start
```

*Sample meta data*  
```{r eval=TRUE}
#### Meta Data ####
meta <- readRDS(paste0(wd,"/input_files/meta/metadata_20190811.RData"))
meta$sampl_nameSimple <- substr(meta$sample_name,start = 4,stop=9)
#Create new factor levels (one for each level combination)
meta$SFVrn <- as.factor(paste0("D",meta$SFV))
meta$Sample_Index <- as.factor(meta$sample_index)
meta$TankID <- as.factor(meta$tankID)
```

**Gene Expression Data Manipulation**
```{r}
#### Data Manipulation ####
# Order genes from annotation list to match the order in the count matrix
gene_order <- gene[order(gene$gene_id),]
identical(rownames(rsem_c),gene_order$gene_id) # TRUE confirms the order matches
# Relabel the rows of the count matrix with LOC ID
rownames(rsem_c) <- gene_order$GENEID
geneC_all <- round(rsem_c)
rm(rsem_c)
```

**DNA Methylation Summary**
```{r}
## Gene level Methylation Summary - All CpGs - comp5 code
cpgMean <- read.table("/shared_lab/20180226_RNAseq_2017OAExp/DNAm/processed_samples/08_geneLevelSummary/DNAm_Data/meanBetaPerFeature.txt",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")

## CpG Annotation
# cpg <- read.table("/shared_lab/20180226_RNAseq_2017OAExp/DNAm/processed_samples/08_geneLevelSummary/DNAm_Data/CpGsAnnotated.bed",header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")

#Local machine code
# All cpgs 
cpgMean_all <-  read.table(paste0(wd,"/results/DNAm/meanBetaPerFeature.txt"),header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")
# CpGs with coverage of 5
# cpgMean <-  read.table(paste0(wd,"/results/DNAm/mean_betamean_CVTrtMeans_PerFeature_filteredBetameanGT05.txt"),header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")

cpgMean_all_gene <- cpgMean_all[cpgMean_all$V3 == "gene",]
cpgMean_gene <- cpgMean[cpgMean$V3 == "gene",]
rm(cpgMean)

## Quick tangent to determine # of exons 
cpgMean_all_exon <- cpgMean_all[cpgMean_all$V3 == "exon",]
gene_type_byExon <- sub(".*gbkey=(.*?);.*","\\1",cpgMean_all_exon$V9,perl=TRUE)
# Selecting out exons within genes we care about  (no t or rRNA)
tf_mRNA <- gene_type_byExon == "mRNA" | gene_type_byExon == "ncRNA" |  gene_type_byExon == "misc_RNA" |  gene_type_byExon == "exon"
#Pulling out geneID 
gene_ID_inExons <- sub(".*GeneID:(.[0-9]{1,8}).*","\\1",cpgMean_all_exon$V9,perl=TRUE)
gene_ID_inExons_filtered <- as.numeric(gene_ID_inExons[tf_mRNA])
gene_ID_inExons_filtered <- paste0("LOC",gene_ID_inExons_filtered)
cpgMean_all_exon$GENEID[tf_mRNA] <- gene_ID_inExons_filtered
cpgMean_all_exon$GENEID[!tf_mRNA] <-"NA"
# Generate Exon counts
exon_counts <- data.frame(GENEID=names(table(cpgMean_all_exon$GENEID)),Exon_Count=as.vector(table(cpgMean_all_exon$GENEID)))
# Count number of CpGs in exons
exon_cpg_counts <- aggregate(V11~GENEID,FUN=sum,data=cpgMean_all_exon)
# Take mean MEthylation for cpgs in exons
cpgMean_all_exon$V10 <- as.numeric(cpgMean_all_exon$V10)
exon_cpg_means <- aggregate(V10~GENEID,FUN=mean,data=cpgMean_all_exon,na.rm=TRUE)
# Merge these piece of data together
exons <- left_join(exon_counts,exon_cpg_counts)
exons <- left_join(exons,exon_cpg_means)
#remove the na
exons <- exons[!exons$GENEID == "NA",]
colnames(exons)[3:4] <- c("exon_cpg_counts","exon_cpg_means")

### Extract gene locations
# This is to merge DNA methylation data with gene expression data
#Total CpG data
gene_ID <- sub(".*GeneID:(.*?);.*","\\1",cpgMean_all_gene$V9,perl=TRUE)
LOC_ID <- paste0("LOC",gene_ID)
cpgMean_all_gene <- data.frame(GENEID=LOC_ID,cpgMean_all_gene)
# Filtered data
gene_ID <- sub(".*GeneID:(.*?);.*","\\1",cpgMean_gene$V9,perl=TRUE)
LOC_ID <- paste0("LOC",gene_ID)
cpgMean_gene <- data.frame(GENEID=LOC_ID,cpgMean_gene)

plot(cpgMean_all_gene$V10~cpgMean_gene$V10,
     xlim=c(0,1),ylim=c(0,1),
     xlab="Filter CpGs",
     ylab="Unfiltered CpGs",
     main="Gene Level DNA Methylation Percent")
abline(b=1,a=0,col="red",lwd=2)

# Ordered identically so we can bind them together
identical(cpgMean_gene$GENEID,cpgMean_all_gene$GENEID)

#Creating new data frame with counts an
cpg_final <- data.frame(cpgMean_gene[,c(1,11:13)],
                        cpgMean_total=cpgMean_all_gene[,11],
                        cpgCount_total=cpgMean_all_gene[,12])
colnames(cpg_final)[2:4] <- c("cpgMean_filtered","cpgCV_filtered","cpgCount_filtered")
sum(is.na(cpg_final$cpgCount_total))

### Merging gene expression, gene leve dnam, and exon level dnam
initial <- left_join(gene,cpg_final)
ge_dnam <- left_join(initial,exons)
ge_dnamV2 <- ge_dnam[!is.na(ge_dnam$cpgCount_total),]
out <- match(ge_dnamV2$GENEID,rownames(geneC_all))

geneC_new <- geneC_all[match(ge_dnamV2$GENEID,rownames(geneC_all)),]
identical(rownames(geneC_new),ge_dnamV2$GENEID)


```

## Removing poorly covered genes (same threshold as differential expression)
```{r}
### Genes ### 
# Breaking down expression coverage by treatment*time combination
#Day 9 Trt 2800
keep_D9.2800 <- rowSums(cpm(geneC_new[,meta$SFVrn=="D09.2800"])>=1) >= 5
sum(keep_D9.2800)
#Day 9 Trt 400
keep_D9.400 <- rowSums(cpm(geneC_new[,meta$SFVrn=="D09.400"])>=1) >= 5
sum(keep_D9.400)
#Day 80 Trt 2800
keep_D80.2800 <- rowSums(cpm(geneC_new[,meta$SFVrn=="D80.2800"])>=1) >= 5
sum(keep_D80.2800)
#Day 80 Trt 400
keep_D80.400 <- rowSums(cpm(geneC_new[,meta$SFVrn=="D80.400"])>=1) >= 5
sum(keep_D80.400)

keep_gene_a2 <- rowSums(cbind(keep_D9.2800,keep_D9.400,
                                 keep_D80.2800,keep_D80.400)) >= 1
# Filter low coverage genes
geneC_a2 <- voom(geneC_new[keep_gene_a2, ])
#geneC_Elist <- cpm(geneC_a2)
#geneC_a2 <- geneC_Elist$E

ge_dnamV2 <- ge_dnamV2[keep_gene_a2,]
```

**Summarize Gene Expression**
```{r}
cv <- function(x)(sd(x)/mean(x))
gene_Mean <-  rowMeans(geneC_a2)
gene_cv <-  apply(geneC_a2,1,cv)
gene_sum_all <- data.frame(gene_Mean,gene_cv)

gene_Mean_9C <- rowMeans(geneC_a2[,meta$SFVrn=="D09.400"])
gene_cv_9C <-  apply(geneC_a2[,meta$SFVrn=="D09.400"],1,cv)
gene_sum_9C <- data.frame(gene_Mean_9C,gene_cv_9C)

gene_Mean_9E <- rowMeans(geneC_a2[,meta$SFVrn=="D09.2800"])
gene_cv_9E <-  apply(geneC_a2[,meta$SFVrn=="D09.2800"],1,cv)

gene_Mean_80C <- rowMeans(geneC_a2[,meta$SFVrn=="D80.400"])
gene_cv_80C <-  apply(geneC_a2[,meta$SFVrn=="D80.400"],1,cv)

gene_Mean_80E <- rowMeans(geneC_a2[,meta$SFVrn=="D80.2800"])
gene_cv_80E <-  apply(geneC_a2[,meta$SFVrn=="D80.2800"],1,cv)

trt_cv <- apply(cbind(gene_cv_9C,gene_Mean_9E,gene_Mean_80C,gene_Mean_80E),1,cv)
diff <- gene_Mean_9C-gene_Mean_9E

# Make gene expression data frame
#ge_sum <- data.frame(diff,gene_Mean_9C,gene_cv_9C ,gene_Mean_9E)
#colnames(ge_sum) <- c("Expression_diff","Expression_9C","ExpressionCV_9C","Expression_9E")
ge_sum <- data.frame(gene_Mean,trt_cv)
colnames(ge_sum) <- c("Expression","Expression_CV")
```

**Create dataframes for PCA**
```{r}
col <- c("gene_length","cpgMean_filtered","cpgCV_filtered","cpgCount_total","Exon_Count") 
cpg_pca <- subset(ge_dnamV2,select=col)
colnames(cpg_pca) <- c("gene_length","Methylation","Methylation CV","CpGs","Exons")
cpg_pca$Methylation <- as.numeric(as.character(cpg_pca$Methylation))
cpg_pca$`Methylation CV`<- as.numeric(as.character(cpg_pca$`Methylation CV` ))
cpg_pca_rv <- cpg_pca[!is.na(cpg_pca$Methylation | cpg_pca$`Methylation CV`),]


ge_sum <- ge_sum[!is.na(cpg_pca$Methylation | cpg_pca$`Methylation CV`),]
final_pca <- data.frame(ge_sum,cpg_pca_rv)

sum(is.na(final_pca[,1]))
```

**PCA **
```{r}
# Perform pca
pca_obj <- prcomp(final_pca,scale=TRUE)

# Scree plot 
fviz_eig(pca_obj,addlabels = TRUE)

# Plot top two pcs with variable vectors
fviz_pca_var(pca_obj,
             title="Variable Vectors - GE x DNAm",
             col.var = "contrib", # Color by contributions to the PC
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

# Contributions of variables to PC1
fviz_contrib(pca_obj, choice = "var", axes = 1, top = 10)
# Contributions of variables to PC2
fviz_contrib(pca_obj, choice = "var", axes = 2, top = 10)
# Contributions of variables to PC3
fviz_contrib(pca_obj, choice = "var", axes = 3, top = 10)

fviz_pca_ind(pca_obj,
             label = "none" # hide individual labels
)
ind <- get_pca_ind(pca_obj)
head(ind$coord[order(ind$coord[,1]),])
which(!is.na(match(rownames(final_pca),"LOC111102280")))
final_pca[rownames(final_pca)=="LOC111102280",]
gene[gene$GENEID=="LOC111102280",]
head(gene)
```

**Plotting Expression CV by Methylation**
```{r}
# From PCA we see interesting relationship between DNA methylation and gene CV
ggplot(final_pca,aes(y=Methylation,x=ExpressionCV_9C))+geom_point() + geom_smooth()

## Looking at mean methylation acros quantiles of gene CV
qq<-cut(final_pca$ExpressionCV_9C,
        breaks=seq(min(final_pca$ExpressionCV_9C),max(final_pca$ExpressionCV_9C), length.out=11),
        include.lowest=T, labels=F)
dnam_quantile <- matrix(ncol=2,nrow=length(unique(qq)))
ci <- function(x){1.96*(sd(x)/sqrt(length(x)))}
for(i in 1:length(unique(qq))){
  means<-mean(final_pca[qq ==unique(qq)[order(unique(qq))][i],6])
  ciS<-ci(final_pca[qq ==unique(qq)[order(unique(qq))][i],6])
  dnam_quantile[i,] <- cbind(means,ciS)
}

# Create data frame with quantile methylation information
plot_df <- data.frame(CV_quantiles=c(1:length(unique(qq))),mean=dnam_quantile[,1],ci=dnam_quantile[,2])

# ggplot
ggplot(plot_df,aes(y=mean,x=CV_quantiles))+ geom_col()
# base plot
bp <- barplot(plot_df$mean,ylim=c(0,1))
arrows(x0=bp,x1=bp,y0=plot_df$mean-plot_df$ci,y1=plot_df$mean+plot_df$ci,angle=90,code=3,length=0.1)
```
