---
title: "DEG_RSEM"
author: "adowneywall"
date: "8/13/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
library(edgeR)
library(limma)
library(fdrtool)
library(DESeq2)
library(dplyr)
library(variancePartition)
```

```{r}
RStudio.Version()

#### RSEM counts ####
RSEM <-  readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/STAR_mapping/RSEM_output/RSEM_gene_Summary.Rdata")
# Separate out RSEM counts and rename rows with LOC ID
rsem_c <- RSEM$Count
rm(RSEM)
RSEM_t <-  readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/STAR_mapping/RSEM_output/RSEM_isoform_Summary.Rdata")
rsem_t_c <- RSEM_t$Count
rm(RSEM_t)
meta <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/meta/metadata_20190811.RData")

#### Transcript File ####
tran <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/references/STAR_gnomon_tximportGeneFile.RData")
# Gene File
gene <- tran[!duplicated(tran$GENEID),]

#### Meta Data ####
meta <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/meta/metadata_20190718.RData")
meta$sampl_nameSimple <- substr(meta$sample_name,start = 4,stop=9)
#Creat new factor levels (one for each level combination)
meta$SFVrn <- as.factor(paste0("D",meta$SFV))
meta$Sample_Index <- as.factor(meta$sample_index)
meta$TankID <- as.factor(meta$tankID)
  
### Target List ###
tl <- read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/references/Target_BiomineralizationGenes .csv")

tl$Location <- as.character(tl$Location)

#### Data Manipulation ####
# Separate out RSEM counts and rename rows with LOC ID
loc_rsem <- gene$GENEID[match(row.names(rsem_c),gene$gene_id)]
row.names(rsem_c) <- loc_rsem
loc_rsem_t <- tran$TXNAME[match(row.names(rsem_t_c),tran$TXNAME)]
row.names(rsem_t_c) <- loc_rsem_t

geneC <- rsem_c[order(rownames(rsem_c)),]
tranC <- rsem_t_c[order(rownames(rsem_t_c)),]
```

Histogram target vs total genes
```{r}
# geneC_sub <- geneC[!is.na(match(rownames(geneC),tl$Location)),]
# 
# total <- cbind(type="Total",log10_Expression=log10(rowSums((cpm(geneC$counts)))))
# target <- cbind(type="Target",log10_Expression=log10(rowSums((cpm(geneC_sub)))))
# all_vals <- data.frame(rbind(total,target))
# all_vals$log10_Expression <- as.numeric(as.character(all_vals$log10_Expression))
# ggplot(all_vals,aes(log10_Expression,stat(density),colour=type)) +
#   geom_freqpoly(bins=10) #+
#   scale_fill_discrete(name = "Gene Expression Target vs Total", 
#                       breaks=c("A", "B"), 
#                       labels = c("1","2")) # plot
# 
# head(geneC_sub$counts)
# sub_09 <- meta$ID[]
# subgeneC_sub$counts
```


# Create DGEList and perform normalization
```{r}
geneC <- DGEList(geneC,genes = gene) # counts - rsem
tranC <- DGEList(tranC,genes = tran) # counts - rsem

keep_gene <- rowSums(cpm(geneC)>1) >= (0.5 * 24)

keep_tran <- rowSums(cpm(tranC)>1) >= (0.5 * 24)

geneC <- geneC[keep_gene, ]
#saveRDS(geneC,"/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/STAR_mapping/RSEM_output/GeneCountMatrix_Filtered_DGEListObj.RData")
tranC <- tranC[keep_tran, ]
dim(geneC)
dim(tranC)

gene_norm <- calcNormFactors(geneC)
tran_norm <- calcNormFactors(tranC)
```

Perform voom and lmFit
```{r}
### Design matrix (treating treatment and time as single factor variable to modelling purposes)
#Make new design matrix
design <- model.matrix(~0+SFVrn,data=meta)
#Rename columns
colnames(design) <- levels(meta$SFVrn)
#Create specific contrasts
contr_mat <- makeContrasts(
  CvE_D9 = D09.2800-D09.400,
  CvE_D80 = D80.2800-D80.400,
  C_D9vD80 = D09.400-D80.400,
  Diff = (D09.2800-D09.400)- (D80.2800-D80.400),
  levels=design
)

### Voom 
gene_voom <- voom(gene_norm,design,plot = TRUE)
plotMDS(gene_voom,col=as.numeric(meta$Treatment),main="Genes")
tran_voom <- voom(tran_norm, design,plot = TRUE)
plotMDS(tran_norm,col=as.numeric(meta$Treatment),main="Transcripts")


# Identify correlation between factors in design contrasts with blocking factor
gene_cor <- duplicateCorrelation(gene_voom, design, block = meta$tankID)
# Identify correlation between factors in design contrasts with blocking factor
tran_cor <- duplicateCorrelation(tran_voom, design, block = meta$tankID)

#Rerun lmFit model w/ blocking factor
#genes
gene_fit_contr <- lmFit(gene_voom, design,
                        block = meta$tankID,
                        correlation = gene_cor$consensus.correlation)
#transcripts
tran_fit_contr <- lmFit(tran_voom, design,
                        block = meta$tankID,
                        correlation = tran_cor$consensus.correlation)

#Refit with contrasts
gene_fit2_contr <- contrasts.fit(gene_fit_contr,contr_mat)
tran_fit2_contr <- contrasts.fit(tran_fit_contr,contr_mat)
#Run in eBayes
gene_fit2_contr <- eBayes(gene_fit2_contr,robust = TRUE)
tran_fit2_contr <- eBayes(tran_fit2_contr)
saveRDS(gene_fit2_contr,"/home/downeyam/Github/2017OAExp_Oysters/results/DiffExpression_gene_Limma_voom_summaryObject.RData")
saveRDS(tran_fit2_contr,"/home/downeyam/Github/2017OAExp_Oysters/results/DiffExpression_transcript_Limma_voom_summaryObject.RData")
```

Checking out top hits and volcano plots
```{r}
(toGene <- topTable(gene_fit2_contr))
(topTran <- topTable(tran_fit2_contr))
head(gene_fit2_contr$genes$predict)
expressionSum <- data.frame(meanExpression=gene_fit2_contr$Amean,gene_fit2_contr$coefficients,predict=gene_fit2_contr$genes$predict)
expressionSum_target <- expressionSum[!is.na(match(rownames(expressionSum),tl$Location)),]
expressionSum_target <- expressionSum_target[order(rownames(expressionSum_target)),]
write.csv(expressionSum_target,"/home/downeyam/Github/2017OAExp_Oysters/results/expressionTargetGenes_results_20190811.csv")
head(gene_fit2_contr$t)
gene_fdr <- fdrtool(gene_fit2_contr$t[,1])
tran_fdr <- fdrtool(tran_fit2_contr$t[,4])
min(gene_fdr$qval)
min(tran_fdr$qval)
# CvE_D9
plot(-log(gene_fit2_contr$p.value[,1])~gene_fit2_contr$coefficients[,1],
     xlab="Log2Fold Change",ylab="-log(p)")
temp <- !is.na(match(rownames(gene_fit2_contr),topTran$GENEID))
sub_top_c <- gene_fit2_contr$coefficients[temp,1]
sub_top_p <- gene_fit2_contr$p.value[temp,1]
points(-log(sub_top_p)~sub_top_c,col="red",pch=16)
# CvE_D80
plot(-log(gene_fit2_contr$p.value[,2])~gene_fit2_contr$coefficients[,2],
     xlab="Log2Fold Change",ylab="-log(p)")
temp <- !is.na(match(rownames(gene_fit2_contr),topTran$GENEID))
sub_top_c <- gene_fit2_contr$coefficients[temp,2]
sub_top_p <- gene_fit2_contr$p.value[temp,2]
points(-log(sub_top_p)~sub_top_c,col="red",pch=16)
# C_D9vD80
plot(-log(gene_fit2_contr$p.value[,3])~gene_fit2_contr$coefficients[,2],
     xlab="Log2Fold Change",ylab="-log(p)")
temp <- !is.na(match(rownames(gene_fit2_contr),topTran$GENEID))
sub_top_c <- gene_fit2_contr$coefficients[temp,3]
sub_top_p <- gene_fit2_contr$p.value[temp,3]
points(-log(sub_top_p)~sub_top_c,col="red",pch=16)
# Difference
plot(-log(gene_fit2_contr$p.value[,4])~gene_fit2_contr$coefficients[,4],
     xlab="Log2Fold Change",ylab="-log(p)")
temp <- !is.na(match(rownames(gene_fit2_contr),topTran$GENEID))
sub_top_c <- gene_fit2_contr$coefficients[temp,4]
sub_top_p <- gene_fit2_contr$p.value[temp,4]
points(-log(sub_top_p)~sub_top_c,col="red",pch=16)
```
