---
title: "01 salmon countmatrix"
output: 
  html_document: 
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tximport)
```

**Data**  
```{r}
dir <- "/home/downeyam/Github/2017OAExp_Oysters/input_files/salmon_RNA/out2/"
ls_files <- list.files(dir)
files_input <- file.path(dir,ls_files,"/quant.sf",fsep = "")

trans <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/transcriptome_table.RData")
tr2gene <- data.frame(TXNAME = trans$fullID,GENEID=trans$location)
tr2gene$TXNAME <- as.character(tr2gene$TXNAME)
tr2gene$GENEID <- as.character(tr2gene$GENEID)

model<-read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Time <- as.character(model$timepoint)
model$Time[model$Time == "3"] <- "09"
model$Time[model$Time == "6"] <- "80"
model$Time <- as.factor(model$Time)
```

**Creating Count Matrix**  

**Gene Count Matrix**
```{r}
names(files_input) <- model$sample_name
# Using transcriptome from Genome
txi_transFromGenome_genes_1 <- tximport(files_input[1:6],
                                      type = "salmon",
                                      tx2gene = tr2gene)
txi_transFromGenome_genes_2 <- tximport(files_input[7:12],
                                      type = "salmon",
                                      tx2gene = tr2gene)
txi_transFromGenome_genes_3 <- tximport(files_input[13:18],
                                      type = "salmon",
                                      tx2gene = tr2gene)
txi_transFromGenome_genes_4 <- tximport(files_input[19:24],
                                      type = "salmon",
                                      tx2gene = tr2gene)

gene_countMatrix <- data.frame(cbind(txi_transFromGenome_genes_1$counts,
                 txi_transFromGenome_genes_2$counts,
                 txi_transFromGenome_genes_3$counts,
                 txi_transFromGenome_genes_4$counts))
#saveRDS(gene_countMatrix,"/home/downeyam/Github/2017OAExp_Oysters/input_files/salmon_RNA/out2/gene_countMatrix.RData")
```

**Transcript Count Matrix**
```{r}
txi_transFromGenome_trans_1 <- tximport(files_input[1:3],
                                      type = "salmon",
                                      txOut = TRUE)
t1 <- txi_transFromGenome_trans_1$counts
rm(txi_transFromGenome_trans_1)
txi_transFromGenome_trans_2 <- tximport(files_input[4:6],
                                      type = "salmon",
                                      txOut = TRUE)
t2 <- txi_transFromGenome_trans_2$counts
rm(txi_transFromGenome_trans_2)
txi_transFromGenome_trans_3 <- tximport(files_input[7:9],
                                      type = "salmon",
                                      txOut = TRUE)
t3 <- txi_transFromGenome_trans_3$counts
txi_transFromGenome_trans_4 <- tximport(files_input[10:12],
                                      type = "salmon",
                                      txOut = TRUE)
t4 <- txi_transFromGenome_trans_4$counts

transcript_countMatrix <- data.frame(cbind(txi_transFromGenome_genes_1$counts,
                 txi_transFromGenome_genes_2$counts,
                 txi_transFromGenome_genes_3$counts,
                 txi_transFromGenome_genes_4$counts))
saveRDS(transcript_countMatrix,"/home/downeyam/Github/2017OAExp_Oysters/input_files/salmon_RNA/out2/trans_countMatrix.RData")
```

