---
title: "3 salmon DESeq2"
author: "adowneywall"
date: "5/12/2019"
output: 
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(ashr)
library(kableExtra)
```

**Data**
```{r}
model<-read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Time <- as.character(model$timepoint)
model$Time[model$Time == "3"] <- "09"
model$Time[model$Time == "6"] <- "80"
model$Time <- as.factor(model$Time)
model$Pop <- as.factor(model$population)
model$Lane <- as.factor(model$lane)
model$SFV <-  interaction(model$Time,model$Treatment) # Creates single factor variable for combination of time and treatment

gc <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/salmon_RNA/run20180512_gene_countMatrix_.RData")
gc <- gc[-c(1:205),] #removing some no LOC genes (most trna)
gc_round <- round(gc,digits=0)

trans <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/transcriptome_table.RData")
```

**Full Model** - Lane, Pop, Treatment, Time, Treatment:Time as factors  
```{r}
# Full Model
filtered_full <- DESeqDataSetFromMatrix(countData = gc_round,
                              colData = model,
                              design = ~ Lane + Pop + Treatment + Time + Treatment:Time)
dds_filtered_full <- DESeq(filtered_full,minmu = 120)
# Interaction
res_filtered_full <- results(dds_filtered_full)
summary(res_filtered_full)
pVal_interaction <- res_filtered_full$padj
# Time
res_filtered_full_Time <- results(dds_filtered_full,contrast=c("Time","09","80"))
summary(res_filtered_full_Time)
pVal_time <- res_filtered_full_Time$padj
# Treatment
res_filtered_full_Treatment <- results(dds_filtered_full,contrast=c("Treatment","400","2800"))
summary(res_filtered_full_Treatment)
pVal_treatment <- res_filtered_full_Treatment$padj

pvals_df <- data.frame(location=row.names(gc),pVal_interaction,pVal_time,pVal_treatment)


## Looking at most significant genes
#Interaction (lowest pval)
ordered_val_full <- order(res_filtered_full$padj)
plotCounts(dds_filtered_full, gene=ordered_val_full[1], intgroup="SFV",normalized = FALSE)
# Treatment (lowest pval)
ordered_val_treatment <- order(res_filtered_full_Treatment$padj)
plotCounts(dds_filtered_full, gene=ordered_val_treatment[1], intgroup="SFV",normalized = FALSE)
#Time (lowest pval)
ordered_val_time <- order(res_filtered_full_Time$padj)
plotCounts(dds_filtered_full, gene=ordered_val_time[1], intgroup="SFV",normalized = FALSE)

gene_GE <- merge(trans,pvals_df,by="location")
```

**Top Twenty Genes Associated with Treatment** 
```{r}
gene_GE_trt <- gene_GE[order(gene_GE$pVal_treatment),]
kable(gene_GE_trt[1:20,]) %>%
  kable_styling()
```

**Top Twenty Genes Associated with Time** 
```{r}
gene_GE_t <- gene_GE[order(gene_GE$pVal_time),]
kable(gene_GE_t[1:20,]) %>%
  kable_styling()
```

**Top Twenty Genes Associated with Time:Treatment** 
```{r}
gene_GE_ttrt <- gene_GE[order(gene_GE$pVal_interaction),]
kable(gene_GE_ttrt[1:20,]) %>%
  kable_styling()
```

### A few targeted gene queries  

**Carbonic Anhydrase**  
```{r}
### Carbonic Anhydrase  
# All potential carbonic anhydrase gene locations based on CV genome on NCBI 
ca_LOC <- c("LOC111134699",
  "LOC111133036",
  "LOC111130107",
  "LOC111134700",
  "LOC111123083",
  "LOC111113402",
  "LOC111137424",
  "LOC111117514",
  "LOC111114059",
  "LOC111112162",
  "LOC111133640",
  "LOC111134104",
  "LOC111133247",
  "LOC111130106",
  "LOC111127254",
  "LOC111127093",
  "LOC111122701",
  "LOC111120606",
  "LOC111120607",
  "LOC11125708")

# Subset for locations matching known carbonic anhydrase genes
match(ca_LOC,row.names(gc),nomatch = FALSE)
out_ca <- match(ca_LOC,row.names(gc))
out_ca <- out_ca[!is.na(out_ca)]
ca_data <- data.frame(t(as.matrix(gc[out_ca,])))
ca_data$Treatment <- model$Treatment
ca_data$Day <- model$Day
pvals_df[out_ca,]

ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,1])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,2])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,3])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,4])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,5])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,6])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,7])) + geom_boxplot()
```

**Calbindin**
```{r}
gc[match("LOC111136252",row.names(gc)),]
pvals_df[match("LOC111136252",row.names(gc)),]
```

**EGF**
```{r}
gc[match("LOC111134661",row.names(gc)),]
pvals_df[match("LOC111134661",row.names(gc)),]
```

