---
title: "3 salmon DESeq2"
author: "adowneywall"
date: "5/12/2019"
output: 
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(ashr)
library(kableExtra)
```

### **Data**
```{r}
model<-read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Time <- as.character(model$timepoint)
model$Time[model$Time == "3"] <- "09"
model$Time[model$Time == "6"] <- "80"
model$Time <- as.factor(model$Time)
model$Pop <- as.factor(model$population)
model$Lane <- as.factor(model$lane)
model$SFV <-  interaction(model$Time,model$Treatment) # Creates single factor variable for combination of time and treatment

# Transcript abundances - scaled by isoform length
ga <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/salmon_pipeline/run20190610/geneMatrixAbundance_default.RData")
# Raw counts (tximport does not recommend using these for DE)
gc <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/salmon_pipeline/run20190610/geneMatrixCounts_default.Rdata")
# Round counts before create DESeq matrix
ga_round <- round(ga,digits=0)
# Additional info about each gene
trans <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/transcriptome_table.RData")
```

**Full Model** - Lane, Pop, Treatment, Time, Treatment:Time as factors  
```{r}
# Full Model
full <- DESeqDataSetFromMatrix(countData = ga_round,
                              colData = model,
                              design = ~ Lane + Pop + Treatment + Time + Treatment:Time)

dds_full_Wald <- DESeq(full,minmu = 80,test = "Wald")
# Interaction
res_full <- results(dds_full_Wald )
summary(res_full)
pVal_interaction <- res_full$padj
# Time
res_full_Time <- results(dds_full_Wald,contrast=c("Time","09","80"))
summary(res_full_Time)
pVal_time <- res_full_Time$padj
# Treatment
res_full_Treatment <- results(dds_full_Wald,contrast=c("Treatment","400","2800"))
summary(res_full_Treatment)
pVal_treatment <- res_full_Treatment$padj

pvals_df <- data.frame(location=row.names(gc),pVal_interaction,pVal_time,pVal_treatment)


## Looking at most significant genes
#Interaction (lowest pval)
ordered_val_full <- order(res_filtered_full$padj)
plotCounts(dds_filtered_full, gene=ordered_val_full[1], intgroup="SFV",normalized = FALSE)
# Treatment (lowest pval)
ordered_val_treatment <- order(res_filtered_full_Treatment$padj)
plotCounts(dds_filtered_full, gene=ordered_val_treatment[1], intgroup="SFV",normalized = FALSE)
#Time (lowest pval)
ordered_val_time <- order(res_filtered_full_Time$padj)
plotCounts(dds_filtered_full, gene=ordered_val_time[1], intgroup="SFV",normalized = FALSE)

gene_GE <- merge(trans,pvals_df,by="location")
```

**Top Twenty Genes Associated with Treatment** 
```{r}
gene_GE_trt <- gene_GE[order(gene_GE$pVal_treatment),]
kable(gene_GE_trt[1:20,]) %>%
  kable_styling()
```

**Top Twenty Genes Associated with Time** 
```{r}
gene_GE_t <- gene_GE[order(gene_GE$pVal_time),]
kable(gene_GE_t[1:20,]) %>%
  kable_styling()
```

**Top Twenty Genes Associated with Time:Treatment** 
```{r}
gene_GE_ttrt <- gene_GE[order(gene_GE$pVal_interaction),]
kable(gene_GE_ttrt[1:20,]) %>%
  kable_styling()
```

### A few targeted gene queries  

**Carbonic Anhydrase**  
```{r}
### Carbonic Anhydrase  
# All potential carbonic anhydrase gene locations based on CV genome on NCBI 
ca_LOC <- c("LOC111134699",
  "LOC111133036",
  "LOC111130107",
  "LOC111134700",
  "LOC111123083",
  "LOC111113402",
  "LOC111137424",
  "LOC111117514",
  "LOC111114059",
  "LOC111112162",
  "LOC111133640",
  "LOC111134104",
  "LOC111133247",
  "LOC111130106",
  "LOC111127254",
  "LOC111127093",
  "LOC111122701",
  "LOC111120606",
  "LOC111120607",
  "LOC11125708")

# Subset for locations matching known carbonic anhydrase genes
match(ca_LOC,row.names(gc),nomatch = FALSE)
out_ca <- match(ca_LOC,row.names(gc))
out_ca <- out_ca[!is.na(out_ca)]
ca_data <- data.frame(t(as.matrix(gc[out_ca,])))
ca_data$Treatment <- model$Treatment
ca_data$Day <- model$Day
pvals_df[out_ca,]

ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,1])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,2])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,3])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,4])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,5])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,6])) + geom_boxplot()
ggplot(ca_data,aes(x=interaction(Treatment,Day),y=ca_data[,7])) + geom_boxplot()
```

**Calbindin**
```{r}
gc[match("LOC111136252",row.names(gc)),]
pvals_df[match("LOC111136252",row.names(gc)),]
```

**EGF**
```{r}
gc[match("LOC111134661",row.names(gc)),]
pvals_df[match("LOC111134661",row.names(gc)),]
```

## DE specifically for individuals from TP9
```{r}
ga_tp9 <-  ga_round[,model$Time == "09"]
dim(ga_tp9)
head(ga_tp9)
model_tp9 <- model[model$Time == "09",]
dim(model_tp9)
tp9 <- DESeqDataSetFromMatrix(countData = ga_tp9,
                              colData = model_tp9,
                              design = ~ Treatment)

# Wald test using negative binomial
dds_tp9_Wald <- DESeq(tp9,minmu = 80,test = "Wald")
# Likelihood ratio test
dds_tp9_LTR <- DESeq(tp9,minmu=80,test="LRT",reduced = ~1)

# Treatment at timepoint 9 with Wald
res_tp9 <- results(dds_tp9_Wald )
summary(res_tp9)
pVal_tp9 <- res_tp9$padj

# Treatment at timepoint 9 with LRT
res_tp9_lrt <- results(dds_tp9_LTR)
summary(res_tp9_lrt)
pVal_tp9_lrt <- res_tp9_lrt$padj

pvals_tp9_df <- data.frame(location=row.names(gc),wald_p=pVal_tp9,lrt_p=pVal_tp9_lrt)

gene_GE_tp9 <- merge(trans,pvals_tp9_df,by="location")
gene_GE_tp9 <- gene_GE_tp9[order(gene_GE_tp9$wald_p),]
kable(gene_GE_tp9[1:50,]) %>%
  kable_styling()
```

## DE specifically for individuals from TP80
```{r}
ga_tp80 <-  ga_round[,model$Time == "80"]
dim(ga_tp80)
head(ga_tp80)
model_tp80 <- model[model$Time == "80",]
dim(model_tp80)
tp80 <- DESeqDataSetFromMatrix(countData = ga_tp80,
                              colData = model_tp80,
                              design = ~ Treatment)

# Wald test using negative binomial
dds_tp80_Wald <- DESeq(tp80,minmu = 80,test = "Wald")
# Likelihood ratio test
dds_tp80_LTR <- DESeq(tp80,minmu=80,test="LRT",reduced = ~1)

# Treatment at timepoint 9 with Wald
res_tp80 <- results(dds_tp80_Wald )
summary(res_tp80)
pVal_interaction <- res_tp80$padj

# Treatment at timepoint 9 with LRT
res_tp80_lrt <- results(dds_tp80_LTR)
summary(res_tp80_lrt)
pVal_interaction_lrt <- res_tp80_lrt$padj

pvals_tp80_df <- data.frame(location=row.names(gc),wald_p=pVal_tp9,lrt_p=pVal_tp9_lrt)

gene_GE_tp80 <- merge(trans,pvals_tp80_df,by="location")
gene_GE_tp80 <- gene_GE_tp80[order(gene_GE_tp80$wald_p),]
kable(gene_GE_tp80[1:50,]) %>%
  kable_styling()
```
