---
title: "04_salmon_sleuth"
author: "adowneywall"
date: "5/13/2019"
output: 
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#source("https://bioconductor.org/biocLite.R")
#biocLite("biomaRt")
library(biomaRt)
library(wasabi)
library(sleuth)
library(cowplot)
```

Using Wasabi to convert Salmon files to ```.h5``` file format for sleuth
```{r eval=FALSE}
dir <- "/home/downeyam/Github/2017OAExp_Oysters/input_files/salmon_RNA"
setwd(dir)
sfdirs <- file.path(dir,"run20180512", list.files("run20180512"))
prepare_fish_for_sleuth(sfdirs)
```


```{r}
model<-read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/metadata_cvirginica_rna_meta.txt", header=TRUE)
model$Treatment <- as.factor(model$treatment)
model$Time <- as.character(model$timepoint)
model$Time[model$Time == "3"] <- "09"
model$Time[model$Time == "6"] <- "80"
model$Time <- as.factor(model$Time)
model$Pop <- as.factor(model$population)
model$Lane <- as.factor(model$lane)
model$SFV <-  interaction(model$Time,model$Treatment) # Creates single factor variable for combination of time and treatment
samp_names <- substr(model$sample_name,4,10)
model <- dplyr::mutate(model,
  path = file.path(dir,'/run20180512/',samp_names,'/abundance.h5',fsep = ""))

metadata <- dplyr::select(model,c("sample_name","Time","Treatment","SFV","Pop","Lane","path"))
colnames(metadata)[1] <- "sample"

trans <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/transcriptome_fromGenome_table.RData")

ttg <- dplyr::select(trans,c("fullID","location","product"))
colnames(ttg) <- c("target_id","ens_gene","product")
```

```{r eval=FALSE}
so <- sleuth_prep(metadata, target_mapping = ttg,aggregation_column = 'ens_gene', extra_bootstrap_summary = TRUE)
#saveRDS(so,"/home/downeyam/Github/2017OAExp_Oysters/results/20190513_sluethobject_fullmodelparameters.RData")
```

```{r}
so <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/results/20190513_sluethobject_fullmodelparameters.RData")
#reduced models
so <- sleuth_fit(so, ~Lane + Pop, 'reduced')
so <- sleuth_fit(so, ~Lane + Pop + Time, 'reduced_time')
so <- sleuth_fit(so, ~Lane + Pop + Treatment, 'reduced_treatment')
#full model v1
so <- sleuth_fit(so, ~Lane + Pop + Time + Treatment, 'full')
#full model v2
so <- sleuth_fit(so, ~Lane + Pop + SFV, 'full_single_factor')

#Likelihood ratio test
so <- sleuth_lrt(so, 'reduced', 'full')
so <- sleuth_lrt(so, 'reduced_time', 'full')
so <- sleuth_lrt(so, 'reduced_treatment', 'full')

so <- sleuth_lrt(so, 'reduced', 'reduced_treatment')
so <- sleuth_lrt(so, 'reduced', 'reduced_time')

so <- sleuth_lrt(so, 'reduced', 'full_single_factor')
```

##Extracting significant values

### Gene Level

#### Treatment
```{r}
# Treatment comparison with Time in as additional factor
sleuth_table_treatment_gene_1 <- sleuth_results(so, 'reduced_time:full', 'lrt', show_all = FALSE)
#sleuth_table_treatment_gene_1 <- dplyr::filter(sleuth_table_treatment_gene_1, qval <= 0.05)
#head(sleuth_table_treatment_gene_1, 20)
# Treatment comparison without Time
sleuth_table_treatment_gene_2 <- sleuth_results(so, 'reduced:reduced_treatment', 'lrt', show_all = FALSE)
#sleuth_table_treatment_gene_2 <- dplyr::filter(sleuth_table_treatment_gene_2, qval <= 0.05)
#head(sleuth_table_treatment_gene_2, 20)
# Still produce same significant genes
```

#### Time
```{r}
# Time comparison with Time in as additional factor
sleuth_table_time_gene_1 <- sleuth_results(so, 'reduced_treatment:full', 'lrt', show_all = FALSE)
time_genes <- dplyr::select(sleuth_table_time_gene_1,c("target_id",time_qval="qval"))
#sleuth_table_time_gene_1 <- dplyr::filter(sleuth_table_time_gene_1, qval <= 0.05)
#head(sleuth_table_time_gene_1, 20)
# Time comparison with Time in as additional factor
sleuth_table_time_gene_2 <- sleuth_results(so, 'reduced:reduced_time', 'lrt', show_all = FALSE)
#sleuth_table_time_gene_2 <- dplyr::filter(sleuth_table_time_gene_2, qval <= 0.05)
#head(sleuth_table_time_gene_2, 20)
```

#### Time + Treatment
```{r}
# Time comparison with Time in as additional factor
sleuth_table_timeTrt_gene_1 <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)
timeTrt_genes <- dplyr::select(sleuth_table_timeTrt_gene_1,c("target_id",timeTrt_qval="qval"))
#sleuth_table_timeTrt_gene_1 <- dplyr::filter(sleuth_table_timeTrt_gene_1, qval <= 0.05)
#head(sleuth_table_timeTrt_gene_1, 20)
```

#### SFV
```{r}
# Time comparison with Time in as additional factor
sleuth_table_SFV_gene <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE)
#sleuth_table_SFV_gene <- dplyr::filter(sleuth_table_SFV_gene, qval <= 0.05)
#head(sleuth_table_SFV_gene, 20)
```

### Combining outputs
```{r}
step1 <- merge(sleuth_table_treatment_gene_1,time_genes,by="target_id")
combine <- merge(step1,timeTrt_genes,by="target_id")
combine_unique <- unique(combine)
combine_sig <- subset(combine_unique, qval<=0.05 | time_qval<=0.05 | timeTrt_qval<=0.05)
final_gene_list <- combine_sig[which(!duplicated(combine_sig$target_id)),]

trt <- round(final_gene_list$qval + 0.4499999)
time <- round(final_gene_list$time_qval + 0.4499999)*100
timeTrt <- round(final_gene_list$timeTrt_qval + 0.4499999)*1000
all <- trt+time+timeTrt

all_significant <- length(all[all == 0])
time_significant <- length(all[all == 1001])
trt_significant <- length(all[all == 1100])
Timetrt_significant <- length(all[all == 101])
TimeInter_significant <- length(all[all == 1])
TreatmentInter_significant <- length(all[all == 100])
saveRDS(TreatmentInter_significant,"/home/downeyam/Github/2017OAExp_Oysters/results/20190513")
```

### Transcript Level  

#### Treatment
```{r}
sleuth_table_Trt_tx <- sleuth_results(so, 'reduced_time:full', 'lrt', show_all = FALSE, pval_aggregate = FALSE)
sleuth_table_Trt_tx <- dplyr::filter(sleuth_table_Trt_tx, qval <= 0.05)
head(sleuth_table_Time_tx, 20)
```

#### Time 
```{r}
sleuth_table_Time_tx <- sleuth_results(so, 'reduced_treatment:full', 'lrt', show_all = FALSE, pval_aggregate = FALSE)
sleuth_table_Time_tx <- dplyr::filter(sleuth_table_Time_tx, qval <= 0.05)
head(sleuth_table_Time_tx, 20)
```

#### Time + Treatment 
```{r}
sleuth_table_timeTrt_tx <- sleuth_results(so, 'reduced:full', 'lrt', show_all = FALSE, pval_aggregate = FALSE)
sleuth_table_timeTrt_tx <- dplyr::filter(sleuth_table_timeTrt_tx, qval <= 0.05)
head(sleuth_table_timeTrt_tx, 20)
```

```{r}
sleuth_live(so)
sleuth_live(so1)
```
