---
title: "Differential Methylation with bayesian mixed binomial model implemented using BRMS"
date: "10/22/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE,message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
library(dplyr)
library(VennDiagram)
wd <- "/home/downeyam/Github/2017OAExp_Oysters"

library(RColorBrewer)
library("wesanderson")
pal <- brewer.pal(n = 12, name = 'Dark2')
```

**Data**
```{r}
# Sample Information
meta <- readRDS(paste0(wd,"/input_files/meta/metadata_20190811.RData"))
meta <- meta[meta$ID != "17099",]
gene_ID <- readRDS(paste0(wd,"/input_files/RNA/references/Exon_GeneLoc.RData"))
cds <- readRDS(paste0(wd,"/input_files/RNA/references/CDS_wGoTerms_GeneLOC.RData"))

# Gene Expression 
gc <- readRDS(paste0(wd,"/results/Transcriptomic/DGEListObj_withIndWeights_filterApproach2_plannedContrastMatrix.RData"))

# Methylation Data
d_meth <- readRDS(paste0(wd,"/results/DNAm/Final_DNAm_gene_BRMS_plannedComparisons.RData"))
d_meta <- readRDS(paste0(wd,"/input_files/DNAm/CG_unstranded_summaryTable_geneOnly_5.RData"))
d_beta <- readRDS(paste0(wd,"/input_files/DNAm/CG_unstranded_beta_geneOnly_5.RData"))

LOC <- sub(".*Name=(.*?);.*","\\1",d_meta$attribute,perl=TRUE) # Extract LOC ID for each gene we have 
## Gene level summary information
cpgMean_all <-  read.table(paste0(wd,"/results/DNAm/meanBetaPerFeature.txt"),header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")
```

**Differential Methylation Summary Table**

Summarized CpGs that were significantly different after correction for all comparisons. Hypomethylation refers to CpGs that had significant lower methylation in our OA treatment compared to the ambient conditions or on day 80 compared to day 9. The opposity is the case (methylation is higher) if hypermethylated.  

```{r}

cols <- colnames(d_meth$Significant)
sum_mat <- matrix(0,nrow=1,ncol=length(cols))
for(i in 1:length(cols)){
  sum_mat[1,i] <- sum(d_meth$Significant[,i]=="*")
}
final_mat <- matrix(0,nrow=2,ncol=9)
colnames(final_mat) <- c("Trt","Time","Interaction",
                         "Trt_D9",
                         "Ambient_Time",
                         "Ambient_D9_Exposed_D80",
                         "Exposed_D9_Ambient_D80",
                         "Exposed_Time",
                         "Trt_D80"
                         )
row.names(final_mat) <- c("Hypomethylated","Hypermethylated")
final_mat[1,] <- sum_mat[,c(1,3,5,6,8,10,12,14,16)]
final_mat[2,] <- sum_mat[,c(2,4,5,7,9,11,13,15,17)]
final_mat[2,3] <- NA

kableExtra::kable(final_mat) %>% 
  kableExtra::kable_styling()
```

Basic Summary - Long Format
```{r}
name <- colnames(d_meth$Significant)[1]
pos <- which(d_meth$Significant[,1]=="*")
estimate <- d_meth$Estimate[which(d_meth$Significant[,1]=="*"),1]
sig_list <- data.frame(pos=as.numeric(pos),estimate=as.numeric(estimate),name=name)
for(i in 2:ncol(d_meth$Significant)){
  name <- colnames(d_meth$Significant)[i]
  pos <- which(d_meth$Significant[,i]=="*")
  estimate <- d_meth$Estimate[which(d_meth$Significant[,i]=="*"),i]
  temp_bind <- data.frame(pos=as.numeric(pos),estimate=as.numeric(estimate),name=name)
  sig_list <- rbind(sig_list,temp_bind)
}
total_significant <- length(unique(sig_list$pos))

# Order by estimate
sig_list <- sig_list[order(sig_list$estimate),]
```

There were a total of `r total_significant` CpGs significantly differentially methylated in at least one comparison.  

**Table of significant CpGs ranked by the mode estimate (a measure of the magnitude difference between groups)**  
```{r}
kable(head(sig_list)) %>%
  kable_styling()
```

**Venn Diagram -  Initial Explanatory Variables (Trt, Time, Interaction)**
```{r}
dnam_sigTable <- d_meth$Significant
head(dnam_sigTable)
dnam_sigTable_PrimaryVariables <- dnam_sigTable[,1:5]
sum_sig <- function(s){
  return(which(s=="*"))
}
sig_positions <- function(x){
  out <- lapply(x,FUN=sum_sig)
  temp <- do.call(c,out)
  return(unique(temp))
}
is.integer0 <- function(x)
{
  is.integer(x) && length(x) == 0L
}
val_test <- function(x){
  if(is.integer0(x)){return(0)}else{return(x)}
}
sig_pos_trt <- sig_positions(list(dnam_sigTable_PrimaryVariables[,1],
              dnam_sigTable_PrimaryVariables[,2]))
sig_pos_time <- sig_positions(list(dnam_sigTable_PrimaryVariables[,3],
              dnam_sigTable_PrimaryVariables[,4]))
sig_pos_inter <- sig_positions(list(dnam_sigTable_PrimaryVariables[,5]))

Complete_overlap <- Reduce(intersect, list(sig_pos_trt,sig_pos_time,sig_pos_inter))
trt_time_overlap <- Reduce(intersect, list(sig_pos_trt,sig_pos_time))
trt_inter_overlap <- Reduce(intersect, list(sig_pos_trt,sig_pos_inter))
time_inter_overlap <- Reduce(intersect, list(sig_pos_time,sig_pos_inter))

grid.newpage()
draw.triple.venn(area1 = 22, area2 = 20, area3 = 13, n12 = 11, n23 = 4, n13 = 5, 
    n123 = 1, category = c("Dog People", "Cat People", "Lizard People"), lty = "blank", 
    fill = c("skyblue", "pink1", "mediumorchid"))
pal <- wes_palette()
grid.newpage()
draw.triple.venn(area1 = length(sig_pos_trt),
                 area2 = length(sig_pos_time),
                 area3 = length(sig_pos_inter),
                 n12 = length(trt_time_overlap),
                 n23 = length(time_inter_overlap),
                 n13 = length(trt_inter_overlap),
                 n123 = length(Complete_overlap),
                 category = c("Treatment", "Time", "Interaction"),
                 fill = c("white","white","white"),
                 lty = c(1,2,3),
                 alpha = 0.6,
                 cex = 2,
                 cat.pos = c(30,330,180),
                 cat.dist = c(0.08,0.04,0.06),
                 cat.cex = c(3,3,3),
                 rotation = 1,
                 rotation.degree = 180)
```

**Venn Diagram -  Post Hoc Comparisons**

```{r}
dnam_sigTable_pH <- dnam_sigTable[dnam_sigTable[,5] == "*",6:ncol(dnam_sigTable)]
#Subset for each post hoc comparison
sig_pos_C9E9 <- sig_positions(list(dnam_sigTable_pH[,1],
              dnam_sigTable_pH[,2]))
sig_pos_C9C80 <- sig_positions(list(dnam_sigTable_pH[,3],
              dnam_sigTable_pH[,4]))
sig_pos_C9E80 <- sig_positions(list(dnam_sigTable_pH[,5],
              dnam_sigTable_pH[,6]))
sig_pos_E9C80 <- sig_positions(list(dnam_sigTable_pH[,7],
              dnam_sigTable_pH[,8]))
sig_pos_E9E80 <- sig_positions(list(dnam_sigTable_pH[,9],
              dnam_sigTable_pH[,10]))
sig_pos_C80E80 <- sig_positions(list(dnam_sigTable_pH[,11],
              dnam_sigTable_pH[,12]))

# Overlaps between different pairwise comparisons
# All groups
Complete_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_C80E80,sig_pos_C9C80,sig_pos_E9E80))
# three groups
C9E9_C80E80_C9C80_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_C80E80,sig_pos_C9C80)) # 1,2,3
C9E9_C80E80_E9E80_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_C80E80,sig_pos_E9E80)) # 1,2,4
C9E9_C9C80_E9E80_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_C9C80,sig_pos_E9E80)) # 1,3,4
C80E80_C9C80_E9E80_overlap <- Reduce(intersect, list(sig_pos_C80E80,sig_pos_C9C80,sig_pos_E9E80)) # 2,3,4

# Two groups
C9E9_C80E80_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_C80E80)) # 1v2
C9E9_C9C80_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_C9C80)) #1v3
C9E9_E9E80_overlap <- Reduce(intersect, list(sig_pos_C9E9,sig_pos_E9E80)) #1v4
C9C80_C80E80_overlap <- Reduce(intersect, list(sig_pos_C9C80,sig_pos_C80E80)) #2v3
C80E80_E9E80_overlap <- Reduce(intersect, list(sig_pos_E9E80,sig_pos_C80E80)) #2v4
C9C80_E9E80_overlap <- Reduce(intersect, list(sig_pos_C9C80,sig_pos_E9E80)) #3v4

grid.newpage()
draw.quad.venn(area1 = length(sig_pos_C9E9),
               area2 = length(sig_pos_C80E80),
               area3 = length(sig_pos_C9C80),
               area4 = length(sig_pos_E9E80),
               n12 = length(C9E9_C80E80_overlap),
               n13 = length(C9E9_C9C80_overlap),
               n14 = length(C9E9_E9E80_overlap),
               n23 = length(C9C80_C80E80_overlap),
               n24 = length(C80E80_E9E80_overlap),
               n34 = length(C9C80_E9E80_overlap),
               n123 = length(C9E9_C80E80_C9C80_overlap),
               n124 = length(C9E9_C80E80_E9E80_overlap),
               n134 = length(C9E9_C9C80_E9E80_overlap),
               n234 = length(C80E80_C9C80_E9E80_overlap),
               n1234 = length(Complete_overlap),
               category = c("C9_E9", "C80_E80", 
                            "C9_C80", "E9_E80"), 
               lty = 1, 
               fill = c("white"),
               cex = 2,
               cat.cex = 2.2)

```

Significant Loci by Treatment
```{r fig.width=12}
# Short investigation of caherin, which had a significantly differentially methylated CpG by treatment
# Row ID : 243889 - Significant by treatment - Cadherin - LOC111106820
# Row ID : 3240 - Significant by treatment (day 9) - transmembrane - LOC111103288
#LOC_pos <- sig_pos_C9E9[18]
LOC_pos <- sig_pos_C80E80
#LOC <- 3240
# Gene / Exon Information
gene_ID[match(LOC[LOC_pos],gene_ID$gene_id),]
# CDS and Function annotation
cds[match(LOC[LOC_pos],cds$gene_id),]
# Confirm the ids match up with the gene expression data
gc$genes[match(LOC[LOC_pos],gc$genes$GENEID),]
# Subset gene expression data to the same number of individual and the DNA methylation data

LOC_pos2 <-  sig_pos_C80E80[51] 
gc$genes[match(LOC[LOC_pos2],gc$genes$GENEID),]

temp <- gc$E[match(LOC[LOC_pos2],gc$genes$GENEID),colnames(gc$E) != "17099"]
beta_temp <- d_beta[,colnames(d_beta) != "17099"]

#Plots
par(mfrow=c(1,3))
boxplot(unlist(beta_temp[LOC_pos2,])~meta$SFV,ylim=c(0,1),
        xlab="Day.Treatment",ylab="Prop. Methylation")
boxplot(temp~meta$SFV,
        xlab="Day.Treatment",ylab="log2 CPM")
plot(unlist(beta_temp[LOC_pos2,])~temp,ylim=c(0,1),
     ylab="Prop.Methylation",xlab="log2 CPM")
```

```{r}
library(matrixStats)
par(mfrow=c(1,1))
hist(d_meth$Evid.Ratio)

d_meth_evidOrder <- d_meth$Evid.Ratio[rev(order(rowMaxs(d_meth$Evid.Ratio[,6:7]))),]
head(d_meth_evidOrder)
sum(rowMaxs(d_meth$Evid.Ratio[,6:7])>5000)
d_beta_order <- d_beta[rev(order(rowMaxs(d_meth$Evid.Ratio[,6:7]))),]
d_meta_order <- d_meta[rev(order(rowMaxs(d_meth$Evid.Ratio[,6:7]))),]
LOC_order <- LOC[rev(order(rowMaxs(d_meth$Evid.Ratio[,6:7])))]
pos <- 6
cds[match(LOC_order[pos],cds$gene_id),]
gene_ID[match(LOC_order[pos],gene_ID$gene_id),]
gc$genes[match(LOC_order[pos],gc$genes$GENEID),]
d_meta_order[pos,]
boxplot(d_beta_order[pos,]~meta$SFV)
```

**Subsetting significant genes and matching with gene annotations**
```{r eval=FALSE}
#CpG info for
cpg_sub <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),stringsAsFactors = FALSE) # Reduce CpG meta data dataframe
# Extract GO slim terms
go_tab_slim <- cds[,c(12,14)]
go_tab <- cds[,c(12,17)]
go_tab_unique <- go_tab[!duplicated(go_tab$gene_id),]
go_tab_unique$InterproScan.GO.ID <- as.character(go_tab_unique$InterproScan.GO.ID)
go_tab_unique$InterproScan.GO.ID[is.na(go_tab_unique$InterproScan.GO.ID)] <- "unknown"
go_tab_unique$InterproScan.GO.ID[go_tab_unique$InterproScan.GO.ID==""] <- "unknown"
# Join cpg annotations with go slim terms
comb <- left_join(cpg_sub,go_tab_unique,"gene_id")
comb_red <- comb[c(1,3)]

# Comparison specific diff methylation tables
C9E9_up <- data.frame(ID=d_meta$ID,evid=d_meth$Evid.Ratio[,6],stringsAsFactors = FALSE)
C9E9_down <- data.frame(ID=d_meta$ID,evid=d_meth$Evid.Ratio[,7],stringsAsFactors = FALSE)
C80E80_up <- data.frame(ID=d_meta$ID,evid=d_meth$Evid.Ratio[,16],stringsAsFactors = FALSE)
C80E80_down <- data.frame(ID=d_meta$ID,evid=d_meth$Evid.Ratio[,17],stringsAsFactors = FALSE)
# Log transform EVID values
C9E9_up_log <- data.frame(ID=d_meta$ID,evid=log(d_meth$Evid.Ratio[,6]),stringsAsFactors = FALSE)
C9E9_down_log <- data.frame(ID=d_meta$ID,evid=log(d_meth$Evid.Ratio[,7]),stringsAsFactors = FALSE)
C80E80_up_log <- data.frame(ID=d_meta$ID,evid=log(d_meth$Evid.Ratio[,16]),stringsAsFactors = FALSE)
C80E80_down_log <- data.frame(ID=d_meta$ID,evid=log(d_meth$Evid.Ratio[,17]),stringsAsFactors = FALSE)

## Select only the largest EVID Ratio for an individual CpG per gene
C9E9_up2 <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=d_meth$Evid.Ratio[,6],stringsAsFactors = FALSE)
C9E9_up2_log <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=log(d_meth$Evid.Ratio[,6]),stringsAsFactors = FALSE)
C9E9_up_max <- aggregate(evid~gene_id,FUN=max,data=C9E9_up2)
C9E9_up_max_log <- aggregate(evid~gene_id,FUN=max,data=C9E9_up2_log)

C9E9_down2 <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=d_meth$Evid.Ratio[,7],stringsAsFactors = FALSE)
C9E9_down2_log <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=log(d_meth$Evid.Ratio[,7]),stringsAsFactors = FALSE)
C9E9_down_max <- aggregate(evid~gene_id,FUN=max,data=C9E9_down2)
C9E9_down_max_log <- aggregate(evid~gene_id,FUN=max,data=C9E9_down2_log)

C80E80_up2 <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=d_meth$Evid.Ratio[,16],stringsAsFactors = FALSE)
C80E80_up2_log <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=log(d_meth$Evid.Ratio[,16]),stringsAsFactors = FALSE)
C80E80_up_max <- aggregate(evid~gene_id,FUN=max,data=C80E80_up2)
C80E80_up_max_log <- aggregate(evid~gene_id,FUN=max,data=C80E80_up2_log)

C80E80_down2 <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=d_meth$Evid.Ratio[,17],stringsAsFactors = FALSE)
C80E80_down2_log <- data.frame(ID=d_meta$ID,gene_id=as.character(LOC),evid=log(d_meth$Evid.Ratio[,17]),stringsAsFactors = FALSE)
C80E80_down_max <- aggregate(evid~gene_id,FUN=max,data=C80E80_down2)
C80E80_down_max_log <- aggregate(evid~gene_id,FUN=max,data=C80E80_down2)

# GO terms annotation table
write.table(x = comb_red,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_Goterms.tab",sep = "\t",row.names = FALSE)
write.table(x = go_tab_unique,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_Goterms.tab",sep = "\t",row.names = FALSE)
# EVID values for all cpgs
write.csv(x = C9E9_up,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C9E9_UP.csv",row.names = FALSE)
write.csv(x = C9E9_down,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C9E9_DOWN.csv",row.names = FALSE)
write.csv(x = C80E80_up,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C80E80_UP.csv",row.names = FALSE)
write.csv(x = C80E80_down,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C80E80_DOWN.csv",row.names = FALSE)
write.csv(x = C9E9_up_log,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C9E9_UP_log.csv",row.names = FALSE)
write.csv(x = C9E9_down_log,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C9E9_DOWN_log.csv",row.names = FALSE)
write.csv(x = C80E80_up_log,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C80E80_UP_log.csv",row.names = FALSE)
write.csv(x = C80E80_down_log,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_allCpG_C80E80_DOWN_log.csv",row.names = FALSE)
# EVID values for gene max subsets
write.csv(x = C9E9_up_max ,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C9E9_UP.csv",row.names = FALSE)
write.csv(x = C9E9_down_max ,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C9E9_DOWN.csv",row.names = FALSE)
write.csv(x = C80E80_up_max,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C80E80_UP.csv",row.names = FALSE)
write.csv(x = C80E80_down_max,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C80E80_DOWN.csv",row.names = FALSE)
write.csv(x = C9E9_up_max_log ,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C9E9_UP_log.csv",row.names = FALSE)
write.csv(x = C9E9_down_max_log ,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C9E9_DOWN_log.csv",row.names = FALSE)
write.csv(x = C80E80_up_max_log,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C80E80_UP_log.csv",row.names = FALSE)
write.csv(x = C80E80_down_max_log,file = "/home/downeyam/Github/2017OAExp_Oysters/results/DNAm/GOMWU_maxCpG_C80E80_DOWN_log.csv",row.names = FALSE)
```

