---
title: "Differential Methylation with bayesian mixed binomial model implemented using BRMS"
date: "10/22/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE,message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(kableExtra)
library(dplyr)
wd <- "/home/downeyam/Github/2017OAExp_Oysters"
```

**Data**
```{r}
# Sample Information
meta <- readRDS(paste0(wd,"/input_files/meta/metadata_20190811.RData"))
meta <- meta[meta$ID != "17099",]
gene_ID <- readRDS(paste0(wd,"/input_files/RNA/references/Exon_GeneLoc.RData"))
cds <- readRDS(paste0(wd,"/input_files/RNA/references/CDS_wGoTerms_GeneLOC.RData"))

# Gene Expression 
gc <- readRDS(paste0(wd,"/results/Transcriptomic/DGEListObj_withIndWeights_filterApproach2_plannedContrastMatrix.RData"))

# Methylation Data
d_meth <- readRDS(paste0(wd,"/results/DNAm/Final_DNAm_gene_BRMS_plannedComparisons.RData"))
d_meta <- readRDS(paste0(wd,"/input_files/DNAm/Final_meta_gene_5.RData"))
d_beta <- readRDS(paste0(wd,"/input_files/DNAm/Final_beta_gene_5.RData"))
```

**Differential Methylation Summary Table**

Summarized CpGs that were significantly different after correction for all comparisons. Hypomethylation refers to CpGs that had significant lower methylation in our OA treatment compared to the ambient conditions or on day 80 compared to day 9. The opposity is the case (methylation is higher) if hypermethylated.  

```{r}

cols <- colnames(d_meth$Significant)
sum_mat <- matrix(0,nrow=1,ncol=length(cols))
for(i in 1:length(cols)){
  sum_mat[1,i] <- sum(d_meth$Significant[,i]=="*")
}
final_mat <- matrix(0,nrow=2,ncol=9)
colnames(final_mat) <- c("Trt","Time","Interaction",
                         "Trt_D9",
                         "Ambient_Time",
                         "Ambient_D9_Exposed_D80",
                         "Exposed_D9_Ambient_D80",
                         "Exposed_Time",
                         "Trt_D80"
                         )
row.names(final_mat) <- c("Hypomethylated","Hypermethylated")
final_mat[1,] <- sum_mat[,c(1,3,5,6,8,10,12,14,16)]
final_mat[2,] <- sum_mat[,c(2,4,5,7,9,11,13,15,17)]
final_mat[2,3] <- NA

kableExtra::kable(final_mat) %>% 
  kableExtra::kable_styling()
```

Basic Summary - Long Format
```{r}
name <- colnames(d_meth$Significant)[1]
pos <- which(d_meth$Significant[,1]=="*")
estimate <- d_meth$Estimate[which(d_meth$Significant[,1]=="*"),1]
sig_list <- data.frame(pos=as.numeric(pos),estimate=as.numeric(estimate),name=name)
for(i in 2:ncol(d_meth$Significant)){
  name <- colnames(d_meth$Significant)[i]
  pos <- which(d_meth$Significant[,i]=="*")
  estimate <- d_meth$Estimate[which(d_meth$Significant[,i]=="*"),i]
  temp_bind <- data.frame(pos=as.numeric(pos),estimate=as.numeric(estimate),name=name)
  sig_list <- rbind(sig_list,temp_bind)
}
total_significant <- length(unique(sig_list$pos))

# Order by estimate
sig_list <- sig_list[order(sig_list$estimate),]
```

There were a total of `r total_significant` CpGs significantly differentially methylated in at least one comparison.  

**Table of significant CpGs ranked by the mode estimate (a measure of the magnitude difference between groups)**  
```{r}
kable(head(sig_list)) %>%
  kable_styling()
```

Significant Loci by Treatment
```{r fig.width=12}
# Short investigation of caherin, which had a significantly differentially methylated CpG by treatment
# Row ID : 243889 - Significant by treatment - Cadherin - LOC111106820
# Row ID : 3240 - Significant by treatment (day 9) - transmembrane - LOC111103288
LOC <- 3240
# Gene / Exon Information
gene_ID[match(d_meta$gene_id[LOC],gene_ID$gene_id),]
# CDS and Function annotation
cds[match(d_meta$gene_id[LOC],cds$gene_id),]
# Confirm the ids match up with the gene expression data
gc$genes[match(d_meta$gene_id[LOC],gc$genes$GENEID),]
# Subset gene expression data to the same number of individual and the DNA methylation data
temp <- gc$E[match(d_meta$gene_id[LOC],gc$genes$GENEID),colnames(gc$E) != "17099"]

#Plots
par(mfrow=c(1,3))
boxplot(unlist(d_beta[LOC,])~meta$SFV,ylim=c(0,1),
        xlab="Day.Treatment",ylab="Prop. Methylation")
boxplot(temp~meta$SFV,
        xlab="Day.Treatment",ylab="log2 CPM")
plot(unlist(d_beta[LOC,])~temp,ylim=c(0,1),
     ylab="Prop.Methylation",xlab="log2 CPM")
```
