---
title: "AE17_DNAm_GOMWU"
author: "adowneywall"
date: "11/27/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## **DATA**

**All CpGs**
```{r}
setwd("/shared_lab/20180226_RNAseq_2017OAExp/DNAm/scripts/GO_MWU")
all_files <- list.files(pattern="GOMWU_allCpG")
# Save annotation file name
goAnnotations=paste0(all_files[grep("Goterms",all_files)])
input=paste0(all_files[-grep("Goterms",all_files)])
goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDivision="BP" # either MF, or BP, or CC
source("gomwu.functions.R")
```

**Max E Ratio for an individual cpg for each gene represented**
```{r}
setwd("/shared_lab/20180226_RNAseq_2017OAExp/DNAm/scripts/GO_MWU")
all_files <- list.files(pattern="GOMWU_maxCpG")
# Save annotation file name
goAnnotations=paste0(all_files[grep("Goterms",all_files)])
input=paste0(all_files[-grep("Goterms",all_files)])
goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDivision="BP" # either MF, or BP, or CC
source("gomwu.functions.R")
```

**Diff Gene Expression Data**
```{r}
setwd("/shared_lab/20180226_RNAseq_2017OAExp/DNAm/scripts/GO_MWU")
all_files <- list.files(pattern="GOMWU_LogFoldChangeGE_")
# Save annotation file name
goAnnotations=paste0(all_files[grep("Goterms",all_files)])
input=paste0(all_files[-grep("Goterms",all_files)])
goDatabase="go.obo" # download from http://www.geneontology.org/GO.downloads.ontology.shtml
goDivision="BP" # either MF, or BP, or CC
source("gomwu.functions.R")
```

## GO MWU functions  

**Perform GO MWU**
```{r}
for(i in 1:length(input)){
  # Calculating stats. It might take ~3 min for MF and BP. Do not rerun it if you just want to replot the data with different cutoffs, go straight to gomwuPlot. If you change any of the numeric values below, delete the files that were generated in previos runs first.
  gomwuStats(input[i], goDatabase, goAnnotations, goDivision,
  	perlPath="perl",
  	largest=0.1,
  	smallest=2,
  	clusterCutHeight=0.25
  )
  # do not continue if the printout shows that no GO terms pass 10% FDR.
}
```

**Follow up plots - check if there are any that are significant**
```{r}
  results=gomwuPlot(input,goAnnotations,goDivision,
  	absValue=2,  # genes with the measure value exceeding this will be counted as "good genes".
  	level1=0.1, # FDR threshold for plotting. 
  	#Specify level1=1 to plot all GO categories containing genes exceeding the absValue.
  	level2=0.05, # FDR cutoff to print in regular (not italic) font.
  	level3=0.01, # FDR cutoff to print in large bold font.
  	txtsize=1.2, # decrease to fit more on one page, or increase (after rescaling the plot so the tree fits the text) for better "word cloud" effect
  	treeHeight=0.5, # height of the hierarchical clustering tree
  #	colors=c("dodgerblue2","firebrick1","skyblue2","lightcoral") # these are default colors, un-remar and change if needed
  )
  
  png(paste0(unlist(strsplit(input[i],split = "/|.csv"))[2],".png"))
  results
  dev.off()
```

