---
title: "DNAm Multivariate Analysis"
author: "adowneywall"
date: "9/2/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(edgeR)
library(ashr)
library(vegan)
library(sf)
library(adegenet)
library(ggplot2)
library(ape)
```

### Current files
```{r}
## Meta Data for the Model
model<-readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/meta/metadata_20190811.RData")
model <- model[model$ID !="17099",]
#### Transcript File ####
tran <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/references/STAR_gnomon_tximportGeneFile.RData")
# Gene File
# gene <- tran[!duplicated(tran$GENEID),]
# rm(tran)
# # This is the filtered count matrix
# RSEM <-  readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/STAR_mapping/RSEM_output/RSEM_gene_Summary.Rdata")
# # Separate out RSEM counts and rename rows with LOC ID
# rsem_c <- RSEM$Count
# rm(RSEM)
#Rename loci
# loc_rsem <- gene$GENEID[match(row.names(rsem_c),gene$gene_id)]
# row.names(rsem_c) <- loc_rsem
# #Remove low variation (or no variation) loci (needed for downstream stuff to work)
# keep_rsem <- rowSums(cpm(rsem_c)>1) >= (0.5 * 24)
# rsem_red <- rsem_c[keep_rsem, ]
#counts_Rdat <- readRDS(file = "/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/Normalized_CountMatrix/Scenario1_normalizedVoom.RData")
#log2(seq(0,1,by=0.01))
#counts <- cpm(rsem_red)
#gc_reduce <-  counts #log2(counts+1)

dnaM_beta <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/DNAm/Final_beta_gene_10.RData")
length(which(is.na(dnaM_beta)))
#dnaM_beta[which(is.na(dnaM_beta[,1])),] <- 0

dnaM_betaE <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/DNAm/Final_beta_exon_10.RData")
length(which(is.na(dnaM_betaE)))
dnaM_betaE[which(is.na(dnaM_betaE[,1])),] <- 0
# Subsetting introns
# length(setdiff(rownames(dnaM_beta),rownames(dnaM_betaE)))
# beta_intron_10 <-dnaM_beta[as.numeric(setdiff(rownames(dnaM_beta),rownames(dnaM_betaE))),]
# beta_intron_10 <-dnaM_beta[1:10,]
# dnaM_betaI <- subset(t(dnaM_beta),select=setdiff(rownames(dnaM_beta),rownames(dnaM_betaE)))
# ?setdiff
```

```{r}
(out_gc <- adonis(t(dnaM_beta)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
(out_gc <- adonis(t(dnaM_betaE)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
```

### Plotting data with RDA  

### Plot in multivariate space with RDA (treatment and time)
```{r}
#prin_comp<-rda(t(test2),scale = TRUE)

prin_comp<-rda(t(dnaM_beta))
sum_pri <- summary(prin_comp)
pca_scores<-scores(prin_comp)
colnames(dnaM_beta)
#pca <- prcomp(t(gc_reduce))
pca <- prcomp(t(dnaM_beta))
eigs <- pca$sdev^2

color_comb <- c("lightblue4","royalblue3","tomato","darkred") # colors for population 
model$colors <- "" 
model$colors[model$SFV == unique(model$SFV)[1]] <-  color_comb[2]
model$colors[model$SFV == unique(model$SFV)[2]] <-  color_comb[1]
model$colors[model$SFV == unique(model$SFV)[3]] <-  color_comb[4]
model$colors[model$SFV == unique(model$SFV)[4]] <-  color_comb[3]
model$pch <- 16
model$pch[model$colors == color_comb[2] | model$colors == "darkred"] <- 10

ordiplot(prin_comp,type="n",
         xlab=paste0("PC1  (",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
         ylab=paste0("PC2  (",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))

orditorp(prin_comp,display="sites",labels = FALSE,col=model$colors,cex = 2,pch = model$pch)
#ordiellipse(prin_comp,model$SFV,conf=0.90,col = color_comb,lwd = 3)
ordispider(prin_comp,model$SFV,col = color_comb,lwd=2.5)
legend(x=c(4,14),y=c(12,17),
       legend = c("Day 09 : Ambient",
                  "Day 80 : Ambient",
                  "Day 09 : 2800",
                  "Day 80 : 2800"),
       pch = c(16,10,16,10),col=color_comb,xpd = .25)
text(x = 6 ,y = -5, 
     paste0("P_Treatment = ",round(out_gc$aov.tab$`Pr(>F)`[1],3),"*"),pos = 4)
text(x = 6 ,y = -6, 
     paste0("P_Time = ",round(out_gc$aov.tab$`Pr(>F)`[2],5),"*"),pos=4)
text(x = 6 ,y = -7, 
     paste0("P_Interaction = ",round(out_gc$aov.tab$`Pr(>F)`[5],3)),pos=4)
```




