---
title: "Brief Summary of OA2018 RNASeq Read and Gene Count Data"
date: "March 27, 2019"
output: 
  html_document:
      keep_md: true
editor_options: 
  chunk_output_type: console
---

### Overview
This script provided a brief run down of the reads data from our RNAseq dataset, and looks at how many reads, alignments, and gene counts are produced using our current bioinformatics pipeline. The reason for this is to troubleshoot the reason for the substantial loss of data as we progress to our final gene count matrix.

## Bioinformatics Steps
```{r, echo = FALSE}
library(kableExtra)
Step <- c("Trimming Raw Reads",
          "Mapping Reads",
          "Filtering",
          "Create Count Matrix",
          "Filter Reads with low frequency")
Program <- c("Trimmomatic (implimented in dDocent)",
             "STAR",
             "SamTools",
             "HT-Seq",
             "R")
Link <- c("https://github.com/DrK-Lo/2017OAExp_Oysters/blob/master/markdown_files/01B_CV17_RNA_Alignment_trimming.Rmd",
          "https://github.com/DrK-Lo/2017OAExp_Oysters/blob/master/markdown_files/01C_CV17_RNA_Alignment_ReadMapping.Rmd",
          "https://github.com/DrK-Lo/2017OAExp_Oysters/blob/master/markdown_files/01D_CV17_RNA_Alignment_Filtering.Rmd",
          "https://github.com/DrK-Lo/2017OAExp_Oysters/blob/master/markdown_files/02_CV17_RNA_createCountMatrix.Rmd",
          "https://github.com/epigeneticstoocean/2017OAExp_Oysters/blob/master/markdown_files/STAR_pipeline/03A_CV17_RNA_countFilteringandAnalysis.md")

tab <- data.frame(cbind(Step,Program,Link))
(kable(tab) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed")) -> s_table)
```
  
```{r}
## Raw Data
#Raw Reads from GENEWIZ
setwd(dir = "/home/downeyam/Github/2017OAExp_Oysters/")
rawRead<-read.delim(file = "/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/RNASeq_RawReadInfo_spaceDelim.txt",sep = " ",header = TRUE)
rawRead_sub<-subset(rawRead,select=c("Sample_ID","RawRead","meanQualityScore"))
```
  
```{r, echo = FALSE}
## Adapter Removal and Initial QC
# READ Info from STAR Aligner
library(stringr)
setwd(dir = "/home/downeyam/Github/2017OAExp_Oysters/input_files/STAR_metadata/temp/")
name<-NULL
inp<-NULL
ua<-NULL
pa<-NULL
for(i in 1:length(list.files())){
out<-readLines(con = list.files()[i])
#name
temp_lab<-str_split(list.files()[i],pattern = "_")
temp_sub<-substr(temp_lab,7,11)
name<-append(name,temp_sub)
# input read number
inp <- append(inp,str_split(out[6],pattern = "\t")[[1]][2])
# unique alignment reads
ua <- append(ua,str_split(out[9],pattern = "\t")[[1]][2])
# percent unique alignment reads
pa <- append(pa,str_split(out[10],pattern = "\t")[[1]][2])
}

star_readInfo <- data.frame(Sample_ID=name,InputRead=inp,uniqAlign=ua,perAlign=pa)

merged<-merge(rawRead_sub,star_readInfo,by="Sample_ID")
```
  
```{r,echo=FALSE}
## Gene Count Matrix
##Created with HTSEQ script

#setwd(dir = "/home/downeyam/Github/2017OAExp_Oysters/")
geneCount<-read.delim(file = "/home/downeyam/Github/2017OAExp_Oysters/results/C_virginica_gene_count_final.txt",sep = " ")
colnames(geneCount)<-substr(colnames(geneCount),start = 4,stop = 9)
env<-read.delim(file = "/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/trait_wc.csv",sep = ",")
env_sub<-env[match(as.numeric(colnames(geneCount)),env$bivalve_ID),]

gene_control<-geneCount[,env_sub$treatment == 400]
gene_exposed<-geneCount[,env_sub$treatment == 2800]
```
  
**Complete Summary Table**
```{r,echo=FALSE}
merged<-merged[with(merged, order(Sample_ID)), ]

rawCounts<- data.frame(counts=rowSums(t(geneCount)))
merged_ReadInfo <- data.frame(merged,geneCounts = rawCounts$counts)

kable(merged_ReadInfo) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

**Numbers of genes identified**
```{r, echo=FALSE}
length(geneCount$`17005`)
```

**Histogram of Counts for each Gene (max total number of counts set to 500)**
```{r}
hist(rowSums(geneCount),breaks = 1000000,xlim = c(5,500))
```

### Percent of Count data in the largest 1,5,10% of genes
**1%**  
```{r}
# Minimum counts in the upper 10%
(one_perc <- quantile(rowSums(geneCount),probs = .99))
geneCount_oneperc <- geneCount[rowSums(geneCount) >= one_perc,]
# Number of Genes
nrow(geneCount_oneperc)
# Comparisons of top 1% by treatment
plot(rowSums(geneCount_oneperc[,env_sub$treatment == 400])~rowSums(geneCount_oneperc[,env_sub$treatment == 2800]),
     ylim=c(0,5000),xlim=c(0,5000),main="Total Counts comparison between treatment - top 1%",ylab= "pCO2 400",xlab="pCO2 2800")
abline(col="red",a = 0,b=1)
# Percent of total counts
(sum(geneCount_oneperc/sum(geneCount))*100)
```
**5%**
```{r}
# Minimum counts in the upper 10%
(five_perc <- quantile(rowSums(geneCount),probs = .95))
geneCount_fiveperc <- geneCount[rowSums(geneCount) >= five_perc,]
# Number of Genes
nrow(geneCount_fiveperc)
# Comparisons of top 1% by treatment
plot(rowSums(geneCount_fiveperc[,env_sub$treatment == 400])~rowSums(geneCount_fiveperc[,env_sub$treatment == 2800]),
     ylim=c(0,5000),xlim=c(0,5000),main="Total Counts comparison between treatment - top 5%",ylab= "pCO2 400",xlab="pCO2 2800")
abline(col="red",a = 0,b=1)
# Percent of total counts
(sum(geneCount_fiveperc/sum(geneCount))*100)
```
**10%**
```{r}
# Minimum counts in the upper 10%
(ten_perc <- quantile(rowSums(geneCount),probs = .90))

geneCount_tenperc <- geneCount[rowSums(geneCount) >= ten_perc,]
# Number of Genes
nrow(geneCount_tenperc)
# Comparisons of top 1% by treatment
plot(rowSums(geneCount_tenperc[,env_sub$treatment == 400])~rowSums(geneCount_tenperc[,env_sub$treatment == 2800]),
     ylim=c(0,5000),xlim=c(0,5000),main="Total Counts comparison between treatment - top 5%",ylab= "pCO2 400",xlab="pCO2 2800")
abline(col="red",a = 0,b=1)
# Percent of total counts
(sum(geneCount_tenperc/sum(geneCount))*100)
```
### Gene with Largest Number of Read Counts
```{r}
big_gene <- geneCount[rowSums(geneCount) == max(rowSums(geneCount)),]
# Counts
sum(big_gene)
#Name
row.names(big_gene)
# Percent of Total Reads
(sum(big_gene)/sum(geneCount))*100
```
