---
title: "AE17 EPF pH Timeseries Analysis"
author: "Alan Downey-Wall"
date: "Aug 11, 2019"
output: 
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(mgcv)
library(dplyr)
library(car)
library(lme4)
library(lmerTest)
library(kableExtra)
library(multcomp)
library(multcompView)
library(factoextra)
```

## Data Input and Formatting

**RData version of data**
```{r}
pheno <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/Exposure_SummaryPhenotypeData_without81.RData")
pheno_red <- pheno[pheno$timepoint != 81,]
```

  
## Plans
  
* Look at EPF pH for all timepoints and treatments (6 tp, 3 treatments, 108 individuals, mostly balanced)  
* Look at EPF pH for all timepoints and for the treatments we have sequence data for (6 tp, 2 treatments, ~72 individuals, mostly balanced)  
* Look at EPF pH for all individuals we sequenced (2 tp, 2 treatments, 24 individuals)  
* Look at EPF complete chemistry for samples we sequenced and have complete chemistry (2 tp, 2 treatments, 14 individuals, not balanced)  

## Analysis of EPF pH  - Total data  
  
* Filtering out NAs (one entry) and including only timepoints from the exposure (not acclimation)
```{r echo=FALSE}
epf_exp <- pheno_red[!is.na(pheno_red$EPF_pH),]
epf_exp <-epf_exp[as.numeric(epf_exp$timepoint) > 0,]
```

**Exploratory plot**
```{r echo = FALSE}
n_sample <- epf_exp %>% group_by(pCO2_fac,timepoint_fac) %>% summarise(n=length(pCO2_fac),y.pos=quantile(pH_meas)[4])

give.n <- function(x){
   return(c(y = 7.8, label = length(x)))
}
trt_color <- c("deepskyblue3", "darkorange", "firebrick")
ggplot(epf_exp,aes(y=EPF_pH,x=timepoint_fac,group=interaction(timepoint_fac,pCO2_fac),colour=pCO2_fac)) + 
  geom_boxplot() +
  scale_color_manual(values=trt_color) + 
  theme_bw() +
  stat_summary(fun.data = give.n,
               geom = "text",
               position = position_dodge(width = 0.9), vjust = -0.25) + 
  labs(title="EPF pH",x="Time",y="EPF pH",colour="pCO2")
```
  
### **Statistical Analysis**  
  
#### **EPF pH total data - treatment (factor) time (continuous)**  
**Overview**  
Test: ANCOVA - Full Model (Fixed and Random Factors)  
* Predictors (fixed effects): Treatment (pCO2_fac) and Time (timepoint)  
* Random Factors: Population, Shelf, Tank (nested in Shelf)   
* Tested for normality and variance assumptions  *

**Full Model**  
```{r}
epfAllTP_full <- lmer(EPF_pH~pCO2_fac*timepoint + (1|PopOrigin) + (1|shelf/tank),data=epf_exp) 
# Warning singular fit - Indicates model is overfit and some of the the random effects will need to be removed. I use a ranova below to determine which effects are uninformative and are candidates for removal.

# Check distribution of residuals
plot(epfAllTP_full)
# They look fairly homoscedastic
qqnorm(resid(epfAllTP_full))
qqline(resid(epfAllTP_full))

anova(whole_factor_full)
# Nothing significant here (although the interaction is close)
ranova(whole_factor_full)
# We should remove population and shelf as random factors and rerun analysis
```
  
**Rerun Model without Population and Shelf**
```{r}
epfAllTP_red <- lmer(EPF_pH~pCO2_fac*timepoint + (1|tank:shelf),data=epf_exp)
# Prevented singular fit issue this time

# Check distribution of residuals
plot(epfAllTP_red)
# They look fairly homoscedastic
qqnorm(resid(epfAllTP_red))
qqline(resid(epfAllTP_red))
# Fairly normal
# Check autocorrelation
ggplot(epf_exp,aes(y=resid(epfAllTP_red),x=timepoint_fac,group=interaction(pCO2_fac,timepoint_fac),colour=pCO2_fac)) + geom_boxplot()
ggplot(epf_exp,aes(y=resid(epfAllTP_red),x=timepoint,colour=pCO2_fac)) + geom_point()
# Nothing obvious
summary(epfAllTP_red)
# Significant interaction between pCO2_2800:time

# Compare full model to one with tank as the only random effect
anova(whole_factor_full,whole_factor_tank)
ranova(epfAllTP_red) # Significant tank effect
# Comparing the two models  (full and w/ tank only there is no difference)
# We keep the simpler model
```
  
#### **EPF pH total data - treatment (continuous) and time (continuous)**  
**Overview**  
Test: Regression
* Predictors: Mean Tank Treatment (pCO2_calc) and Time (timepoint)  
* Random Factors: Population, Shelf, Tank  
* Tested for normality and variance assumptions  
  
```{r}
# Had to scale these values due to large differences in variance
whole_cont_full <- lmer(EPF_pH~scale(epf_exp$pCO2_calc)*scale(epf_exp$timepoint) + (1|PopOrigin) + (1|tank:shelf) ,data=epf_exp)
# Warning -  looks like this is driven by too many explanatory variables
ranova(whole_cont_full)
step(whole_cont_full)
# Looks like we can remove population
# Ill start by removing population and shelf and reruning
```

**Rerun model with just tank as random factor**
```{r}
whole_cont_tank <- lmer(EPF_pH~scale(epf_exp$pCO2_calc)*scale(epf_exp$timepoint) + (1|tank:shelf) ,data=epf_exp)
# Better no warning this time 

# Check distribution of residuals
plot(whole_cont_tank)
# They look fairly homoscedastic
qqnorm(resid(whole_cont_tank))
qqline(resid(whole_cont_tank))
# Fairly normal

summary(whole_cont_tank)
# Treatment and intereaction significant.
```
  
#### **EPF pH total data - treatment (factor) time (factor)**  
**Overview**  
Test: 2-way ANOVA - Full Model (Fixed and Random Factors)  
* Prdictors (fixed effects): Treatment (pCO2_fac) and Time (timepoint)  
* Random Factors: Population, Shelf, Tank (nested in Shelf)   
* Tested for normality and variance assumptions  *

**Full Model**  
```{r}
epfAllTP_full <- lmer(EPF_pH~pCO2_fac*timepoint_fac + (1|PopOrigin) + (1|shelf/tank),data=epf_exp) 
# Warning singular fit - Indicates model is overfit and some of the the random effects will need to be removed. I use a ranova below to determine which effects are uninformative and are candidates for removal.

# Check distribution of residuals
plot(epfAllTP_full)
# They look fairly homoscedastic
qqnorm(resid(epfAllTP_full))
qqline(resid(epfAllTP_full))

anova(whole_factor_full)
# Nothing significant here (although the interaction is close)
ranova(whole_factor_full)
# We should remove population and shelf as random factors and rerun analysis
```
  
**Rerun Model without Population and Shelf**
```{r}
epfAllTP_red <- lmer(EPF_pH~pCO2_fac*timepoint_fac + (1|tank:shelf),data=epf_exp)
# Prevented singular fit issue this time

# Check distribution of residuals
plot(epfAllTP_red)
# They look fairly homoscedastic
qqnorm(resid(epfAllTP_red))
qqline(resid(epfAllTP_red))
# Fairly normal
# Nothing obvious
anova(epfAllTP_red)
Anova(epfAllTP_red,type = c("3"))


# Compare full model to one with tank as the only random effect
anova(whole_factor_full,whole_factor_tank)
ranova(epfAllTP_red) # Significant tank effect
# Comparing the two models  (full and w/ tank only there is no difference)
# We keep the simpler model
summary(epfAllTP_red)


## Planned comparisons ##
group <- paste0(pheno_red$pCO2_fac,":",pheno_red$timepoint_fac)
mod_matrix <- model.matrix(epfAllTP_red)
agg_mod_matrix <- aggregate(mod_matrix~group,FUN=mean)
rownames(agg_mod_matrix) <- agg_mod_matrix$group
agg_mod_matrix <- agg_mod_matrix[,-1]

lc2 <- as.matrix(agg_mod_matrix)

summary(glht(epfAllTP_red,linfct=lc2),adjusted(type = "fdr"))

k <- rbind("400v900_1"=lc2["900:1",]-lc2["400:1",],
           "400v2800_1"=lc2["2800:1",]-lc2["400:1",],
           "900v2800_1"=lc2["2800:1",]-lc2["900:1",],
           "400v900_2"=lc2["900:2",]-lc2["400:2",],
           "400v2800_2"=lc2["2800:2",]-lc2["400:2",],
           "900v2800_2"=lc2["2800:2",]-lc2["900:2",],
           "400v900_9"=lc2["900:9",]-lc2["400:9",],
           "400v2800_9"=lc2["2800:9",]-lc2["400:9",],
           "900v2800_9"=lc2["2800:9",]-lc2["900:9",]
           )
summary(glht(epfAllTP_red,linfct=k),adjusted(type = "fdr"))

# Nothing seems significant thus far
```

### Relative EPF pH
  
#### **EPF pH total data - treatment (continuous) and time (continuous)**  
**Overview**  
Test: 2-way ANCOVA 
* Predictor: Mean Tank Treatment (pCO2_calc)   
* Continuous covariate: Time (timepoint)  
* Random Facotrs: Population, Shelf, Tank  
* Tested for normality and variance assumptions  
  
```{r}
# Had to scale these values due to large differences in variance
whole_cont_full <- lmer(EPF_pH~scale(epf_exp$pCO2_calc)*scale(epf_exp$timepoint) + (1|PopOrigin) + (1|tank:shelf) ,data=epf_exp)
# Warning -  looks like this is driven by too many explanatory variables
ranova(whole_cont_full)
step(whole_cont_full)
# Looks like we can remove population
# Ill start by removing population and shelf and reruning
```

**Rerun model with just tank as random factor**
```{r}
whole_cont_tank <- lmer(EPF_pH~scale(epf_exp$pCO2_calc)*scale(epf_exp$timepoint) + (1|tank:shelf) ,data=epf_exp)
# Better no warning this time 

# Check distribution of residuals
plot(whole_cont_tank)
# They look fairly homoscedastic
qqnorm(resid(whole_cont_tank))
qqline(resid(whole_cont_tank))
# Fairly normal

summary(whole_cont_tank)
# Treatment and intereaction significant.
```
  
### Pairwise T-tests for each experimental timepoint  
  
```{r}
c_exp <- epf_exp[epf_exp$pCO2_fac == 400,]
e_exp <- epf_exp[epf_exp$pCO2_fac == 2800,]

pair_T <- NULL
for (i in 1:length(unique(c_exp$timepoint))){
  day1_c <- c_exp[c_exp$timepoint == unique(c_exp$timepoint[order(c_exp$timepoint)])[i],]
  day1_e <- e_exp[e_exp$timepoint == unique(c_exp$timepoint[order(c_exp$timepoint)])[i],]
  temp <- t.test(day1_c$EPF_pH,day1_e$EPF_pH)
  pair_T <- c(pair_T,temp$p.value)
}
names(pair_T) <- unique(c_exp$timepoint[order(c_exp$timepoint)])
(final_p <- p.adjust(pair_T,method = "BH"))

```
  
## Analysis of EPF pH  - Only the two treatments we analyzed for sequencing.  
```{r}
ee_treat <- epf_exp[epf_exp$pCO2_fac != 900,]
```

**Exploratory plot**  
```{r}
n_sample <- ee_treat %>% group_by(pCO2_fac,timepoint_fac) %>% summarise(n=length(pCO2_fac),y.pos=quantile(pH_meas)[4])

give.n <- function(x){
   return(c(y = 7.8, label = length(x)))
}
trt_color <- c("deepskyblue3","firebrick")

ggplot(ee_treat,aes(y=EPF_pH,x=timepoint_fac,group=interaction(timepoint_fac,pCO2_fac),colour=pCO2_fac)) + 
  geom_boxplot() +
  scale_color_manual(values=trt_color) + 
  theme_bw() +
  stat_summary(fun.data = give.n,
               geom = "text",
               position = position_dodge(width = 0.9), vjust = -0.25) + 
  labs(title="EPF pH",x="Time",y="EPF pH",colour="pCO2")

ee_treat$timepoint_num <- as.numeric(ee_treat$timepoint)
ggplot(ee_treat, aes(y=EPF_pH,x=timepoint_num,group=pCO2_fac,color=pCO2_fac)) +
  geom_point() +
  geom_smooth(method=loess,se=FALSE,span=0.9) +
  labs(title="Scatter Plot fit with loess line")

treatSeq_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,ee_treat,FUN=mean)
SE <- function(x){
  sd(x)/(sqrt(length(x)))
}
treatSeq_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,ee_treat,FUN=SE)

mean_config <- rbind(treatSeq_means$EPF_pH[1:6],treatSeq_means$EPF_pH[7:12])
SE_config <- rbind(treatSeq_SE $EPF_pH[1:6],treatSeq_SE $EPF_pH[7:12])

bp <- barplot(mean_config,ylim=c(6.5,8.1),col=c("lightblue","tomato"),ylab = "EPF pH",xpd = FALSE,beside = TRUE)
arrows(x0 = bp, x1 = bp, y0 = mean_config - SE_config, y1 = mean_config +
SE_config, angle = 90, len = 0.1, code = 3, xpd = NA)
text(x = bp, y = 6.4, label = c("Control","OA","Control","OA"), xpd = NA)
text(x = mean(bp[1:2]), y = 6.3, label = "24 Hr", xpd = NA)
text(x = mean(bp[3:4]), y = 6.3, label = "36 Hr", xpd = NA)
text(x = mean(bp[5:6]), y = 6.3, label = "Day 9", xpd = NA)
text(x = mean(bp[7:8]), y = 6.3, label = "Day 22", xpd = NA)
text(x = mean(bp[9:10]), y = 6.3, label = "Day 50", xpd = NA)
text(x = mean(bp[11:12]), y = 6.3, label = "Day 80", xpd = NA)
box()
```
  
### **Statistcal Analysis**  
  
#### **EPF pH total data - time as factor**  
**Overview**  
Test: 2-way ANOVA - Full Model (Fixed and Random Factors)  
* Explanatory Factors: Treatment (pCO2_fac) and Time (timepoint_fac)  
* Random Factors: Population, Shelf, Tank (nested in Shelf)   
* Tested for normality and variance assumptions  
  
```{r}
partial_factor_full <- lmer(EPF_pH~timepoint_fac*pCO2_fac + (1|PopOrigin) + (1|shelf/tank),data=ee_treat) #
# Warning singular fit

# Check distribution of residuals
plot(partial_factor_full)
# They look fairly homoscedastic
qqnorm(resid(partial_factor_full))
qqline(resid(partial_factor_full))

anova(partial_factor_full)
# Nothing significant here (although the interaction is close)

ranova(partial_factor_full)
# We should remove population and shelf as random factors and rerun analysis
```
  
**Rerun Model with Population and Shelf removed as random facteors**
```{r}
partial_factor_tank <- lmer(EPF_pH~timepoint_fac*pCO2_fac + (1|tank:shelf),data=ee_treat)

# Prevented singular fit issue this time

# Check distribution of residuals
plot(partial_factor_tank)
# They look fairly homoscedastic
qqnorm(resid(partial_factor_tank))
qqline(resid(partial_factor_tank))
# Fairly normal

anova(partial_factor_tank)

anova(partial_factor_full,partial_factor_tank)
# Comparing the two models  (full and w/ tank only there is no difference)
# We keep the simpler model

ee_treat$timepCO2 <-  factor(interaction(ee_treat$timepoint_fac,ee_treat$pCO2_fac))

hl_tank_v2 <- lmer(EPF_pH~timepCO2 + (1|tank:shelf),data=ee_treat)
# If you wanted to do a complete pairwise comparison
#summary(glht(hl_tank_v2, linfct = mcp(timepCO2 = "Tukey")), test = adjusted("holm"))
```

#### **EPF pH total data - treatment as factor and time as continuous variable**  
**Overview**  
Test: 2-way ANCOVA 
* Explainatory Factors: Treatment (pCO2_fac)   
* Continuous covariate: Time (timepoint)  
* Random Facotrs: Population, Shelf, Tank  
* Tested for normality and variance assumptions  
  
```{r}
partial_cont_full <- lmer(EPF_pH~timepoint*pCO2_fac + (1|PopOrigin) + (1|shelf/tank) ,data=ee_treat)
# ERROR -  looks like this is driven by too many explanatory parameters

ranova(partial_cont_full)
# Looks like we can remove the random effects anyways
# Ill start by removing population and shelf and reruning
```

**Rerun model with just tank as random factor**
```{r}
partial_cont_tank <- lmer(EPF_pH~timepoint*pCO2_fac + (1|tank:shelf) ,data=ee_treat)

# Check distribution of residuals
plot(partial_cont_tank)
# They look fairly homoscedastic
qqnorm(resid(partial_cont_tank))
qqline(resid(partial_cont_tank))
# Fairly normal

ranova(partial_cont_tank)
# Not significant but good to keep i think

anova(partial_cont_tank)
summary(partial_cont_tank)
# Nothing significant :(
```  

**Automated Model Selection**
```{r}
step(partial_cont_full)
# Model Select only pCO2fac and tank
```
  
## Analysis of EPF pH -  Only samples we sequenced  

#### **EPF pH total data - time as factor**  
**Overview**  
Test: 2-way ANOVA - Full Model (Fixed and Random Factors)  
* Explanatory Factors: Treatment (pCO2_fac) and Time (timepoint_fac)  
* Random Factors: Population, Shelf, Tank (nested in Shelf)   
* Tested for normality and variance assumptions  

```{r}
pheno_seq <- pheno_red[pheno_red$sequenced == "Y",]
seq_full <- lmer(EPF_pH~timepoint_fac*pCO2_fac + (1|PopOrigin) + (1|shelf/tank),data=pheno_seq)
# Singular fit warning

anova(seq_full) # Everything is signficant

plot(seq_full)
# Some heteroscedascity but not too bad
qqnorm(resid(seq_full))
qqline(resid(seq_full))

ranova(seq_full) # None of these seem important

anova(seq_full) # Everything except pCO2 significant

# Decide to remove random factors since they aren't informative
```

**Rerun with standard lm**
```{r}
seq_fixed <- lm(EPF_pH~timepoint_fac*pCO2_fac, data=pheno_seq)

plot(seq_fixed)
# We can see more clearly that some of the mild issues with the assumptions are being driven by 1 (61).

anova(seq_fixed) # without random factors, pCO2 is now even less significant
```
  
**Rerun but remove the one problematic points**
```{r}
# New dataset w/o the one point
pheno_seq2 <- pheno_seq[row.names(pheno_seq) != c(61),]

seq_fixed_outlierRM <- lm(EPF_pH~timepoint_fac*pCO2_fac, data=pheno_seq2)

plot(seq_fixed_outlierRM)
# On a whole this looks like a slight improvement

anova(seq_fixed_outlierRM) # Now everything is significant
```
  
**Pairwise comparisons of fixed factor model (with points rm)**  
```{r}
seq_pairwise <- TukeyHSD(aov(EPF_pH~timepoint_fac*pCO2_fac, data=pheno_seq2))

plot(seq_pairwise$`timepoint_fac:pCO2_fac`)
```
  
**Final Figure**  
```{r echo=FALSE}

seq2_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,pheno_seq2,FUN=mean)
SE <- function(x){
  sd(x)/(sqrt(length(x)))
}
seq2_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,pheno_seq2,FUN=SE)

seq_labels <- multcompLetters(seq_pairwise$`timepoint_fac:pCO2_fac`[, "p adj"])$Letters
seq_labels <- seq_labels[c(1,2,4,3)]


bp <- barplot(seq2_means$EPF_pH,ylim=c(6.5,8.1),col=c("blue","red","blue","red"),ylab = "EPF pH",
              xpd = FALSE)
arrows(x0 = bp, x1 = bp, y0 = seq2_means$EPF_pH - seq2_SE$EPF_pH, y1 = seq2_means$EPF_pH +
seq2_SE$EPF_pH, angle = 90, len = 0.1, code = 3, xpd = NA)
text(x = bp, y = 6.4, label = c("Control","OA","Control","OA"), xpd = NA)
text(x = mean(bp[1:2]), y = 6.3, label = "Day 9", xpd = NA)
text(x = mean(bp[3:4]), y = 6.3, label = "Day 80", xpd = NA)
box()
text(x = bp, y = seq2_means$EPF_pH + seq2_SE$EPF_pH, pos = 3, lab = seq_labels, xpd = NA)
```
  
Final bar plot has SE bars and significance levels based on pairwise comparison test (tukey method) based on model selected above.
  
**Automated Model Selection**
```{r}
#step(seq_full)
# Automated processed converged on the same model as above
```
  
## Analysis of EPF pH, DIC, and Ca Saturation - All samples data with **Complete Carbonate Chemistry** (WARNING VERY LOW SAMPLE SIZE)  

* This is data that we measured both the EPF pH and DIC of extract EPF Fluid in order to calculate the complete carbonate chemistry  
```{r}
#Reducing data to online those with DIC
epf_wDIC <- pheno_red[pheno_red$carb_chem == "Y",]
DICnum <- as.numeric(as.character(epf_wDIC$DIC))

epf_wDIC$DICnum <- DICnum
epf_wDIC_red <- epf_wDIC[!is.na(epf_wDIC$DICnum),]

```
**Exploratory Plots**  
```{r echo=FALSE}
#Color Scheme
# Default orderL 400,900,2800
trt_color <- c("deepskyblue3", "darkorange", "firebrick")
#pH
give.n <- function(x){
   return(c(y = 7.9, label = length(x)))
}
ggplot(epf_wDIC_red,aes(x=timepoint_fac,y=EPF_pH,group=interaction(pCO2_fac,timepoint),colour=pCO2_fac)) + geom_boxplot()
  
ggplot(epf_wDIC_red,aes(x=timepoint_fac,y=EPF_DIC,group=interaction(pCO2_fac,timepoint),colour=pCO2_fac)) + geom_boxplot()

# geom_hline(yintercept= sum_WC$pH,col=trt_color,size=1) + 
#   geom_boxplot() + 
#   scale_color_manual(values=trt_color) + 
#   theme_bw() +
#   stat_summary(fun.data = give.n,
#                geom = "text",
#                position = position_dodge(width = 0.9), vjust = -0.25) +
#   labs(title="EPF pH",x="Time",y="EPF pH",colour="pCO2")
# Calcite
# ggplot(epf_s,aes(x=timepoint_fac,y=EPF_Ca_Start,group=interaction(pCO2_fac,timepoint_fac),colour=pCO2_fac)) + 
#   geom_hline(yintercept= sum_WC$Calcite,col=trt_color,size=1) + 
#   geom_boxplot() + 
#   scale_color_manual(values=trt_color) + 
#   theme_bw() +
#   labs(title="EPF Calcite",x="Time",y="EPF Ca",colour="pCO2")
# DIC
# ggplot(epf_s,aes(x=timepoint_fac,y=EPF_DIC_Start,group=interaction(pCO2_fac,timepoint_fac),colour=pCO2_fac)) + 
#   geom_hline(yintercept= sum_WC$DIC,col=trt_color,size=1) + 
#   geom_boxplot() + 
#   scale_color_manual(values=trt_color) + 
#   theme_bw() +
#   labs(title="EPF DIC",x="Time",y="EPF DIC",colour="pCO2")
```
**Figure Notes**: Horizontal lines represent the mean environment value for each the the treatment levels.
  
**Subsetting data for stats**
```{r}
# # Lets only look at exposure timepoints here
# epf_s_Exp <- epf_s[epf_s$timepoint > 2,]
# # Only two treatment levels we have molecular data for (~400,~2800)
# epf_s_Exp <- epf_s_Exp[!epf_s_Exp$pCO2_fac==900,]
```
  
#### **Statistical Analysis**

**Overview**  
Test: 2-way ANOVA - 
* Explainatory Factors: Treatment (pCO2_fac) and Time (timepoint_fac)
* Tested for normality and variance assumptions - log transformed EPF data if needed
  
**NOTE** : Due to extremely lower power hear I don't think we really even have the option of using random effects  

**EPF Calcite Saturation**  
```{r}
mod_ca <- lm(EPF_Ca_Start~pCO2_fac*timepoint_fac,data=epf_wDIC_red)

qqnorm(mod_ca$residuals)
qqline(mod_ca$residuals)

# resids a little wonky trying a transformation
mod_ca_log <- lm(log(EPF_Ca_Start)~pCO2_fac*timepoint_fac,data=epf_wDIC_red)

leveneTest(log(epf_wDIC_red$EPF_Ca_Start)epf_wDIC_red$pCO2_fac*epf_wDIC_red$timepoint_fac)
# Seem to be OK

qqnorm(mod_ca_log$residuals)
qqline(mod_ca_log$residuals)
# Better than mod_1 before transformation

mod_ca_2 <-  lmer(log(EPF_Ca_Start)~pCO2_fac*timepoint_fac+(1|pop)+(1|shelf/tank),data=epf_wDIC_red)
qqnorm(resid(mod_ca_2))
qqline(resid(mod_ca_2))

## RANOVA to check random effects  
ranova(mod_ca_2)
# None are significant, check to see if the model is better
anova(mod_ca_2,mod_ca_log)
# Nope
#Using simple lm, random effects aren't improving the model.
# NOTE i didn't continuing using random effects because i don't think we have the sample size to try.

## Running ANOVA
(aov_ca <-aov(mod_ca_log))
summary(aov_ca)
# :( time is significant but nothing else
```
  
**DIC**  
```{r}
mod_dic <- lm(log(EPF_DIC)~pCO2_fac*timepoint_fac,data=epf_wDIC_red)

qqnorm(mod_dic$residuals)
qqline(mod_dic$residuals)

leveneTest(log(epf_wDIC_red$EPF_DIC)~epf_wDIC_red$pCO2_fac*epf_wDIC_red$timepoint_fac)
# Seem to be OK

## Running ANOVA
(aov_dic <-aov(mod_dic))
summary(aov_dic)
# :( close but not significant (might be limited power here)
```

**pH**  
```{r}
mod_pH <- lm(EPF_pH~pCO2_fac*timepoint_fac,data=epf_wDIC_red)

qqnorm(mod_pH$residuals)
qqline(mod_pH$residuals)

leveneTest(epf_wDIC_red$EPF_pH~epf_wDIC_red$pCO2_fac*epf_wDIC_red$timepoint_fac)
# Seem to be OK

## Running ANOVA
(aov_pH <-aov(mod_pH))
summary(aov_pH)
# Everything is significant
```

## Final Summary  
  
**General Observations**: The data was analyzed primarily using ANOVAs based on linear mixed models that included tank as a random effect and pCO2 treatment and time as fixed effects. Time in this data set was primarily treated as a factor (along with pCO2 treatment) given the non linear relationship with EPF pH and time and the lack of a truly robust timeseries for most comparisons (however it might still be a good idead to examine the complete EPF pH timeseries with additional analyses). Generally, pCO2 was found to have a significant effect on EPF pH, and the interaction of time and treatment was found to also often be significant. Importantly, when just considering the two timepoints with sequence data (Day 9 and day 80) treatment, time, and the interaction were all significant.

**Break down of the final models by organizational level**  
  
Full Dataset (All timepoints, 3 treatments)  
* Best Model : EPF_pH ~ Treatment * Time + (1|Tank:Shelf)  
```{r echo=FALSE}
kable(anova(whole_factor_tank)) %>% kable_styling()
```
  
Full Timeseries w high and control treatments (All timepoints, 2 treatments)  

* Best Model : EPF_pH ~ Treatment * Time + (1|Tank:Shelf)  
```{r echo=FALSE}
kable(anova(partial_factor_tank)) %>% kable_styling()
```
  
EPF pH from sequenced individuals (2 timepoints, 2 treatments, 24 samples)  

* Best Model : EPF_pH ~ Treatment * Time + (1|Tank:Shelf)  
```{r echo=FALSE}
kable(anova(seq_fixed_outlierRM)) %>% kable_styling()
```
  
All samples with complete carbonate chemistry (~3 timepoints, 2 treatments )  
  
* Best Model pH :  EPF_pH ~ Treatment * Time  
```{r echo=FALSE}
kable(anova(aov_pH)) %>% kable_styling()
```
* Best Model DIC :  EPF_DIC ~ Treatment * Time  
```{r echo=FALSE}
kable(anova(aov_dic)) %>% kable_styling()
```
* Best Model Calcite Saturation :  EPF_CaSat ~ Treatment * Time  
```{r echo=FALSE}
kable(anova(aov_ca)) %>% kable_styling()
```
  
  
## Final Figure  
  
**Option 1**  
  
```{r echo=FALSE}
par(mfrow=c(1,2))
treatSeq_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,ee_treat,FUN=mean)
treatSeq_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,ee_treat,FUN=SE)
mean_config <- rbind(treatSeq_means$EPF_pH[1:6],treatSeq_means$EPF_pH[7:12])
SE_config <- rbind(treatSeq_SE $EPF_pH[1:6],treatSeq_SE $EPF_pH[7:12])

bp <- barplot(mean_config,ylim=c(6.5,8.1),col=c("lightblue","tomato"),ylab = "EPF pH",xpd = FALSE,beside = TRUE)
arrows(x0 = bp, x1 = bp, y0 = mean_config - SE_config, y1 = mean_config +
SE_config, angle = 90, len = 0.05, code = 3, xpd = NA, lwd = 2)
text(x = bp, y = 6.45, label = c("Ctrl","OA","Ctrl","OA"), xpd = NA,srt=90)
text(x = mean(bp[1:2]), y = 6.35, label = "24 Hr", xpd = NA)
text(x = mean(bp[3:4]), y = 6.35, label = "36 Hr", xpd = NA)
text(x = mean(bp[5:6]), y = 6.35, label = "Day 9", xpd = NA)
text(x = mean(bp[7:8]), y = 6.35, label = "Day 22", xpd = NA)
text(x = mean(bp[9:10]), y = 6.35, label = "Day 50", xpd = NA)
text(x = mean(bp[11:12]), y = 6.35, label = "Day 80", xpd = NA)
text(x=-2,y=8.2,label="A",cex = 2, xpd = NA)
box()


# With outliers removed
# seq2_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,just_seq2,FUN=mean)
# SE <- function(x){
#   sd(x)/(sqrt(length(x)))
# }
# seq2_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,just_seq2,FUN=SE)
# 
# seq_labels <- multcompLetters(seq_pairwise$`timepoint_fac:pCO2_fac`[, "p adj"])$Letters
# seq_labels <- seq_labels[c(1,2,4,3)]
# 
# seq2_meanMat <- rbind(seq2_means$EPF_pH[1:2],seq2_means$EPF_pH[3:4])
# seq2_SEMat <- rbind(seq2_SE$EPF_pH[1:2],seq2_SE$EPF_pH[3:4])
# 
# bp <- barplot(seq2_meanMat,ylim=c(6.5,8.1),col=c("lightblue","tomato"),ylab = "EPF pH",
#               xpd = FALSE,beside=TRUE)
# arrows(x0 = bp, x1 = bp, y0 = seq2_meanMat - seq2_SEMat, y1 = seq2_meanMat +
# seq2_SEMat, angle = 90, len = 0.05, code = 3, xpd = NA,lwd=2)
# text(x = bp, y = 6.45, label = c("Ctrl","OA","Ctrl","OA"), xpd = NA)
# text(x = mean(bp[1:2]), y = 6.35, label = "Day 9", xpd = NA)
# text(x = mean(bp[3:4]), y = 6.35, label = "Day 80", xpd = NA)
# box()
# text(x = bp, y = seq2_meanMat + seq2_SEMat, pos = 3, lab = seq_labels, xpd = NA)

# With outliers retained
seq_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,pheno_seq2,FUN=mean)
SE <- function(x){
  sd(x)/(sqrt(length(x)))
}
seq_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,pheno_seq2,FUN=SE)

seq_labels <- multcompLetters(seq_pairwise$`timepoint_fac:pCO2_fac`[, "p adj"])$Letters
seq_labels <- seq_labels[c(1,2,4,3)]

seq_meanMat <- rbind(seq_means$EPF_pH[1:2],seq_means$EPF_pH[3:4])
seq_SEMat <- rbind(seq_SE$EPF_pH[1:2],seq_SE$EPF_pH[3:4])

bp <- barplot(seq_meanMat,ylim=c(6.5,8.1),col=c("lightblue","tomato"),ylab = "EPF pH",
              xpd = FALSE,beside=TRUE)
arrows(x0 = bp, x1 = bp, y0 = seq_meanMat - seq_SEMat, y1 = seq_meanMat +
seq_SEMat, angle = 90, len = 0.05, code = 3, xpd = NA,lwd=2)
text(x = bp, y = 6.45, label = c("Ctrl","OA","Ctrl","OA"), xpd = NA)
text(x = mean(bp[1:2]), y = 6.35, label = "Day 9", xpd = NA)
text(x = mean(bp[3:4]), y = 6.35, label = "Day 80", xpd = NA)
text(x=-.15,y=8.2,label="B",cex = 2, xpd = NA)
text(x = bp, y = seq_meanMat + seq_SEMat, pos = 3, lab = seq_labels, xpd = NA)
box()
  ```
  
Fig 1. EPF pH in different OA treatments over time. (A) Complete timeseries plotted with timepoint as a factor on the x-axis and standard error bars. (B) The two timepoints we have sequenced for gene expression and DNA methylation.  

**Option 2**  
```{r echo=FALSE}
treatSeq_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,ee_treat,FUN=mean)
treat_means_ctrl <- treatSeq_means[treatSeq_means$pCO2_fac == "400",]
treat_means_oa <- treatSeq_means[treatSeq_means$pCO2_fac == "2800",]

treatSeq_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,ee_treat,FUN=SE)
treat_SE_ctrl <- treatSeq_SE[treatSeq_SE$pCO2_fac == "400",]
treat_SE_oa <- treatSeq_SE[treatSeq_SE$pCO2_fac == "2800",]

# Environment mean
c_mean <- mean(c_exp$pH_meas)
e_mean <- mean(e_exp$pH_meas)

pdf("/home/downeyam/Desktop/Plot3.pdf", width = 11.5, height = 5.5)
m <- rbind(c(1, 1), c(2, 3))
layout(m,height=c(2,2),width=(c(2,2)))
par(mar = c(5, 4, 0, 0))
#par(mfrow=c(1,2))
  
bp <- plot(treat_means_ctrl$EPF_pH~as.numeric(as.character(treat_means_ctrl$timepoint_fac)),ylab="EPF pH",xlab="Time (Days)",col="lightblue4",ylim = c(6.5,8.1))
    
    abline(h=c_mean,col="lightblue4",lty=2)
    abline(h=e_mean,col="tomato",lty=2)
    
    lines(treat_means_ctrl$EPF_pH~as.numeric(as.character(treat_means_ctrl$timepoint_fac)),col="lightblue4")
    arrows(x0 = as.numeric(as.character(treat_means_ctrl$timepoint_fac)), x1 = as.numeric(as.character(treat_means_ctrl$timepoint_fac)), y0 = c(treat_means_ctrl$EPF_pH - treat_SE_ctrl$EPF_pH), y1 = treat_means_ctrl$EPF_pH +
    treat_SE_ctrl$EPF_pH, angle = 90, len = 0.05, code = 3, xpd = NA, lwd = 2,col="lightblue3")
    
    points(treat_means_oa$EPF_pH~as.numeric(as.character(treat_means_oa$timepoint_fac)),col="tomato")
    lines(treat_means_oa$EPF_pH~as.numeric(as.character(treat_means_oa$timepoint_fac)),col="tomato")
    arrows(x0 = as.numeric(as.character(treat_means_oa$timepoint_fac)), x1 = as.numeric(as.character(treat_means_oa$timepoint_fac)), y0 = c(treat_means_oa$EPF_pH - treat_SE_oa$EPF_pH), y1 = treat_means_oa$EPF_pH +
    treat_SE_oa$EPF_pH, angle = 90, len = 0.05, code = 3, xpd = NA, lwd = 2,col="tomato")
    legend(x=45,y=8.15,legend = c("Control","OA"),col = c("lightblue3","tomato"),lty = 1, cex = 0.9,lwd = 2,bty = "n")
    text(x=-10,y=8.27,label="A",cex = 2, xpd = NA)
    box()
  # With outliers retained
  seq_means <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,just_seq,FUN=mean)
  SE <- function(x){
    sd(x)/(sqrt(length(x)))
  }
  seq_SE <- aggregate(EPF_pH~timepoint_fac+pCO2_fac,just_seq,FUN=SE)
  
  seq_labels <- multcompLetters(seq_pairwise$`timepoint_fac:pCO2_fac`[, "p adj"])$Letters
  seq_labels <- seq_labels[c(1,2,4,3)]
  
  seq_meanMat <- rbind(seq_means$EPF_pH[1:2],seq_means$EPF_pH[3:4])
  seq_SEMat <- rbind(seq_SE$EPF_pH[1:2],seq_SE$EPF_pH[3:4])
  
  bp <- barplot(seq_meanMat,ylim=c(6.5,8.1),col=c("lightblue","tomato"),ylab = "EPF pH",
                xpd = FALSE,beside=TRUE)
  arrows(x0 = bp, x1 = bp, y0 = seq_meanMat - seq_SEMat, y1 = seq_meanMat +
  seq_SEMat, angle = 90, len = 0.05, code = 3, xpd = NA,lwd=2)
  text(x = bp, y = 6.45, label = c("Ctrl","OA","Ctrl","OA"), xpd = NA)
  text(x = mean(bp[1:2]), y = 6.35, label = "Day 9", xpd = NA)
  text(x = mean(bp[3:4]), y = 6.35, label = "Day 80", xpd = NA)
  text(x=-.15,y=8.2,label="B",cex = 2, xpd = NA)
  text(x = bp, y = seq_meanMat + seq_SEMat, pos = 3, lab = seq_labels, xpd = NA)
  box()
  
  dev.off()
```
  
Fig 2. EPF pH in different OA treatments over time. (A) Complete timeseries plotted with timepoint as a continuous variable on the x-axis and standard error bars. (B) The two timepoints we have sequenced for gene expression and DNA methylation.  

## Creating RData files for phenotype and water chemistry data using raw data files
```{r, eval=FALSE}
### EPF fluid ###
#epf_cal <- read.csv(file = "/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/EPF/AE17_EPF_Calcification_20190403.csv")

# # Remove and NAs at end of dataframe
# epf_cal <- epf_cal[!is.na(epf_cal$ID),]
# # Making Date object
# epf_cal$Sample_Date <- as.Date(as.character(epf_cal$sample_date),origin="%Y%m%d",format="%Y%m%d")
# # Removing samples outside of experiment (those for louises exp)
# epf_cal <- epf_cal[epf_cal$Sample_Date < "2017-08-25",]
# epf_cal <- epf_cal[!is.na(epf_cal$Sample_Date),]
# #aggreate final two timepoints
# epf_cal$Sample_Date[epf_cal$Sample_Date > "2017-08-20"] <-"2017-08-23"
#   
# # Convert Dates into timepoints (time prior to experiment exposure starting)
# epf_cal$timepoint <-  epf_cal$Sample_Date - as.Date("2017-06-04")
# epf_cal$timepoint_fac <-  as.factor(epf_cal$timepoint)

### EPF ALL CHEM AND RESPIRATION DATA SHEET ###
epf_cal2 <- read.csv(file="/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/EPF/AE17_Exp_2017_CalcificationInfo_20190321_JBR.csv")
# Making Date object
epf_cal2$Sample_Date <- as.Date(as.character(epf_cal2$sample_date),origin="%Y%m%d",format="%Y%m%d")
# Removing samples outside of experiment (those for louises exp)
epf_c <- epf_cal2[epf_cal2$Sample_Date < "2017-08-25",]
# Remove an NA
epf_c <- epf_c[!is.na(epf_c$Sample_Date),]
# Creating a timepoint
epf_c$timepoint <-  epf_c$Sample_Date - as.Date("2017-06-04")
epf_c$timepoint_fac <-  as.factor(epf_c$timepoint)
## Deciding to treat time and treatment as factors
# Aggregating data pre exposure levels to 400
epf_c$pCO2[epf_c$timepoint < 0] <- 400
#Convert to factor
epf_c$pCO2_fac <-  as.factor(epf_c$pCO2)

#Select useful columns
c_sel <- c("ID","Sample_Date","timepoint","timepoint_fac",
           "WaterSample_ID","shelf","tank","pCO2","pCO2_fac",
           "DIC","TA","pCO2_calc","Calcite","pH_meas",
           "dry_end_meas","Daily_Pcnt_Change_Corrected")
epf_c %>% dplyr::select(c_sel) -> epf_cf
unique(epf_cf$Sample_Date)
epf_cf$Date_Order <- NA
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-05-09")] <- 1
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-05-17")] <- 2
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-06-02")] <- 3
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-06-05")] <- 4
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-06-06")] <- 5
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-06-13")] <- 6
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-06-26")] <- 7
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-07-24")] <- 8
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-08-22")] <- 9
epf_cf$Date_Order[epf_cf$Sample_Date==as.Date("2017-08-24")] <- 10
epf_cf$Date_Tank <- paste0(epf_cf$Date_Order,"_",epf_cf$WaterSample_ID)
### EPF DIC DATA SHEET ###

epf_wDIC <- read.csv(file="/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/EPF/AE17_EPF_DIC_20190707.csv")

### Water Chemistry ###
wc <-  read.csv(file="/home/downeyam/Github/2017OAExp_Oysters/input_files/WC/AE17_pHSalinity_20180305.csv")
# Making Date Column
wc$Sample_Date <- as.Date(as.character(wc$Date_WQ),origin="%Y%m%d",format="%Y%m%d")
# subsetting to only needed columns
wc_sel <- c("Shelf","Tank","TankID","Sample_Date",
            "PCO2","Sal_Corr","pH","Temp")
wc %>% dplyr::select(wc_sel) -> wc_sub
# These dates are for selecting the nearest wc sampling tp either on or proceeding the epf tp
date_sel <- c("2017-05-08","2017-05-17","2017-06-02","2017-06-05","2017-06-06",
              "2017-06-12","2017-06-26","2017-07-24","2017-08-21","2017-08-23")
wc_sub <- wc_sub[which(wc_sub$Sample_Date %in% as.Date(date_sel)),]
# Create order for Dates
wc_sub$Date_Order <- NA
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-05-08")] <- 1
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-05-17")] <- 2
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-06-02")] <- 3
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-06-05")] <- 4
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-06-06")] <- 5
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-06-12")] <- 6
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-06-26")] <- 7
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-07-24")] <- 8
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-08-21")] <- 9
wc_sub$Date_Order[wc_sub$Sample_Date==as.Date("2017-08-23")] <- 10
wc_sub$Date_Tank <- paste0(wc_sub$Date_Order,"_",wc_sub$TankID)
colnames(wc_sub)
wc_sub %>% dplyr::select("Date_Tank",Tank_Salinity="Sal_Corr",Tank_pH="pH",Tank_temp="Temp")-> wc_f


# Check to see the samples are evenly spread across tank and time (some missing values for last tps)
table(epf_cf$WaterSample_ID,epf_cf$Date_Order)
table(wc_sub$TankID,wc_sub$Date_Order)

# Merge the oyster and chem data
pheno_wc <- merge(epf_cf,wc_f,by="Date_Tank")

### Data set with population information ###
sp <- read.csv("/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/AE17_Exp_2017_CollectionInfo_20190227.csv")

# Making Date object
sp$Sample_Date <- as.Date(as.character(sp$sample_date),origin="%Y%m%d",format="%Y%m%d")
# Removing samples outside of experiment (those for louises exp)
sp <- sp[sp$Sample_Date < "2017-08-25",]
# Remove an NA
sp <- sp[!is.na(sp$Sample_Date),]
colnames(sp)
sp_col <- c("ID","PopOrigin","length","width","wet_wgtcorr","dry_wgt")
sp %>% dplyr::select(sp_col) -> sp_red

pheno_wc <- merge(pheno_wc,sp_red,by="ID")
# Dry Weight / Length Condition index
pheno_wc$Len_Wgt_Index <- pheno_wc$dry_wgt / pheno_wc$length
# Hand and Nell Condition Index : Dry Weight * 1000 / (Wet Wgt)
pheno_wc$HN_Index <- (pheno_wc$dry_wgt * 1000)/(pheno_wc$wet_wgtcorr)

### Meta data with sample sequence info ###
seqSamples <- read.csv(file="/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/EPF/AE17_EPF_DIC_20190707.csv")
seqSamples <- seqSamples[!is.na(seqSamples$sample_index),]

#subset full pheno_wc_Seq to just thoses we sequenced
pheno_wc_Seq <- pheno_wc[which(pheno_wc$ID %in% seqSamples$ID),]
#merge with sampleSeq datasheet
sampleSeq_full <- merge(seqSamples,pheno_wc_Seq,by="ID")
colnames(sampleSeq_full)
sampleSeq_full %>%
dplyr::select("ID","sample_name","sample_index","Day",
              "treatment","pCO2_fac",tank="tank.x",shelf="shelf.x",
              "population","PopOrigin",
              "extraction_order","seq_order","lane",
              "epf_pH",epf_DIC="DIC.x",epf_Salinity="Salinity",
              env_pH="pH_meas",env_DIC="DIC.y",env_TA="TA",env_pCO2="pCO2_calc",env_Calcite="Calcite",
              Calcification_perDay = "Daily_Pcnt_Change_Corrected",
              "length","width",wet_wgt="wet_wgtcorr","dry_wgt",
              "Len_Wgt_Index","HN_Index") -> sample_Seq_Final

# Store full phenotype and water chemistry dataframe as RData for easier use
saveRDS(pheno_wc,"/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/EPF/full_data_20190716.Rdata")
# Store phenotype and wc data from just the 24 sequenced samples for easier use
saveRDS(sample_Seq_Final,"/home/downeyam/Github/2017OAExp_Oysters/input_files/Phenotype/EPF/seq_data_20190716.Rdata")
```

## Additional Analysis  
  
### Examining water chem in multi dimensions
```{r}
par(mfrow=c(1,1))
epf_cCarbChem <- epf_cal[!is.na(epf_cal$EPF_DIC_Start),]

chemNames <- c("EPF_pH","EPF_DIC_Start","EPF_Ca_Start")
epf_cCarbChem_r <- subset(epf_cCarbChem,select = chemNames )

cCarb_pca <- prcomp(epf_cCarbChem_r,scale=TRUE)
fviz_eig(cCarb_pca)

fviz_pca_var(cCarb_pca,
             title="Variable Vectors - Total Carbonate Chem Data ",
             col.var = "contrib", # Color by contributions to the PC
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )

group1 <- as.factor(epf_cCarbChem$pCO2)
fviz_pca_biplot(cCarb_pca, repel = TRUE,
                title="Individuals and Variable Vectors - Total Carbonate Chem Data (Ellipse = CI) ",
                col.ind = group1, # color by groups
                palette = c("#00AFBB",  "#FC4E07","darkseagreen"),
                addEllipses = TRUE,
                ellipse.type = "confidence",
                legend.title = "pCO2",
                col.var = "#2E9FDF" # Variables color
                #col.ind = "#696969"  # Individuals color
                )

group2 <- interaction(epf_cCarbChem$pCO2,epf_cCarbChem$timepoint)
fviz_pca_biplot(cCarb_pca, repel = TRUE,label = c("var"),
                col.ind = group2, # color by groups
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
                addEllipses = TRUE,
                ellipse.type = "confidence",
                legend.title = "pCO2.SampleDay",
                col.var = "#2E9FDF" # Variables color
                #col.ind = "#696969"  # Individuals color
                )

### Just Samples from the actual exposure (acclimation points removed)
exposure_cCarbChem <- epf_cCarbChem[epf_cCarbChem$timepoint > 0,]
chemNames <- c("EPF_pH","EPF_DIC_Start","EPF_Ca_Start")
exposure_cCarbChem_r <- subset(exposure_cCarbChem,select = chemNames )

cCarb_pca2 <- prcomp(exposure_cCarbChem_r,scale=TRUE)
fviz_eig(cCarb_pca2)

group3 <- interaction(exposure_cCarbChem$pCO2,exposure_cCarbChem$timepoint)
fviz_pca_biplot(cCarb_pca2, repel = TRUE,label = c("var"),
                title="Individuals and Variable Vectors - Exposure Only Carbonate Chem Data (Ellipse = CI) ",
                col.ind = group3, # color by groups
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
                addEllipses = TRUE,
                ellipse.type = "confidence",
                legend.title = "pCO2",
                col.var = "#2E9FDF" # Variables color
                #col.ind = "#696969"  # Individuals color
                )

### Just Samples we sequenced 
seq_cCarbChem <- just_seq[!is.na(just_seq$EPF_DIC_Start),]
chemNames <- c("EPF_pH","EPF_DIC_Start","EPF_Ca_Start")
seq_cCarbChem_r <- subset(seq_cCarbChem,select = chemNames )

cCarb_pca3 <- prcomp(seq_cCarbChem_r,scale=TRUE)
fviz_eig(cCarb_pca3)

group3 <- interaction(seq_cCarbChem$pCO2,seq_cCarbChem$timepoint)
fviz_pca_biplot(cCarb_pca3, repel = TRUE,label = c("var"),
                title="Individuals and Variable Vectors - Sequenced Samples Only Carbonate Chem Data (Ellipse = CI) ",
                col.ind = group3, # color by groups
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
                addEllipses = TRUE,
                ellipse.type = "confidence",
                legend.title = "pCO2",
                col.var = "#2E9FDF" # Variables color
                #col.ind = "#696969"  # Individuals color
                )
```
