---
title: "Cvirginica_palialpH_RNAseq_countmatrix"
output: 
  github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

###Post Processing- Creating Count Matrix
First, sort all bam files by name to use in htseq-count. Sorting by name seems to work the best when working with bam files output from STAR alignments. You'll need to create a tmp file to hold temporary files required for samtools
```{bash, eval=FALSE}
mkdir tmp
```
Then sort files:
```{bash, eval=FALSE}
#!/bin/bash

for i in $( ls RNA*.q4.bam ); do
        ind=$( echo $i | cut -d'.' -f 1)
        samtools sort -n -@32 -T /tmp/"$i".q4.namesorted \
        -o "$ind".q4.namesorted.bam $i 2> "$ind"_bam_namesorted_error.log
done
```
I then converted the gff file to a gtf file using gffread (see here: https://github.com/gpertea/gffread for installation instructions), because it was easier to use gtf file for htseq, the program that gives a count matrix. Make sure that gffread executable is in the current directory (`./RNA` subdirectory).
```{bash, eval=FALSE}
./gffread ref_C_virginica-3.0_top_level.gff3 -T -o top.gtf
```
At this point there may be missing gene_name data in the `top.gtf` file. Because the gene_name is the id attribute used to count transcripts in the code below, it is essential that each element has a gene_name. Therefore, I manually added a gene_name for elements without one starting "unknown_gene1" all the way up to "unknown_gene26"

```{bash, eval=FALSE}
subl top.gtf
```

Now use the name sorted bam files and the gtf file in htseq-count to count the average number of transcripts (exons specifically) per gene. 
```{bash,eval=FALSE}
#!/bin/bash

for i in $( ls RNA*.q4.namesorted.bam ); do
        ind=$( echo $i | cut -d'.' -f 1| cut -c 4-)
        htseq-count -f bam -r name -s yes -i gene_name \
        --additional-attr gene_id -t exon -m union \
        $i top.gtf > \
        "$ind"_exon_count.txt 2> "$ind"_exon_count_error.log
done
```
  
**Options for `htseq-count`**  
* `-f` : format; bam or sam  
* `-r` : order; 'pos' or 'name' Sorting order of alignment file. Paired-end sequencing data must be sorted either by position or by read name, and the sorting order must be specified.  
* `-s` : stranded; {yes,no,reverse}, whether from strand specific assay.  
* `-i` : IDATTR; GFF attribute to be used as feature ID  
* `-t` : Featuretype; (third column of GFF) all other features ignored (default for GTF files: exon)  
* `-m` : mode; {union,intersection-strict,intersection-nonempty}, mode to handle reads overlapping more than one feature.  
* `--additional-attr` : additional-attr; additional feature attributes { default: none, suitable, for GTF files: gene_name}
  
This creates files for each individual. I then created an Rscript to combine all of these files into a single count matrix with each individual as a separate column and each gene as a separate row.
```{r, eval=FALSE}
base_path <- ("/home/downeyam/Github/2017OAExp_Oysters/RNA")
exon_count_filenames <- list.files(path=base_path, pattern=".+_exon_count.txt")

k <- 1
column_names <- c("gene_name", "gene_id")

for (i in exon_count_filenames) {
  file_contents <- unlist(strsplit(i, split="[_]"))
  ind <- file_contents[1]
  table_name <- paste0(ind, "_exon_count")
  assign(table_name, read.table(i, header=F, sep = "\t"))
  if(k==1) {
    C_virginica_gene_count <- data.frame(get(table_name))
  }
  else {
    C_virginica_gene_count <- cbind(C_virginica_gene_count, get(table_name)$V3)
  }
  k=k+1
  column_names <- c(column_names, paste0("RNA",ind))
}

names(C_virginica_gene_count) <- column_names
write.table(C_virginica_gene_count, file="C_virginica_gene_count.txt", col.names = T)
```


