---
title: "Multivariate Visualization of Transcriptomic and DNA Methylation Data"
author: "adowneywall"
date: "6/11/2019"
output: 
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(DESeq2)
library(edgeR)
library(ashr)
library(vegan)
library(sf)
library(adegenet)
library(ggplot2)
library(ape)
library(matrixStats)
```

## **Script Description**  

**Brief Overview** : This script performs a series of basic multivariate approaches to explore the whole genome expression profiles of the 24 oyster RNAseq samples. In partiular this includes:
* A permanova (implemented in the package vegan, called adonis)  
* An RDA (implemented in vegan)
* A DAPC with a focus on treatment(implemented in adegenet)

## **Data**  

Steps:  
* Read in dataframe with metadata for each samples including, treatment, time, population, lane of sequencing, and variable that contains each unique combination level between treatment and time.  
* Read in gene (or transcript) abundance matrix generated from the '01_salmon_countMatrix.Rmd' script.  
* Remove uninformative genes from count matrix.  

### Current files
```{r}
## Meta Data for the Model
model<-readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/meta/metadata_20190811.RData")

#### Transcript File ####
tran <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/references/STAR_gnomon_tximportGeneFile.RData")
# Gene File
gene <- tran[!duplicated(tran$GENEID),]
rm(tran)
# This is the filtered count matrix
RSEM <-  readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/STAR_mapping/RSEM_output/GeneCountMatrix_Filtered_DGEListObj.RData")
# Separate out RSEM counts and rename rows with LOC ID
rsem_c <- RSEM$counts
#counts_Rdat <- readRDS(file = "/home/downeyam/Github/2017OAExp_Oysters/input_files/RNA/Normalized_CountMatrix/Scenario1_normalizedVoom.RData")
rna_count <- cpm(rsem_c)

#DNA Methlyation Data
dnaM_beta <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/DNAm/Final_beta_gene_10.RData")
dnaM_beta[which(is.na(dnaM_beta[,1])),] <- 0
dnaM_beta_no99 <- dnaM_beta[,-11]
colnames(dnaM_beta_no99)
model_no99 <- model[-11,]
model_no99$sample_name
```

# Based on Gene (and isoform) level data

### PERMANOVA (implements using adonis from vegan package)

* Testing for statistic differential between treatment, time, and treatment:time  

**RNA**
```{r}
## RNA
#Scales each gene to have mean of 0 and sd of 1
rna_scale <- t(scale(t(rna_count)))
# Adjusts distribution so that min value is <= 0 (gets rid of negative numbers)
rna_scale2 <- rna_scale + abs(rowMins(rna_scale))
# Quick check to see that scaling seems to be performed correctly (i.e. sd = 1 for each gene)
sd(rna_scale2[5000,])

# Without Scaling
(out_gc <- adonis(t(rna_count)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
# With Scaling but w/o adjusting for negative values
(out_gc <- adonis(t(rna_scale)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
# Scaling and adjusting for negative values (final model)
(out_gc <- adonis(t(rna_scale2)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
(out_gc <- adonis(t(rna_scale2)~Treatment*Time,data=model,permutations = 5000))
```

**DNAm**
```{r}
# All samples
(out_dnam <- adonis(t(dnaM_beta)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
# With 17099 removed
(out_dnam_no99 <- adonis(t(dnaM_beta_no99)~Treatment*Time+Pop+Lane,data=model_no99,permutations = 5000))
# There is a significant effect due to treatment
```

### Plotting data with PCA  

#### RNA (treatment and time)
```{r}
#prin_comp<-rda(t(rna_scale2))
#sum_pri <- summary(prin_comp)
#pca_scores<-scores(prin_comp)
pca <- prcomp(t(rna_scale2))
eigs <- pca$sdev^2

pca$x[,1]
prin_comp$CA$v[,1]
head(prin_comp$CA)
plot(pca$x[,1]~prin_comp$CA$u[,1])

color_comb <- c("lightblue4","royalblue3","tomato","darkred") # colors for population 
model$colors <- "" 
model$colors[model$SFV == unique(model$SFV)[1]] <-  color_comb[2]
model$colors[model$SFV == unique(model$SFV)[2]] <-  color_comb[1]
model$colors[model$SFV == unique(model$SFV)[3]] <-  color_comb[4]
model$colors[model$SFV == unique(model$SFV)[4]] <-  color_comb[3]
model$pch <- 16
model$pch[model$colors == color_comb[2] | model$colors == "darkred"] <- 10

ordiplot(pca,type="n",
         xlab=paste0("PC1  (",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
         ylab=paste0("PC2  (",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))

orditorp(pca,display="sites",labels = FALSE,col=model$colors,cex = 2,pch = model$pch)
ordispider(pca,model$SFV,col = color_comb,lwd=2.5)
legend(x=c(50,70),y=c(80,120),
       legend = c("Day 09 : Ambient",
                  "Day 80 : Ambient",
                  "Day 09 : OA 2800",
                  "Day 80 : OA 2800"),
       pch = c(16,10,16,10),col=color_comb,xpd = .25,bty="n")
text(x = 70 ,y = -25, 
     paste0("P_Treatment = ",round(out_gc$aov.tab$`Pr(>F)`[1],3),"*"),pos = 4)
text(x = 70 ,y = -35, 
     paste0("P_Time = ",round(out_gc$aov.tab$`Pr(>F)`[2],5),"*"),pos=4)
text(x = 70 ,y = -45, 
     paste0("P_Interaction = ",round(out_gc$aov.tab$`Pr(>F)`[5],3)),pos=4)
```

#### RNA : population  
```{r}
ordiplot(prin_comp,type="n",
         xlab=paste0("PC1 ",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
         ylab=paste0("PC2 ",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))
orditorp(prin_comp,display="sites",labels = FALSE,col=model$Pop,cex = 2,pch = 16)
legend(x=100,y=100,legend = c("Ipswich","Rowley 1","Rowley 2"),pch = 16,col=unique(model$Pop),xpd = .25)
```
  
#### DNAm (treatment and time) : All samples
```{r}
pca <- prcomp(t(dnaM_beta))
eigs <- pca$sdev^2
pca$x
pca$x[,1]
prin_comp$CA$u[,1]
plot(pca$x[,1]~prin_comp$CA$u[,1])
# Sample 17099 Looks like a weird outlier along the first PC axis

color_comb <- c("lightblue4","royalblue3","tomato","darkred") # colors for population 
model$colors <- "" 
model$colors[model$SFV == unique(model$SFV)[1]] <-  color_comb[2]
model$colors[model$SFV == unique(model$SFV)[2]] <-  color_comb[1]
model$colors[model$SFV == unique(model$SFV)[3]] <-  color_comb[4]
model$colors[model$SFV == unique(model$SFV)[4]] <-  color_comb[3]
model$pch <- 16
model$pch[model$colors == color_comb[2] | model$colors == "darkred"] <- 10

ordiplot(pca,type="n",
         xlab=paste0("PC1  (",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
         ylab=paste0("PC2  (",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))

orditorp(pca,display="sites",labels = FALSE,col=model$colors,cex = 2,pch = model$pch)
ordispider(pca,model$SFV,col = color_comb,lwd=2.5)
legend(x=c(30,40),y=c(-2,2),
       legend = c("Day 09 : Ambient",
                  "Day 80 : Ambient",
                  "Day 09 : OA 2800",
                  "Day 80 : OA 2800"),
       pch = c(16,10,16,10),col=color_comb,xpd = .25,bty="n")
text(x = 30 ,y = -20, 
     paste0("P_Treatment = ",round(out_dnam$aov.tab$`Pr(>F)`[1],3)),pos = 4)
text(x = 30 ,y = -25, 
     paste0("P_Time = ",round(out_dnam$aov.tab$`Pr(>F)`[2],5)),pos=4)
text(x = 30 ,y = -30, 
     paste0("P_Interaction = ",round(out_dnam$aov.tab$`Pr(>F)`[5],3)),pos=4)
```  
  
#### PCA DNAm time and treatment : with 17099 removed
```{r}
pca <- prcomp(t(dnaM_beta_no99))
eigs <- pca$sdev^2

color_comb <- c("lightblue4","royalblue3","tomato","darkred") # colors for population 
model_no99$colors <- "" 
model_no99$colors[model_no99$SFV == unique(model_no99$SFV)[1]] <-  color_comb[2]
model_no99$colors[model_no99$SFV == unique(model_no99$SFV)[2]] <-  color_comb[1]
model_no99$colors[model_no99$SFV == unique(model_no99$SFV)[3]] <-  color_comb[4]
model_no99$colors[model_no99$SFV == unique(model_no99$SFV)[4]] <-  color_comb[3]
model_no99$pch <- 16
model_no99$pch[model_no99$colors == color_comb[2] | model_no99$colors == "darkred"] <- 10

ordiplot(pca,type="n",
         xlab=paste0("PC1  (",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
         ylab=paste0("PC2  (",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))

orditorp(pca,display="sites",labels = FALSE,col=model_no99$colors,cex = 2,pch = model_no99$pch)
ordispider(pca,model_no99$SFV,col = color_comb,lwd=2.5)
legend(x=c(15,40),y=c(10,25),
       legend = c("Day 09 : Ambient",
                  "Day 80 : Ambient",
                  "Day 09 : 2800",
                  "Day 80 : 2800"),
       pch = c(16,10,16,10),col=color_comb,xpd = .25,bty="n")
text(x = 8 ,y = -20, 
     paste0("P_Treatment = ",round(out_dnam_no99$aov.tab$`Pr(>F)`[1],3),"*"),pos = 4)
text(x = 8 ,y = -23, 
     paste0("P_Time = ",round(out_dnam_no99$aov.tab$`Pr(>F)`[2],5)),pos=4)
text(x = 8 ,y = -26, 
     paste0("P_Interaction = ",round(out_dnam_no99$aov.tab$`Pr(>F)`[5],3)),pos=4)
```  
  
#### DNAm : population  
```{r}
ordiplot(prin_comp,type="n",
         xlab=paste0("PC1 (",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
         ylab=paste0("PC2 (",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))
orditorp(prin_comp,display="sites",labels = FALSE,col=model$Pop,cex = 2,pch = 16)
legend(x=100,y=100,legend = c("Ipswich","Rowley 1","Rowley 2"),pch = 16,col=unique(model$Pop),xpd = .25)

prin_com
```
  
### Cumulative Variance Plot  
```{r}
out <- prcomp(gc_reduce)
vars <- apply(out$x, 2, var)  
props <- vars / sum(vars)
cumsum(props)
plot(cumsum(props)*100~c(1:24),ylab=c("Cumulative Variance (%)"),xlab=c("PCs"))
```

### TWO STEP DAPC: first create discriminant function from TP 9 samples and predict coordinates on df for day 80 samples.  

#### RNA 
**Creating DF by treatment with first timepoint**  
```{r}
early_time_counts <- rna_scale2[,model$Day == 9]
early_time_meta <- model[model$Day == 9,]

dapc_treatment_10<-dapc(t(early_time_counts),early_time_meta$treatment,n.pca=8,n.da=2)
# PCs = 8
# clusters = 1
early_time_meta$dt <- unlist(dapc_treatment_10$ind.coord[,1])

ggplot(early_time_meta,aes(dt,fill=as.factor(treatment),colour=as.factor(treatment))) + 
  geom_density(alpha=0.1) + xlim(-15,15) + 
  labs(title="Discriminant Function for Treatment on Day 9 (based on 8 PCs)",
       x="Discriminant function 1",
       colour="Treatment",
       fill="Treatment") +
  theme_bw() +
  scale_color_manual(values=c("deepskyblue2","firebrick1")) +
  scale_fill_manual(values=c("deepskyblue2","firebrick1"))
```

Looking at which genes are driving the patterns between the two treatments
```{r}
#contrib_treatment <- loadingplot(dapc_treatment_10$var.contr, axis=1,thres=.05, lab.jitter=1)
```

Looking at the most important locus based on loading
```{r}
# head(gc)
# gc_majorLoading <- gc[row.names(gc) == "LOC111102518",]
# majorLoading_counts <- as.data.frame(cbind(Trt=model$Treatment,Time=model$Time,SFV=model$SFV,Pop=model$Pop,(LOC111104151=gc_majorLoading)))
# 
# ggplot(majorLoading_counts,aes(x=as.factor(SFV),y=log10(LOC111104151))) + geom_boxplot() + 
#   labs(x= c("SFV"))
```

**Mapping 
Day 80 samples**  
```{r}
late_time_counts <- rna_scale2[,model$Day == 80]
late_time_meta <- model[model$Day == 80,]

predict_values <- predict.dapc(dapc_treatment_10,t(late_time_counts))
late_time_meta$dt <-unlist(predict_values$ind.scores[,1])

whole_meta<- rbind(early_time_meta,late_time_meta)

ggplot(whole_meta,aes(dt,fill=as.factor(interaction(Day,treatment)),colour=as.factor(interaction(Day,treatment)))) + 
  geom_density(alpha=0.1) + xlim(-12,12) + 
  labs(x="Discriminant function 1",
       colour="Day.Treatment",
       fill="Day.Treatment") +
  theme_bw() +
  scale_color_manual(values=c("lightblue4","blue4","tomato","darkred")) +
  scale_fill_manual(values=c("lightblue4","blue4","tomato1","darkred"))

D9_400 <- whole_meta[whole_meta$SFV == "09.400",]
D9_400_Density <- density(D9_400$dt)
D9_2800 <- whole_meta[whole_meta$SFV == "09.2800",]
D9_2800_Density <- density(D9_2800$dt)
D80_400 <- whole_meta[whole_meta$SFV == "80.400",]
D80_400_Density <- density(D80_400$dt)
D80_2800 <- whole_meta[whole_meta$SFV == "80.2800",]
D80_2800_Density <- density(D80_2800$dt)

plot(D9_400_Density, main="",xlim=c(-10,12),ylim=c(0,0.9),
     xlab="Discriminant Function 1")
polygon(D9_400_Density, col=alpha("lightblue",0.8), border="lightblue")
polygon(D9_2800_Density, col=alpha("tomato",0.8), border="red3")
polygon(D80_400_Density, col=alpha("blue",0.9), 
        border="blue",density = 20,cex=100)
polygon(D80_2800_Density, col=alpha("red3",0.9),
        border="red3",density = 20,angle=0,cex=5)
legend(x=c(0,10),y=c(0.9,0.73),legend=c("Day 09 : Ambient","Day 09 : 2800 ","Day 80 : Ambient","Day 80 : 2800"),
       fill = c(alpha("lightblue",0.8),alpha("tomato",0.8),alpha("blue",0.7),alpha("red3",0.7)),
       density = c(1000,1000,50,50),angle = c(0,0,45,0),bty="n")
```

### DAPC based on combined time*treatment factor  
  
**Creating DF by treatment with first timepoint**  
```{r}
dapc_SFV_10<-dapc(t(rna_scale2),model$SFV,n.pca=8,n.da=3)
# PCs = 10
# clusters = 3
output <- data.frame(Trt=model$Treatment,Time=model$Time,dapc_SFV_10$ind.coord)

ggplot(output,aes(x=LD1,y=LD2,fill=as.factor(interaction(Trt,Time)),colour=as.factor(interaction(Trt,Time)))) + 
  geom_point(aes(size=5)) + #geom_density(alpha=0.1) + #xlim(-28,28) + 
  labs(title="DAPC for Treatment*Time Combination",
       x="Discriminant function 1",
       y="Discriminant function 2",
       colour="Treatment",
       fill="Treatment") +
    theme_bw() +
  scale_color_manual(values=c("deepskyblue2","firebrick1","blue4","darkred")) +
  scale_fill_manual(values=c("deepskyblue2","firebrick1","blue4","darkred"))
```

Looking at contribution of individual genes
```{r}
contrib_SFV <- loadingplot(dapc_SFV_10$var.contr, axis=2,thres=.07, lab.jitter=1)
```

### DNAm

```{r}
dnaM_beta_no99_v2 <- as.matrix(dnaM_beta_no99)
early_time_counts <- dnaM_beta_no99_v2[,model_no99$Day == 9]
early_time_meta <- model_no99[model_no99$Day == 9,]

dapc_treatment_10<-dapc(t(early_time_counts),early_time_meta$treatment,n.pca=8,n.da=2)
# PCs = 8
# clusters = 1
early_time_meta$dt <- unlist(dapc_treatment_10$ind.coord[,1])

ggplot(early_time_meta,aes(dt,fill=as.factor(treatment),colour=as.factor(treatment))) + 
  geom_density(alpha=0.1) + xlim(-6,6) + 
  labs(title="Discriminant Function for Treatment on Day 9 (based on 8 PCs)",
       x="Discriminant function 1",
       colour="Treatment",
       fill="Treatment") +
  theme_bw() +
  scale_color_manual(values=c("deepskyblue2","firebrick1")) +
  scale_fill_manual(values=c("deepskyblue2","firebrick1"))
```

**Mapping 
Day 80 samples**  
```{r}
late_time_counts <- dnaM_beta_no99_v2[,model_no99$Day == 80]
late_time_meta <- model_no99[model_no99$Day == 80,]

predict_values <- predict.dapc(dapc_treatment_10,t(late_time_counts))
late_time_meta$dt <-unlist(predict_values$ind.scores[,1])

whole_meta<- rbind(early_time_meta,late_time_meta)

ggplot(whole_meta,aes(dt,fill=as.factor(interaction(Day,treatment)),colour=as.factor(interaction(Day,treatment)))) + 
  geom_density(alpha=0.1) + xlim(-6,6) + 
  labs(x="Discriminant function 1",
       colour="Day.Treatment",
       fill="Day.Treatment") +
  theme_bw() +
  scale_color_manual(values=c("lightblue4","blue4","tomato","darkred")) +
  scale_fill_manual(values=c("lightblue4","blue4","tomato1","darkred"))

D9_400 <- whole_meta[whole_meta$SFV == "09.400",]
D9_400_Density <- density(D9_400$dt)
D9_2800 <- whole_meta[whole_meta$SFV == "09.2800",]
D9_2800_Density <- density(D9_2800$dt)
D80_400 <- whole_meta[whole_meta$SFV == "80.400",]
D80_400_Density <- density(D80_400$dt)
D80_2800 <- whole_meta[whole_meta$SFV == "80.2800",]
D80_2800_Density <- density(D80_2800$dt)

plot(D9_400_Density, main="",xlim=c(-5,3),ylim=c(0,0.9),
     xlab="Discriminant Function 1")
polygon(D9_400_Density, col=alpha("lightblue",0.8), border="lightblue")
polygon(D9_2800_Density, col=alpha("tomato",0.8), border="red3")
polygon(D80_400_Density, col=alpha("blue",0.9), 
        border="blue",density = 20,cex=100)
polygon(D80_2800_Density, col=alpha("red3",0.9),
        border="red3",density = 20,angle=0,cex=5)
legend(x=c(0,10),y=c(0.9,0.73),legend=c("Day 09 : Ambient","Day 09 : 2800 ","Day 80 : Ambient","Day 80 : 2800"),
       fill = c(alpha("lightblue",0.8),alpha("tomato",0.8),alpha("blue",0.7),alpha("red3",0.7)),
       density = c(1000,1000,50,50),angle = c(0,0,45,0),bty="n")

dapc_SFV_10<-dapc(t(dnaM_beta_no99),model_no99$SFV,n.pca=8,n.da=3)
# PCs = 10
# clusters = 3
output <- data.frame(Trt=model_no99$Treatment,Time=model_no99$Time,dapc_SFV_10$ind.coord,Pop=model_no99$Pop)

ggplot(output,aes(x=LD1,y=LD2,fill=as.factor(interaction(Trt,Time)),colour=as.factor(interaction(Trt,Time)))) + 
  geom_point(aes(size=5)) + #geom_density(alpha=0.1) + #xlim(-28,28) + 
  labs(title="DAPC for Treatment*Time Combination",
       x="Discriminant function 1",
       y="Discriminant function 2",
       colour="Treatment",
       fill="Treatment") +
    theme_bw() +
  scale_color_manual(values=c("deepskyblue2","firebrick1","blue4","darkred")) +
  scale_fill_manual(values=c("deepskyblue2","firebrick1","blue4","darkred"))

ggplot(output,aes(x=LD1,y=LD2,fill=as.factor(interaction(Trt,Time)),colour=as.factor(Pop))) + 
  geom_point(aes(size=5)) + #geom_density(alpha=0.1) + #xlim(-28,28) + 
  labs(title="DAPC for Treatment*Time Combination",
       x="Discriminant function 1",
       y="Discriminant function 2",
       colour="Treatment",
       fill="Treatment") +
    theme_bw() +
  scale_color_manual(values=c("deepskyblue2","firebrick1","blue4","darkred")) +
  scale_fill_manual(values=c("deepskyblue2","firebrick1","blue4","darkred"))
```

## Final figures
```{r eval=FALSE}
par(mfrow=c(2,1))
# Figure A
text(x = -25 ,y = 20, 
     "A",cex = 2.2,xpd=NA)

# Figure B

late_time_counts <- gc_reduce[,model$Day == 80]
late_time_meta <- model[model$Day == 80,]

predict_values <- predict.dapc(dapc_treatment_10,t(late_time_counts))
late_time_meta$dt <-unlist(predict_values$ind.scores[,1])

whole_meta<- rbind(early_time_meta,late_time_meta)

D9_400 <- whole_meta[whole_meta$SFV == "09.400",]
D9_400_Density <- density(D9_400$dt)
D9_2800 <- whole_meta[whole_meta$SFV == "09.2800",]
D9_2800_Density <- density(D9_2800$dt)
D80_400 <- whole_meta[whole_meta$SFV == "80.400",]
D80_400_Density <- density(D80_400$dt)
D80_2800 <- whole_meta[whole_meta$SFV == "80.2800",]
D80_2800_Density <- density(D80_2800$dt)

plot(D9_400_Density, main="",xlim=c(-10,7),ylim=c(0,0.9),
     xlab="Discriminant Function 1",cex=1.2)
polygon(D9_400_Density, col=alpha("lightblue",0.8), border="lightblue")
polygon(D9_2800_Density, col=alpha("tomato",0.8), border="red3")
polygon(D80_400_Density, col=alpha("royalblue3",0.9), 
        border="royalblue3",density = 20,cex=100)
polygon(D80_2800_Density, col=alpha("red3",0.9),
        border="red3",density = 20,angle=0,cex=5)
legend(x=c(-10,-4.9),y=c(0.9,0.65),bty="n",cex=1.2,
       legend=c("Day 09 : Ambient",
                "Day 80 : Ambient",
                "Day 09 : 2800 ",
                "Day 80 : 2800"),
       fill = c(alpha("lightblue",0.8),
                alpha("royalblue3",0.7),
                alpha("tomato",0.8),
                alpha("red3",0.7)),
       density = c(1000,
                   50,
                   1000,
                   50),
       angle = c(0,
                 45,
                 0,
                 0))
text(x = -11.88 ,y = 1.03, 
     "B",cex = 2.2,xpd=NA)
```

# Based on Transcript level data

### PERMANOVA (implements using adonis from vegan package)

* Testing for statistic differential between treatment, time, and treatment:time  

```{r}
# (out <- adonis(t(tc_reduce)~Treatment*Time+Pop+Lane,data=model,permutations = 5000))
```

### Plotting data with RDA with isoform data

### Plot in multivariate space with RDA (treatment and time)
```{r}
# prin_comp<-rda(t(tc_reduce), scale=FALSE)
# sum_pri <- summary(prin_comp)
# pca_scores<-scores(prin_comp)
# 
# pca <- prcomp(t(tc_reduce))
# eigs <- pca$sdev^2
# 
# head(sum_pri$species)
# color_comb <- c("deepskyblue2","blue4","firebrick1","darkred") # colors for population 
# model$colors <- "" 
# model$colors[model$SFV == unique(model$SFV)[1]] <-  color_comb[2]
# model$colors[model$SFV == unique(model$SFV)[2]] <-  color_comb[1]
# model$colors[model$SFV == unique(model$SFV)[3]] <-  color_comb[4]
# model$colors[model$SFV == unique(model$SFV)[4]] <-  color_comb[3]
# 
# ordiplot(prin_comp,type="n",
#          xlab=paste0("PC1 ",round(eigs[1] / sum(eigs)*100,1),"% Variance Explained)"),
#          ylab=paste0("PC2 ",round(eigs[2] / sum(eigs)*100,1),"% Variance Explained)"))
# orditorp(prin_comp,display="sites",labels = FALSE,col=model$colors,cex = 2,pch = 16,)
# #ordiellipse(prin_comp,model$SFV,conf=0.90,col = color_comb,lwd = 3)
# ordispider(prin_comp,model$SFV,col = color_comb,lwd=2.5)
# legend(x=100,y=150,legend = c("Day09_Control","Day80_Control","Day09_OA","Day80_OA"),pch = 16,col=color_comb,xpd = .25)
# text(x = -538 ,y = -300, paste0("Adonis P_Treatment = ",out$aov.tab$`Pr(>F)`[1],"*"))
# text(x = -550 ,y = -360, paste0("Adonis P_Duration = ",out$aov.tab$`Pr(>F)`[2],"*"))
# text(x = -540 ,y = -420, paste0("Adonis P_Interaction = ",out$aov.tab$`Pr(>F)`[5]))
```


