---
title: "Trimming and Removing Adapters"
output: 
  github_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Change to subdirectory with the raw sequence information and download the adapter file corresponding to your library prep kit.
```{bash, eval=FALSE}
cd /shared_lab/20180226_RNAseq_2017OAExp/RNA
# This downloads the adapter info for the type of sequence (TruSeq3 - paired end) method we used
wget https://github.com/timflutre/trimmomatic/blob/master/adapters/TruSeq3-PE-2.fa
```
  
Download custom dDocent v2.2.ed20 that looks for adapters in working directory  
```{bash, eval=FALSE}
wget https://raw.githubusercontent.com/jpuritz/EecSeq/master/Bioinformatics/dDocent
```
  
Download the configuration file  
```{bash, eval=FALSE}
wget https://raw.githubusercontent.com/jpuritz/EecSeq/master/Bioinformatics/RNA.config
```

**Use subl (or nano) to edit config file to appropriate settings**:  
Number of Processors=45,
Maximum Memory=100G,  
Trimming=yes,  
Assembly?=no,  
Type_of_Assembly= ,  
Clustering_Similarity%= ,  
Mapping Reads?=no,  
Mapping_Match_Value= ,  
Mapping_Mismatch_Value= ,  
Mapping_GapOpen_Penalty= ,  
Calling_SNPs?=no,  
Email=downey-wall.a@husky.neu.edu  

```{bash, eval=FALSE}
# subl is for sublime text, switch this to favorite text editor
subl RNA.config
```
  
Use subl (or your favorite text editor) to edit dDocent to trim accoding to custom adapter file (use ctrl+w to find and replace TruSeq2-PE.fa with TruSeq3-PE-2.fa; 2 instances)
```{bash, eval=FALSE}
subl dDocent
```
  
Make sure that the input files are in the format specified by dDocent. That is, filenames should have the extensions \*.R1.fq.gz for forward reads and \*.R2.fq.fz for reverse reads (see http://ddocent.com/quick/ for more details). 
  
Run dDocent (You may have to change permissions to run it):
```{bash, eval=FALSE}
./dDocent RNA.config
```
  
#### What's going on behind the code?  
The dDocent pipeline uses Trimmomatic trimming tool to remove adapter and low quality sequences from the ends of reads. Within the dDocent code it:  
- (i) is specified to be paired-end (which is automatically recognized based on our file naming scheme)  
- (ii) removes adapters based on thresholds for how well the adapter sequences align to reads (2:30:10; see Trimmomatic manual for more details)  
- (iii) removes leading bases with phred quality score less than 20, removes trailing bases with phred quality score   less than 20  
- (iv) scans the reads at a 5bp window and cuts when the average quality of the five bases is less than 10  
- (v) makes sure all reads are a minimum length after this cutting (greater than the shortest read/2).  
  
### Script Outputs  

Example: `P1_17178.F.fq.gz`

Trimmed File Directory: `/shared_lab/20180226_RNAseq_2017OAExp/RNA/rawfiles`

### Filter Reads that Poorly Mapped to Genome
To filter reads that poorly mapped, you can either run the following code each individual:
```{bash, eval=F}
samtools view -@64 -q4 -h -F 0x100 -F 0x400 RNA17005_m3Aligned.sortedByCoord.out.bam| mawk '$6 !~/[8-9].[SH]/ && $6 !~ /[1-9][0-9].[SH]/' | samtools view -b > RNA17005.q4.bam 2> RNA17005.q4_error.log
```
  
**The arguments:**  
- ```-q4```   - Skip alignments with MAPQ smaller than INT [0]. 
- ```-h```    - argument specifies to include header in output
- ```-F```    - Do not output alignments with any bits set in INT present in the FLAG field. INT can be specified in hex by beginning with `0x' (i.e. /^0x[0-9A-F]+/) or in octal by beginning with `0' (i.e. /^0[0-7]+/) [0]. Specifies to not output alignments with with any flag between 100 and 400  
- ```mawk```  - line filters out reads that had significant hard or soft clipping >79 bp.  
- ```b```     - Output in the BAM format. 

Script for filtering reads from all files within a directory:
```{bash, eval=F}
#!/bin/bash

for i in $( ls RNA*m3Aligned.sortedByCoord.out.bam ); do
        ind=$( echo $i | cut -d'_' -f 1)
        samtools view -@64 -q4 -h -F 0x100 -F 0x400 "$i"| \
        mawk '$6 !~/[8-9].[SH]/ && $6 !~ /[1-9][0-9].[SH]/' | \
        samtools view -b > "$ind".q4.bam 2> "$ind".q4_error.log
done
```





**Next** 03_CV17_RNA_Alignment_ReadMapping.Rmd