---
title: "Sequence Cost Script"
author: "adowneywall"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(kableExtra)
library(dplyr)
```


# Initial Values

### Money
```{r}
# Budget 25k
BUD<-25000
```
### Sequencing Coverage
  
#### Whole Genome (C. virginica)
```{r}
# Whole genome (Mb)
GENOME_SIZE <- 685
```
#### Exome (C. virginica)
```{r}
# Size of the whole exome (Mb)
EXOME_SIZE <- 100
```
  
### Regeants and Sequencing Costs
```{r}
## MEC-Seq (not including sequencing costs)
MEC_PSAMPLE <- 25
# based on Puritz et al. Eec-Seq (2018) costs

## WGBS (not including sequencing costs)
WGBS_PSAMPLE <- 12
# based on Raine et al SPLAT (2017) costs

## Sequencing
# Hiseq 4000
# Single cell
# 100x2 paired end sequencing
# Average number of reads for a hiseq (Mb)
# based on sequencing coverage calculator : https://support.illumina.com/downloads/sequencing_coverage_calculator.html
READS <- 62500
# Cost per lane
LANE_COST <- 2700
# Take from Eec-Seq Paper
```
  
### Target seqeuncing Depth and Number of Samples
```{r}
# Initial 40x coverage per base
# for MEC-Seq
MEC_DEPTH <- c(25,30,35,40,45,50)
# for WGBS
WGBS_DEPTH <- c(25,40,50)
# Percent Dulicates (2%)
DUP <- 0.02
# Number of Samples
NUM <- c(1:200)
```

### Simple Functions
```{r}
# SIZE = size of sequencing region
# D = Depth
# DP = % duplicates
# R = # of reads from sequencing platform
sample_per_lane <- function(SIZE,D,DP,R){
  return(R / (SIZE*D + (SIZE*D*DP)))
}

# num = number of samples
# perSamplePrep = cost of library prep per sample
# perLane= number of individuals per lane
# laneCost = cost of lane of sequencing
sample_cost_calc <- function(name,num,perSamplePrep,SIZE,D,DP,R,laneCost,Trouble=0){
  #scenarios <- expand.grid(name="MEC",num=4,perSamplePrep=60,Size_Seq=EXOME_SIZE,Depth=50,Per_Duplication=0.02,Seq_Reads=READS,laneCost=2000,trouble_costs = 0)
  scenarios <- expand.grid(name=name,num=num,perSamplePrep=perSamplePrep,Size_Seq=SIZE,Depth=D,Per_Duplication=DP,Seq_Reads=R,laneCost=laneCost,trouble_costs = Trouble)
  
  scenarios$perLane <- sample_per_lane(scenarios$Size_Seq,scenarios$Depth,scenarios$Per_Duplication,scenarios$Seq_Reads)
  # Cost to prepare samples for sequencing
  scenarios$prep_cost <- scenarios$num * scenarios$perSamplePrep
  # Cost to sequencing
  scenarios$lane_cost <-  ceiling(scenarios$num/floor(scenarios$perLane))*scenarios$laneCost
  # Total cost
  scenarios$total_cost <-  scenarios$prep_cost + scenarios$lane_cost + scenarios$trouble_costs
  # Number of lanes needed (actual cal)
  scenarios$num_lanes_rounded <- ceiling(scenarios$num/floor(scenarios$perLane))
  # Remaining proportion of final lane not used due to design
  scenarios$rem_final_lane <- ceiling(scenarios$num/floor(scenarios$perLane))-scenarios$num/floor(scenarios$perLane)
  # Remaining reads in each lane (if more than one) not used optimally
  scenarios$rem_reads_per_lane <-  ((scenarios$perLane %% floor(scenarios$perLane))/scenarios$perLane) * READS
  # Expected depth given remaining reads in lan
  scenarios$actual_pred_depth <- scenarios$Depth + floor(scenarios$rem_reads_per_lane / (scenarios$Size_Seq + (scenarios$Size_Seq*scenarios$Per_Duplication)))
  # cumulative reads left over among all lanes 
  scenarios$cum_rem_reads_per_lane_cum <- scenarios$rem_reads_per_lane * scenarios$num_lanes_rounded 
  
  return(
    scenarios
  )
}

```

# Cost Calculations

## MEC-Seq
```{r}
sample_range_MEC <- sample_cost_calc("MEC",NUM,MEC_PSAMPLE,EXOME_SIZE,MEC_DEPTH,DUP,READS,LANE_COST, 1500)

## Total Costs
ggplot(sample_range_MEC,aes(x=num,y=total_cost,group=Depth,colour=as.factor(Depth))) + geom_line() + geom_abline(intercept=BUD,colour="red") + xlab("Number of Samples") + ylab("Total Cost (Dollars)")

## Lane Use Efficiency

# Per lane read efficiency
ggplot(sample_range_MEC,aes(x=Depth,y=rem_reads_per_lane)) + geom_line() + xlab("Depth") + ylab("Per Lane Utilization")
ggplot(sample_range_MEC,aes(x=num,y=rem_reads_per_lane,group=Depth,colour=as.factor(Depth))) + geom_line() + xlab("Number of Samples") + ylab("Per Lane Utilization")
# Cumulative
ggplot(sample_range_MEC,aes(x=num,y=cum_rem_reads_per_lane_cum,group=Depth,colour=as.factor(Depth))) + geom_line() + xlab("Number of Samples") + ylab("Cumulative under utilized reads (in mB)")

## Number of Lanes
ggplot(sample_range_MEC,aes(x=num,y=num_lanes_rounded,group=Depth,colour=as.factor(Depth))) + geom_line() + xlab("Number of Samples") + ylab("Number of Sequencing Lanes")

## Number of actual reads
ggplot(sample_range_MEC,aes(x=num,y=actual_pred_depth-Depth,group=Depth,linetype=as.factor(Depth),colour=as.factor(Depth))) + geom_line() + xlab("Number of Samples") + ylab("Expected number of reads more than target sequence depth") + ylim(0,40)
```

## WGBS 
```{r}
sample_range_WGBS <- sample_cost_calc("WGBS",NUM,WGBS_PSAMPLE,GENOME_SIZE,WGBS_DEPTH,DUP,READS,LANE_COST)

# Total Costs
ggplot(sample_range_WGBS,aes(x=num,y=total_cost,group=Depth,colour=as.factor(Depth))) + geom_line() + geom_abline(intercept=BUD,colour="red") + xlab("Number of Samples") + ylab("Total Cost (Dollars)") + ylim(0,40000)
```

# Summary
```{r}
#Combine MEC and WGBS data
sample_range <-rbind(sample_range_MEC,sample_range_WGBS)
# Plot total costs together
ggplot(sample_range,aes(x=num,y=total_cost,group=interaction(name,Depth),linetype=name,colour=as.factor(Depth))) + geom_line() + geom_abline(intercept=BUD,colour="red") + xlab("Number of Samples") + ylab("Total Cost (Dollars)") + ylim(0,40000)


sample_under_budget <- sample_range[sample_range$total_cost < BUD,]

#Summarizes the total number of samples that can be run under a specific budget depending on specific method and depth
sample_under_budget %>% group_by(name,Depth) %>%
  summarise(Seq_Size = max(Size_Seq),lanes_needed = max(num_lanes_rounded),
            TroubleShooting=max(trouble_costs),Library_Prep=max(prep_cost),Sequencing_Cost=max(lane_cost),Total_Cost=max(total_cost),
            sample_number = max(num)) -> sample_table

sample_table$sample_per_lane <- sample_table$sample_number/sample_table$lanes_needed
sample_table$predict_depth_per_sample <- READS / (sample_table$sample_per_lane*sample_table$Seq_Size + (sample_table$sample_per_lane*sample_table$Seq_Size*DUP))

# Print as table
kable(sample_table) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

# DNA Methylation Inheritance in Oysters

## Main Questions  

**General**: Is there evidence of inherited DNA methylation from parent to offspring?  

**Q1**: Does the Methylome of offspring from exposed parents look more similar to exposed parents or to control parents?
  * **Hyp**: Methylome that looks more similar (along a )

**Thoughts from Larval Cross Experiment**  
* The best block for crosses was B1.  
  * 
This is al 

## With MEC-Seq  

**Max Samples**: 84 (50x coverage)

## With WGBS  
 **Max Samples**: 18 samples (44x coverage) or 27 samples (29x coverage)








