---
title: "Compare Salmon runs on sample 17005"
author: "adowneywall"
date: "5/12/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
dir <- "/home/downeyam/Github/salmon_tutorial/test_out/"
ls_files <- list.files(dir)
(files_input <- file.path(dir,ls_files,"/quant.sf",fsep = ""))

trans <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/transcriptome_table.RData")
transFromGenome <- readRDS("/home/downeyam/Github/2017OAExp_Oysters/input_files/transcriptome_fromGenome_table.RData")

tr2gene_transFromGenome <- data.frame(TXNAME = transFromGenome$fullID,GENEID=transFromGenome$location)
tr2gene$TXNAME <- as.character(tr2gene$TXNAME)
tr2gene$GENEID <- as.character(tr2gene$GENEID)
head(tr2gene_transFromGenome)

tr2gene_trans <- data.frame(TXNAME = trans$fullID,GENEID=trans$location)
tr2gene$TXNAME <- as.character(tr2gene$TXNAME)
tr2gene$GENEID <- as.character(tr2gene$GENEID)
head(tr2gene_trans)
```

```{r}
names(files_input) <- paste0("sample", 1:6)
# Using transcriptome from Genome
txi_transFromGenome_genes <- tximport(files_input[c(1,3,5)],countsFromAbundance = "scaledTPM",
                                      type = "salmon",
                                      tx2gene = tr2gene_transFromGenome)
txi_transFromGenome_trans <- tximport(files_input[c(1,3,5)],
                                      type = "salmon",
                                      txOut = TRUE)

# Using transcriptome
txi_tr2gene_genes <- tximport(files_input[c(2,4,6)],countsFromAbundance = "scaledTPM",
                                      type = "salmon",
                                      tx2gene = tr2gene_trans)
txi_tr2gene_trans <- tximport(files_input[c(2,4,6)],
                                      type = "salmon",
                                      txOut = TRUE)
```

```{r}

tfg_counts <- data.frame(txi_transFromGenome_genes$counts)
length(tfg_counts[,1])
sum(tfg_counts[,1])
sum(tfg_counts[,2])
sum(tfg_counts[,3])

cor(tfg_counts[,1],tfg_counts[,2])
cor(tfg_counts[,2],tfg_counts[,3])
cor(tfg_counts[,1],tfg_counts[,3])

t_counts <- data.frame(txi_tr2gene_genes$counts)
length(t_counts[,1])
sum(t_counts[,1])
sum(t_counts[,2])
sum(t_counts[,3])

cor(t_counts[,1],t_counts[,2])
cor(t_counts[,2],t_counts[,3])
cor(t_counts[,1],t_counts[,3])

t_counts$names <- row.names(t_counts)
tfg_counts$names <- row.names(tfg_counts)

# Merging the two Salmon Index version runs after transcripts have been aggreated to genes
comb <- merge(tfg_counts,t_counts,by="names")
colnames(comb) <- c("Locus",ls_files)

# Correlations for all possible samples
cor(comb$`17005_fromGenome`,comb$`17005_fromRawFiles_F`)
cor(comb$`17005_fromGenome`,comb$`17005_fromRawFiles_F2_fromGenomeIndex`)

#Summation
sum(comb$`17005_fromGenome`)
sum(comb$`17005_fromRawFiles_F`)
sum(comb$`17005_fromRawFiles_F2_fromGenomeIndex`)
sum(comb$`17005_fromRawFiles_R1`)
sum(comb$`17005_fromRawFiles_R1_fromGenomeIndex`)
sum(comb$`17005_I2`)

# Read in original gene count matrix
GeneCounts <- as.matrix(read.delim("/home/downeyam/Github/2017OAExp_Oysters/results/C_virginica_gene_count_final.txt",sep = " "))
# Simplify it into a single data frame
countMat_STAR <- data.frame(Locus=row.names(GeneCounts),counts=GeneCounts[,1])
# MErge original star mapped counts with those from salmon
comb_Mappers <- merge(comb,countMat_STAR,by="Locus")

# Comparing sum counts from one salmon run version vs. those generated by star
sum(comb_Mappers$counts)
sum(comb_Mappers$`17005_fromGenome`)
sum(comb_Mappers$`17005_fromRawFiles_F`)
sum(comb_Mappers$`17005_fromRawFiles_F2_fromGenomeIndex`)
sum(comb_Mappers$`17005_fromRawFiles_R1`)
sum(comb_Mappers$`17005_fromRawFiles_R1_fromGenomeIndex`)
sum(comb_Mappers$`17005_I2`)

comb_Mappers$Locus[order(comb_Mappers$`17005_fromGenome`)][1:100]
comb_Mappers$Locus[order(comb_Mappers$`17005_fromRawFiles_F`)][1:100]

cor(comb_Mappers$`17005_fromGenome`,comb_Mappers$counts)
# This is roughly the same for all correlation comparisons between the salmon generated data and the STAR pipeline

c1 <- (comb_Mappers$`17005_fromGenome`/sum(comb_Mappers$`17005_fromGenome`))
c2 <- (comb_Mappers$counts/sum(comb_Mappers$counts))
plot(c1 ~ c2)

length(comb_Mappers$`17005_fromGenome`[comb_Mappers$`17005_fromGenome`<10])
```
  


