---
title: "Untitled"
author: "KE Lotterhos"
date: "10/10/2019"
output: html_document
---

setwd("~/Documents/GitHub/2017OAExp_Oysters/markdown_files/compareGonadMantle")

```{r, install}
packages_needed <- c("scales", "biomartr", "rio", "tidyverse", "googledrive","data.table", "matrixStats"
                     )

for (i in 1:length(packages_needed)){
  if(!(packages_needed[i] %in% installed.packages())){install.packages(packages_needed[i])}
}


for (i in 1:length(packages_needed)){
  library( packages_needed[i], character.only = TRUE)
}

bioc_packages_needed <- c("karyoploteR", "Biostrings", "biomaRt")

if(length(bioc_packages_needed)>0){
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  for (i in 1:length(packages_needed)){
     if(!(bioc_packages_needed[i] %in% installed.packages())){BiocManager::install(packages_needed[i])
     }
  }
}


for (i in 1:length(bioc_packages_needed)){
  library(bioc_packages_needed[i], character.only = TRUE)
}
```

ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/022/765/GCF_002022765.2_C_virginica-3.0/GCF_002022765.2_C_virginica-3.0_genomic.gff.gz

```{r, load gff}
if(!(getwd()=="/Users/lotterhos/Documents/GitHub/2017OAExp_Oysters/markdown_files/compareGonadMantle")){setwd("/Users/lotterhos/Documents/GitHub/2017OAExp_Oysters/markdown_files/compareGonadMantle")}


if(!file.exists("_ncbi_downloads")){
  gff.file <- getGFF(organism="Crassostrea virginica", gunzip=TRUE)
}
#gff.file <-"./ref_C_virginica-3.0_top_level.gff3"
gff.file <- "_ncbi_downloads/annotation/Crassostrea_virginica_genomic_refseq.gff"

gff.sorted <- "_ncbi_downloads/gffsorted.gff"
if(!(file.exists(gff.sorted))){
  system(paste("sortBed -i", gff.file,">", gff.sorted))
}


header.lines <- readLines(gff.file)

#The lines with the standard chromosomes start with "##sequence-region ".
#Select them.
ll <- header.lines[grepl(header.lines, pattern = "##sequence-region ")]


#split them by space, and create a data.frame
gg <- data.frame(do.call(rbind, strsplit(ll, split = " ")))
gg[,3] <- as.numeric(as.character(gg[,3]))
gg[,4] <- as.numeric(as.character(gg[,4]))

CV.genome <- toGRanges(gg[,c(2,3,4)])

```




```{r, exon density}
kp <- plotKaryotype(genome=CV.genome)
kpAddCytobandsAsLine(kp)

#features_df <- read_gff(gff.file)
#head(features_df)
features <- rtracklayer::import(gff.file)
head(features)
seqnames(features)
chrs <- as.character(seqnames(features)@values[1:10])

levels(features$type)
genes <- features[features$type=="gene" ,]
exons <- features[features$type=="exon",]

#gff.file <-"./HSP.gff3"
#features <- import(gff.file)
#HSP <- features[features$type=="exon"]
#HSP <- reduce(HSP)

exons.sc <-IRanges::reduce(exons)
genes.sc <-IRanges::reduce(genes)

kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL)
kpAddCytobandsAsLine(kp)
kpPlotRegions(kp, data=genes)

kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type=2, chromosomes = chrs)
kpAddCytobandsAsLine(kp)
kpPlotRegions(kp, data=genes[strand(genes)=="+"], avoid.overlapping = FALSE)
kpPlotRegions(kp, data=genes[strand(genes)=="-"], avoid.overlapping = FALSE, data.panel=2)


kp <- plotKaryotype(genome=CV.genome[1:10], ideogram.plotter = NULL, chromosomes=chrs)
#kpAddCytobandsAsLine(kp)
kpPlotDensity(kp, data=exons,window.size = 100000, col=adjustcolor("blue", 0.5), lwd=0.5, border=adjustcolor("blue", 0.5), r0=0, r1=1)
kpPlotDensity(kp, data=genes,window.size = 100000, col=adjustcolor("red", 0.3), lwd=0.5, data.panel=2, r0=0, r1=-1)

kp <- plotKaryotype(genome=CV.genome[1:10], ideogram.plotter = NULL, chromosomes=chrs, plot.type=2)
kpAddCytobandsAsLine(kp)
kpPlotDensity(kp, data=exons,window.size = 100000, col=adjustcolor("blue", 0.5), lwd=0.5, border=adjustcolor("blue", 0.5), data.panel=1)
kpPlotDensity(kp, data=genes,window.size = 100000, col=adjustcolor("red", 0.3), lwd=0.5, data.panel=2) #this should plot on the bottom, but doesn't seem to work

```

The first three required BED fields are:

chrom - The name of the chromosome (e.g. chr3, chrY, chr2_random) or scaffold (e.g. scaffold10671).
chromStart - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered 0.
chromEnd - Ending position



### Load methylation data for mantle

Data from Alan Downey-Wall 2017 exposure experiment

`CG_unstranded_BetaMatrix.RData` represents the beta values for each individual at all potential CpG sites based on the oyster genome. In this file i destranded paired CpG cytosines prior to calculating the beta value. The two important complimentary files to this one are  `CG_unstranded_summaryTable.RData` for relevant CpG position information and `CG_unstranded_TotalCountMatrix.RData` for the coverage.

These files contain all cpgs identified by Bismarck using the oyster genome. It has not been filtered.

Remove 17099 because of low coverage and high % duplicated reads.

```{r load mantle data}

list.files("~/Google Drive/LotterhosLab/2017-2020-Oyster OA Epigenetics/2017_CV_AdultExposureExp/Data/DNAm/Data/")
mantle <- readRDS("~/Google Drive/LotterhosLab/2017-2020-Oyster OA Epigenetics/2017_CV_AdultExposureExp/Data/DNAm/Data/CG_unstranded_BetaMatrix.RData")
colnames(mantle) <- paste0("s", colnames(mantle))
head(mantle)

mantle1 <- as.data.frame(mantle) # takes afew seconds
#mantle1 <- data.table(mantle) # gives error with dplyr functions
head(mantle1)

mantle2 <- mantle1 %>% dplyr::select(-c(s17099)) %>% dplyr::mutate(betamean = rowMeans(mantle1))
mantle2$pos <- rownames(mantle)
head(mantle2)
rm(mantle, mantle1)

mantle_pos <- readRDS("~/Google Drive/LotterhosLab/2017-2020-Oyster OA Epigenetics/2017_CV_AdultExposureExp/Data/DNAm/Data/CG_unstranded_summaryTable.RData")
names(mantle_pos)[1:3] <- c("chrom", "chromStart", "chromEnd")
  # bedfile naming system - not sure yet if this is necessary to work with ranges
mantle_pos$pos <- paste(mantle_pos$chrom, mantle_pos$chromStart, sep="_")
head(mantle_pos)

mantle_counts <- readRDS("~/Google Drive/LotterhosLab/2017-2020-Oyster OA Epigenetics/2017_CV_AdultExposureExp/Data/DNAm/Data/CG_unstranded_TotalCountMatrix.RData")
head(mantle_counts)

mantle_info <- readRDS("~/Google Drive/LotterhosLab/2017-2020-Oyster OA Epigenetics/2017_CV_AdultExposureExp/Data/meta/metadata_20190811.Rdata")
head(mantle_info)



mantle_plot <- full_join(mantle_pos, mantle2)
head(mantle_plot)
dim(mantle_plot)

#kp <- plotKaryotype(genome=CV.genome[1:10], ideogram.plotter = NULL, chromosomes=chrs)
#kpAddCytobandsAsLine(kp)
# kp
# kpLines(kp, x=mantle_plot$chromStart, y=mantle_plot$betamean, data.panel = 1, chr=mantle_plot$chrom) DO NOT DO THIS - WAAAY TO MUCH INFO TO PLOT
#kpAxis(kp, data.panel=1)

```


### Calculate CV in treatment means for each CpG in mantle
```{r}
which(colnames(mantle_plot) %in% gsub("RNA", "s", mantle_info$sample_name[mantle_info$treatment==400 & mantle_info$Day==9]))

mantle_plot$mean_400_day9 <- rowMeans(mantle_plot[,which(colnames(mantle_plot) %in% gsub("RNA", "s", mantle_info$sample_name[mantle_info$treatment==400 & mantle_info$Day==9]))])
mantle_plot$mean_400_day80 <- rowMeans(mantle_plot[,which(colnames(mantle_plot) %in% gsub("RNA", "s", mantle_info$sample_name[mantle_info$treatment==400 & mantle_info$Day==80]))])
mantle_plot$mean_2800_day9 <- rowMeans(mantle_plot[,which(colnames(mantle_plot) %in% gsub("RNA", "s", mantle_info$sample_name[mantle_info$treatment==2800 & mantle_info$Day==9]))])
mantle_plot$mean_2800_day80 <- rowMeans(mantle_plot[,which(colnames(mantle_plot) %in% gsub("RNA", "s", mantle_info$sample_name[mantle_info$treatment==2800 & mantle_info$Day==80]))])
  
head(mantle_plot)

#this is slow, it takes a few minutes
  # sd_trt_means <- apply(mantle_plot[,c("mean_400_day9", "mean_400_day80", "mean_2800_day9", "mean_2800_day80")], 1, sd)

# this is so much faster!
sd_trt_means <- rowSds(as.matrix(mantle_plot[,c("mean_400_day9", "mean_400_day80", "mean_2800_day9", "mean_2800_day80")]))
hist(sd_trt_means)

mantle_plot$CV_trt_means <-   sd_trt_means/mantle_plot$betamean

hist(mantle_plot$CV_trt_means)
h<- ggplot(mantle_plot, aes(betamean, CV_trt_means))
# this takes a minute but it works
h + geom_hex()

mantle_plot$CV_trt_means[mantle_plot$betamean<0.05] <- NA
  # check if large CV values were coming from low methylated loci
hist(mantle_plot$CV_trt_means)
head(mantle_plot)

h<- ggplot(mantle_plot, aes(betamean, CV_trt_means))
# this takes a minute but it works
h + geom_hex()
# note how that high density area is gone

h<- ggplot(mantle_plot, aes(sd_trt_means, CV_trt_means))
# this takes a minute but it works
h + geom_hex()
#


```

## Summarize features with Bedtools

```{r}
list.files()
#this takes a few secs
mantle_bed <- "bedfiles/mantle_meanBeta.bed"
whichcols <- c("chrom", "chromStart", "chromEnd", "betamean",  "mean_400_day9", "mean_400_day80", "mean_2800_day9", "mean_2800_day80")

# This takes a minute
write_delim(mantle_plot[,whichcols], mantle_bed, col_names =FALSE , delim="\t")
write_delim(data.frame(t(whichcols)), paste0(mantle_bed, ".header.txt"), col_names=FALSE)

system(paste("head", mantle_bed))

gff.sorted
system(paste0("head -n30 ",gff.sorted))
#bedtools map requires each input file to be sorted by genome coordinate.

# This line calculates the mean of betamean (average methylation across all individuals) for each feature
## Note hard coding for column 4 betamean
# Because what we are inputting are methylation calls for every CpG in the genome, we get teh count of CpGs for each feasture in the output
  system(paste0("bedtools map -a ", gff.sorted, " -b ", mantle_bed, " -c 4 -o mean,count > bedfiles/meanBetaPerFeature.txt"))
  system("head bedfiles/meanBetaPerFeature.txt")
  write_delim(data.frame(t(c("chrom", "source", "type", "chromStart", "chromEnd", "score", "strand", "phase", "attributes", "mean_betamean", "numCpGs"))), 
              paste0("bedfiles/meanBetaPerFeature.txt.header.txt"), col_names=FALSE)
  #seqid - name of the chromosome or scaffold; chromosome names can be given with or without the 'chr' prefix. Important note: the seq ID must be one used within Ensembl, i.e. a standard chromosome name or an Ensembl identifier such as a scaffold ID, without any additional content such as species or assembly. See the example GFF output below.
  #source - name of the program that generated this feature, or the data source (database or project name)
  #type - type of feature. Must be a term or accession from the SOFA sequence ontology
  #start - Start position of the feature, with sequence numbering starting at 1.
  #end - End position of the feature, with sequence numbering starting at 1.
  #score - A floating point value.
  #strand - defined as + (forward) or - (reverse).
  #phase - One of '0', '1' or '2'. '0' indicates that the first base of the feature is the first base of a codon, '1' that the second base is the first base of a codon, and so on..
  #attributes

# This line calculates the mean of betamean (average methylation across all individuals) for each feature  
## Note hard coding for column 5 CV_trt_means
# Note that I previously removed the CpGs with very low differences in mehtylation across treatment but very high CV
# because of the division by a very low mean methylation (filtered betamean > 0.05, see last chunk of code)
  mantle_bed2 <- "bedfiles/mantle_meanBeta2_CV.bed"
  whichcols2 <- c("chrom", "chromStart", "chromEnd", "betamean", "CV_trt_means", "mean_400_day9", "mean_400_day80", "mean_2800_day9", "mean_2800_day80")
  write_delim(mantle_plot[which(!(is.na(mantle_plot$CV_trt_means))),whichcols2], mantle_bed2, col_names =FALSE , delim="\t")
  write_delim(data.frame(t(whichcols2)), paste0(mantle_bed2, ".header.txt"), col_names=FALSE)  
    
  system(paste0("bedtools map -a ", gff.sorted, " -b ", mantle_bed2, " -c 4,5,4 -o mean,mean,count > bedfiles/mean_betamean_CVTrtMeans_PerFeature_filteredBetameanGT05.txt"))
    write_delim(data.frame(t(c("chrom", "source", "type", "chromStart", "chromEnd", "score", "strand", "phase", "attributes", "mean_betamean_filteredBetameanGT05", "mean_CVTrtMeans_filteredBetameanGT05", "numCpGs_filteredBetameanGT05"))), 
              paste0("bedfiles/mean_betamean_CVTrtMeans_PerFeature_filteredBetameanGT05.header.txt"), col_names=FALSE)
  system("head bedfiles/mean_betamean_CVTrtMeans_PerFeature_filteredBetameanGT05.txt")


  # if you want to know if your CpG is in a feature, use this code
  # But replace gff.sorted with a sorted file of the features (e.g. genes only)
  # system(paste0("bedtools annotate -i ", mantle_bed, " -files ", gff.sorted, " | head"))
    # wget http://kitt.uri.edu/sorted.ref3.0*
    # sorted.ref3.0.CDS.bed
    # sorted.ref3.0.exon.bed
    # sorted.ref3.0.UTR.bed
    # sorted.ref3.intergenic.bed
    # sorted.ref3.0.gene.bed
  
  system(paste0("bedtools intersect -wa -wb -a ", mantle_bed ," -b ", gff.sorted, " > bedfiles/CpGsAnnotated.bed"))
  system("head bedfiles/CpGsAnnotated.bed")
   
```



### Load methylation data for gonads

Data from Yaamini V 2016 exposure experiment. Gonads were analyzed from 10 individuals. These files were downloaded from Robert's lab server.

There are 10 files (1 for each individual) and each file was filtered to CpGs with 5x coverage for that individual. There are 5 columns in each file:
`("chrom", "chromStart", "chromEnd", "beta", "coverage")`
where `beta` is the percent methylation and `coverage` is the coverage for that individual.


* MBD Cv 31 =1
* MBD Cv 32 = 2
* MBD Cv 33 = 3
* MBD Cv 35 = 4
* MBD Cv 36 = 5
* MBD Cv 103 = 6
* MBD Cv 104 = 7
* MBD Cv 105 = 8
* MBD Cv 106 = 9
* MBD Cv 108 = 10


```{r}
link <- "https://gannet.fish.washington.edu/spartina/2019-10-15-paper-gonad-meth/analyses/2019-03-07-IGV-Verification/"
file_list <- paste0("zr2096_",1:10,"_s1_R1_val_1_bismark_bt2_pe.deduplicated.bismark.cov_5x.percentMeth.cov")

if(!file.exists(bedfile)){
  dir.create("bedfiles")
}

### Download the files
for (i in 1:length(file_list)){
  if(!file.exists(paste0("bedfiles/", file_list[i]))){
    download.file(paste0(link, file_list[i]), paste0("bedfiles/", file_list[i]))
  }
}


### Read the files
for (i in 1:length(file_list)){
  gonads <- read.table(paste0("bedfiles/", file_list[i]))
  cat(i, "gonads", dim(gonads), '\n')
  colnames(gonads) <- c("chrom", "chromStart", "chromEnd", paste0("beta_",i), paste0("coverage_",i))
  if(i==1){
    gonadsall <- gonads
  }else{
    gonadsall <- full_join(gonadsall, gonads)
    cat(i, "gonadsall", dim(gonadsall), '\n')
  }
}

head(gonadsall)
hist(log10(gonads$coverage))
min(gonads$coverage)

# For now I am going to assume anything not captured by MBD has 0% methylation:
gonadsall[is.na(gonadsall)] <- 0
head(gonadsall)

whichbeta <- grep("beta", colnames(gonadsall))

gonadsall$beta_gonad <- rowMeans(gonadsall[,whichbeta])
gonadsall$beta_gonad <- gonadsall$beta_gonad/100
hist(gonadsall$beta_gonad)
```


Let's summarize this beta data and plot it for a custom window size.

This code uses the split, apply, combine approach to produce data summaries.

For more information, check out Vince Buffalo's Bioinformatics Book.

```{r split and summarize functions}
beta_split_chrom_mantle <- split(mantle_plot$beta, mantle_plot$chrom)
str(beta_split_chrom_mantle)

beta_pos_chrom_mantle  <- split(mantle_plot$chromStart, mantle_plot$chrom)
str(beta_pos_chrom_mantle)

beta_split_chrom_gonads <- split(gonadsall$beta_gonad, gonadsall$chrom)
str(beta_split_chrom_gonads)

beta_pos_chrom_gonads  <- split(gonadsall$chromStart, gonadsall$chrom)
str(beta_pos_chrom_gonads)

summ_beta <- function(beta, pos, window1=10^5){
  bins <- cut(pos, breaks=seq(1, max(pos), by = window1))
  #beta_split <- split(beta, bins)
  out <- do.call(rbind, lapply(split(beta, bins), summary))
  out2 <- as.data.frame(out)
  out2$mid <- seq(window1/2, (max(pos)-window1/2), by = window1)
  out2$window <- rownames(out)
  return(out2)
}
```

Make a plot with Mantle Methylation on the top and Gonad on the bottom
```{r, mantle exon plot}
png("Mantle2017Top_Gonad2016Bottom.png", width=12, height=8, units="in", res=400)
  kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, chromosomes=chrs, 
                      plot.type=2)
  kpAddCytobandsAsLine(kp)
  mantle_summ <- list()
  gonad_summ <- list()
  
  for (i in 2:length(beta_split_chrom_mantle)){
    a<- 
  summ_beta(beta_split_chrom_mantle[[i]], beta_pos_chrom_mantle[[i]])
  head(a)
  kpLines(kp, x=a$mid, y=a$"1st Qu.", data.panel = 1, chr=names(beta_split_chrom_mantle[i]), col="grey")
  kpLines(kp, x=a$mid, y=a$"3rd Qu.", data.panel = 1, chr=names(beta_split_chrom_mantle[i]), col="grey")
  kpLines(kp, x=a$mid, y=a$Mean, data.panel = 1, chr=names(beta_split_chrom_mantle[i]), col="darkviolet")
  mantle_summ[[i]] <- a
  
   b<- 
  summ_beta(beta_split_chrom_gonads[[i]], beta_pos_chrom_gonads[[i]])
  head(b)
  kpLines(kp, x=b$mid, y=b$"1st Qu.", data.panel = 2, chr=names(beta_split_chrom_gonads[i]), col="grey")
  kpLines(kp, x=b$mid, y=b$"3rd Qu.", data.panel = 2, chr=names(beta_split_chrom_gonads[i]), col="grey")
  kpLines(kp, x=b$mid, y=b$Mean, data.panel = 2, chr=names(beta_split_chrom_gonads[i]), col="chocolate3")
  gonad_summ[[i]] <- b
  }
  
  
  kpAxis(kp, data.panel=1, labels = c("", 0.5, 1), cex=0.5)
  kpAxis(kp, data.panel=2, labels = c("", 0.5, 1), cex=0.5)
  
  #kpPlotDensity(kp, data=exons,window.size = 100000, col=adjustcolor("blue", 0.5), lwd=0.5, border=adjustcolor("blue", 0.5), data.panel=2, r0=0, r1=1)
dev.off()  
```

Overall correspondence between mantle and gonad at the regional level
```{r}
pdf("Mantle_vs_Gonad_10KBwindows.pdf", width=6, height=8)
  #str(mantle_summ)
  #str(gonad_summ)
  par(mar=c(2,2,0,0), mfrow=c(5,2), oma=c(3,3,0.1,0.1))
  for (i in 2:length(mantle_summ)){
    plot(mantle_summ[[i]]$Mean, gonad_summ[[i]]$Mean, bty="n", xlab="", ylab="", xlim=c(0,1), ylim=c(0,1), col=adjustcolor("darkorchid4",0.4))
    abline(0,1, col="darkgrey")
    text(0, 0.9, names(beta_split_chrom_gonads[i]), adj=0)
    text(0.9, 0.1, round(cor(mantle_summ[[i]]$Mean, gonad_summ[[i]]$Mean), 2))
  }
  mtext("Mean Mantle Methylation (100 KB windows)", outer=TRUE, side=1, line=1)
  mtext("Mean Gonad Methylation (100 KB windows)", outer=TRUE, side=2, line=1)
dev.off()
```




Before comparing Alan's and Yaamini's data, I want to check to make sure their positions line up (and they don't):

```{r}
dim(mantle_plot)
dim(gonadsall)

levels(gonadsall$chrom)

gonadsall %>% filter(chrom=="NC_007175.2")  %>% arrange(chromStart) %>% head(20)
mantle_plot %>% filter(chrom=="NC_007175.2") %>% arrange(chromStart) %>% head(20)

gonadsall %>% filter(chrom=="NC_035780.1")  %>% arrange(chromStart) %>% head(20)

# The left join leaves us with sites in the gonads data only
allcomp <- full_join(gonadsall[gonadsall$chrom=="NC_007175.2",c("chrom", "chromStart", "chromEnd", "beta_gonad")],
                     mantle_plot[gonadsall$chrom=="NC_007175.2",c("chrom", "chromStart", "chromEnd", "C_motif", "Tri_motif", "pos", "betamean")])  
allcomp <- allcomp[order(allcomp$chromStart),]
head(allcomp, 50)

dim(allcomp)

h <- ggplot(allcomp, aes(allcomp$beta_gonad, allcomp$betamean))
h + geom_bin2d()

```


In general, Yaamini has some data 

What's more, the data Yaamini sent me appear to have different positions mapped in different individuals.

Alan Downey-Wall 10:45

So I used coverage2cytosine from bismark to determine the coverage for each cytosine, which takes the chromosome coordinates from the genome file:

Text from there github: Starting from the coverage output, the Bismark methylation extractor can optionally also output a genome-wide cytosine methylation report. The module coverage2cytosine (part of the Bismark package) may also be run individually. It is also sorted by chromosomal coordinates but also contains the sequence context and is in the following format:
`<chromosome> <position> <strand> <count methylated> <count unmethylated> <C-context> <trinucleotide context>`

The main difference to the bedGraph or coverage output is that every cytosine on both the top and bottom strands will be considered irrespective of whether they were actually covered by any reads in the experiment or not. For this to work one has to also specify the genome that was used for the Bismark alignments using the option --genome_folder <path>. As for the bedGraph mode, this will only consider cytosines in CpG context by default but can be extended to cytosines in any sequence context by using the option --CX (cf. Appendix (III)). Be aware though that this might mean an output with individual lines for more than 1.1 billion cytosines for any large mammalian genome...

Here is a link to the code I used to destrand: https://github.com/epigeneticstoocean/2017OAExp_Oysters/blob/master/markdown_files/DNAm/cluster_scripts/R/04_countMatrixFromCytoSummary.R

I defaulted to having the start position for the cytosine on the positive strand, and the end position for the cytosine on the negative strand.

```{r}
sum(!(mantle_plot$chromStart < mantle_plot$chromEnd))
sum(!(mantle_plot$chromStart - mantle_plot$chromEnd)== -1)
```








bedfile <- "bedfiles/2019-04-09-All-5x-CpGs.bedgraph"
if(!file.exists(bedfile)){
  dir.create("bedfiles")
  download.file("https://gannet.fish.washington.edu/spartina/2018-10-10-project-virginica-oa-Large-Files/2019-05-13-Yaamini-Virginica-Repository/analyses/2019-03-18-Characterizing-CpG-Methylation/2019-04-09-All-5x-CpGs.bedgraph", "bedfiles/2019-04-09-All-5x-CpGs.bedgraph")
}

# gonads <- rtracklayer::import(bedfile) # error

system("head bedfiles/2019-04-09-All-5x-CpGs.bedgraph")
gonads <- read.table("bedfiles/2019-04-09-All-5x-CpGs.bedgraph")

head(gonads)
```


coverage.1 <- import.bed("./Total_EC.2c.bed")
test.cov <- toGRanges(coverage.1)

coverage.2 <- import.bed("./m4q4sc.bed")
test.cov2 <- toGRanges(coverage.2)



kpPlotCoverage(kp, data=test.cov)

kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL,plot.type=2, chromosomes = "NC_035780.1")
kpAddCytobandsAsLine(kp)
kpPlotDensity(kp, data=exons.sc,window.size = 1000, chromosomes="NC_035780.1")
#kpPoints(kp, chr="NC_007175.2", data=test.cov, y=score, data.panel = 2)
kpPlotCoverage(kp, data=test.cov, data.panel=2,chromosomes="NC_035780.1")


kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL,plot.type=2, chromosomes = "NC_035780.1")
kpAddCytobandsAsLine(kp)
kpPlotDensity(kp, data=exons,window.size = 1000000, chr="NC_035780.1")
kpPlotDensity(kp, data=test.cov,window.size=1000000, chr="NC_035780.1", data.panel=2)

EC_2.bam <-"./ECI_2.F.bam"

kpAddChromosomeNumbers <- function(karyoplot, ...){
  karyoplot$beginKpPlot()
  on.exit(karyoplot$endKpPlot())
  bb <- getChromosomeNamesBoundingBox(karyoplot)
  chr.labels <- seq(1,length(karyoplot$chromosomes))
  chr.labels <- paste("Chr ",chr.labels,sep="")
  x <- (bb$x0 + bb$x1)/2
  y <- (bb$y0 + bb$y1)/2
  graphics::text(x = x, y = y, labels = chr.labels, ...)
  invisible(karyoplot)
}


cbPalette <- c("#009E73","#D55E00","#56B4E9", "#0072B2","#E69F00", "#999999","#F0E442" , "#CC79A7")

pie(rep(1, length(cbPalette)), labels = sprintf("%d (%s)", seq_along(cbPalette), 
                                                cbPalette), col = alpha(cbPalette,0.9))
coverage.1 <- import.bed("./Total_EC.2c.bed")
test.cov <- toGRanges(coverage.1)

coverage.2 <- import.bed("./m4q4sc.bed")
test.cov2 <- toGRanges(coverage.2)

kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 3, chromosomes="NC_035785.1")

png(filename="kary1new.png", type="cairo",units="px", width=5600, 
    height=3000, res=300, bg="transparent")
kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 2, chromosomes=c("NC_035780.1","NC_035781.1","NC_035782.1","NC_035783.1","NC_035784.1","NC_035785.1","NC_035786.1","NC_035787.1","NC_035788.1","NC_035789.1"),labels.plotter = kpAddChromosomeNumbers)
kp$plot.params$chromosome.height=1
kp$plot.params$ideogramheight <-1
kp$plot.params$data1inmargin <- -10
kp$plot.params$data2inmargin <-30
kp$plot.params$topmargin <-0
kp$plot.params$bottommargin <-0

kpAddCytobandsAsLine(kp, lwd=5)
kpAddBaseNumbers(kp)
#kpPlotRegions(kp, data=exons.sc)
#kpPlotRegions(kp,data=HSP)
#kpPlotDensity(kp, data=HSP, col = "red", data.panel=1,window.size = 100)
#kpPlotDensity(kp, data=exons.sc,window.size = 100000)
kpPlotDensity(kp, data=test.cov,window.size=100000, col=alpha(cbPalette[4],0.9),data.panel=1, ymax=150000)
kpPlotDensity(kp, data=test.cov2,window.size = 100000, col=alpha(cbPalette[2],0.9),data.panel=2, ymax=40000, r1=0.9)
dev.off()

  png(filename="kary.black.png", type="cairo",units="px", width=5600, 
    height=3000, res=300, bg="transparent")
kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 2, col="white", chromosomes=c("NC_035780.1","NC_035781.1","NC_035782.1","NC_035783.1","NC_035784.1","NC_035785.1","NC_035786.1","NC_035787.1","NC_035788.1","NC_035789.1"),labels.plotter = kpAddChromosomeNumbers)
kp$plot.params$chromosome.height=1
kp$plot.params$ideogramheight <-1
kp$plot.params$data1inmargin <- -10
kp$plot.params$data2inmargin <-30
kp$plot.params$topmargin <-0
kp$plot.params$bottommargin <-0

kpAddCytobandsAsLine(kp, lwd=5)
kpAddBaseNumbers(kp, col="white")
#dev.off()
#kpPlotRegions(kp, data=exons.sc)
#kpPlotRegions(kp,data=HSP)
#kpPlotDensity(kp, data=HSP, col = "red", data.panel=1,window.size = 100)
#kpPlotDensity(kp, data=exons.sc,window.size = 100000, col=alpha(cbPalette[1],0.9), border="white")
#dev.off()
kpPlotDensity(kp, data=test.cov,window.size=100000, col=alpha(cbPalette[4],0.9),border="white",data.panel=2, ymax=30000)
#kpPlotDensity(kp, data=test.cov2,window.size = 100000, col="red",data.panel=1, ymax=40000, r1=0.9)

dev.off()


kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 3, chromosomes="NC_035785.1")

png(filename="kary1.black.png", type="cairo",units="px", width=5600, 
    height=3000, res=300, bg="transparent")
kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 2, col="white", chromosomes=c("NC_035780.1","NC_035781.1","NC_035782.1","NC_035783.1","NC_035784.1","NC_035785.1","NC_035786.1","NC_035787.1","NC_035788.1","NC_035789.1"),labels.plotter = kpAddChromosomeNumbers)
kp$plot.params$chromosome.height=1
kp$plot.params$ideogramheight <-1
kp$plot.params$data1inmargin <- -10
kp$plot.params$data2inmargin <-30
kp$plot.params$topmargin <-0
kp$plot.params$bottommargin <-0

kpAddCytobandsAsLine(kp, lwd=5)
kpAddBaseNumbers(kp, col="white")
#dev.off()
#kpPlotRegions(kp, data=exons.sc)
#kpPlotRegions(kp,data=HSP)
#kpPlotDensity(kp, data=HSP, col = "red", data.panel=1,window.size = 100)
#kpPlotDensity(kp, data=exons.sc,window.size = 100000)
kpPlotDensity(kp, data=test.cov2,window.size = 100000, col=alpha(cbPalette[2],0.9),data.panel=1, border="white",ymax=40000, r1=0.9)
#dev.off()
kpPlotDensity(kp, data=test.cov,window.size=100000, col=alpha(cbPalette[4],0.9),data.panel=2, border="white",ymax=30000)
dev.off()

png(filename="karynew.png", type="cairo",units="px", width=5600, 
    height=3000, res=300, bg="transparent")
kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 2, chromosomes=c("NC_035780.1","NC_035781.1","NC_035782.1","NC_035783.1","NC_035784.1","NC_035785.1","NC_035786.1","NC_035787.1","NC_035788.1","NC_035789.1"),labels.plotter = kpAddChromosomeNumbers)
kp$plot.params$chromosome.height=1
kp$plot.params$ideogramheight <-1
kp$plot.params$data1inmargin <- -10
kp$plot.params$data2inmargin <-30
kp$plot.params$topmargin <-0
kp$plot.params$bottommargin <-0

kpAddCytobandsAsLine(kp, lwd=5)
kpAddBaseNumbers(kp)
#kpPlotRegions(kp, data=exons.sc)
#kpPlotRegions(kp,data=HSP)
#kpPlotDensity(kp, data=HSP, col = "red", data.panel=1,window.size = 100)
kpPlotDensity(kp, data=exons.sc,window.size = 100000, col=alpha(cbPalette[1],0.9))
kpPlotDensity(kp, data=test.cov,window.size=100000, col=alpha(cbPalette[4],0.9),data.panel=2, ymax=150000)
#kpPlotDensity(kp, data=test.cov2,window.size = 100000, col="red",data.panel=1, ymax=40000, r1=0.9)

dev.off()

png(filename="karynewblack.png", type="cairo",units="px", width=5600, 
    height=3000, res=300, bg="transparent")
kp <- plotKaryotype(genome=CV.genome, ideogram.plotter = NULL, plot.type = 2, chromosomes=c("NC_035780.1","NC_035781.1","NC_035782.1","NC_035783.1","NC_035784.1","NC_035785.1","NC_035786.1","NC_035787.1","NC_035788.1","NC_035789.1"),labels.plotter = kpAddChromosomeNumbers,col="white")
kp$plot.params$chromosome.height=1
kp$plot.params$ideogramheight <-1
kp$plot.params$data1inmargin <- -30
kp$plot.params$data2inmargin <-10
kp$plot.params$topmargin <-0
kp$plot.params$bottommargin <-0

kpAddCytobandsAsLine(kp, lwd=5)
kpAddBaseNumbers(kp, col="white")
#kpPlotRegions(kp, data=exons.sc)
#kpPlotRegions(kp,data=HSP)
#kpPlotDensity(kp, data=HSP, col = "red", data.panel=1,window.size = 100)
kpPlotDensity(kp, data=test.cov,window.size = 100000, col=alpha(cbPalette[4],0.9),border="grey",ymax=150000)
kpPlotDensity(kp, data=test.cov2,window.size=100000, col=alpha(cbPalette[2],0.9),data.panel=2,border="grey",ymax=30000, r1=0.9)
#kpPlotDensity(kp, data=test.cov2,window.size = 100000, col="red",data.panel=1, ymax=40000, r1=0.9)

dev.off()

#kpPlotBAMDensity(kp, data = EC_2.bam, window.size = 100000, data.panel=2)




